<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>SpaMTP • SpaMTP</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png">
<link rel="icon" sizes="any" href="favicon.ico">
<link rel="manifest" href="site.webmanifest">
<script src="lightswitch.js"></script><script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="SpaMTP">
<meta name="description" content="This is an easy-to-use spatial metabolomic data analysis package that harnesses the functionality of Cardinal and Seurat. In addition to easier and quicker plotting, this package contains additional function for various downstream analyses such as differential peak analysis, merging of Metabolomic and Spatial Transcriptomic (Visium) data, m/z metabolite annotation, and others.">
<meta property="og:description" content="This is an easy-to-use spatial metabolomic data analysis package that harnesses the functionality of Cardinal and Seurat. In addition to easier and quicker plotting, this package contains additional function for various downstream analyses such as differential peak analysis, merging of Metabolomic and Spatial Transcriptomic (Visium) data, m/z metabolite annotation, and others.">
<meta property="og:image" content="https://genomicsmachinelearning.github.io/SpaMTP/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">SpaMTP</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/SpaMTP.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="articles/Mouse_Urinary_Bladder.html">Spatial Metabolomics Analysis</a></li>
    <li><a class="dropdown-item" href="articles/Multi-Omic_Mouse_Brain.html">Spatial Multi-Omics Analysis</a></li>
    <li><a class="dropdown-item" href="articles/human_brain_tissue_analysis.html">Single Cell Multi-Omics Analysis</a></li>
    <li><a class="dropdown-item" href="articles/SpaMTP_Additional_Features.html">Additional SpaMTP Features</a></li>
    <li><a class="dropdown-item" href="articles/Big_Data.html">Handling Big Datasets</a></li>
    <li><a class="dropdown-item" href="articles/Annotation_Refinement.html">Refinning Metabolite Annotations</a></li>
    <li><a class="dropdown-item" href="articles/Single_Cell_MultiOmics.html">Simulated scMulti-Omics Analysis</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Updates</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/GenomicsMachineLearning/SpaMTP" aria-label="GitHub"><span class="fa fa-github"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="spamtp-">SpaMTP <a class="anchor" aria-label="anchor" href="#spamtp-"></a>
</h1>
</div>
<!-- badges: start -->
<div class="section level2">
<h2 id="r-based-user-friendly-spatial-metabolomic-transcriptomic-and-proteomic-data-analysis-tool">
<em>R</em>-based User-Friendly Spatial Metabolomic, Transcriptomic, and Proteomic Data Analysis Tool<a class="anchor" aria-label="anchor" href="#r-based-user-friendly-spatial-metabolomic-transcriptomic-and-proteomic-data-analysis-tool"></a>
</h2>
<p><br></p>
<!-- badges: end -->
<p>SpaMTP is an R package designed for the integrative analysis of spatial metabolomics and spatial transcriptomics data. SpaMTP inherits functionalities from two well established R packages (Cardinal and Seurat) to present a user-friendly platform for integrative spatial-omics analysis. Build on the foundation of a <a href="https://satijalab.org/seurat/" class="external-link"><em>Seurat Class Object</em></a>, this package has three major functionalities which include; (1) mass-to-charge ratio (m/z) metabolite annotation, (2) various downstream statistical analysis including differential metabolite expression and pathway analysis, and (3) integrative spatial-omics analysis. In addition, this package includes various functions for data visualisation and data import/export, permitting flexible usage with other established R and Python packages.</p>
<p>Please head to the <a href="https://genomicsmachinelearning.github.io/SpaMTP/"><strong>SpaMTP Website</strong></a> for all <strong>tutorials, documentation and additional information</strong>. For a detailed explanation and to cite <em>SpaMTP</em>, please visit our pre-print available <a href="https://www.biorxiv.org/content/10.1101/2024.10.31.621429v1" class="external-link">here</a>.</p>
<p><br></p>
<p><img src="reference/figures/SpaMTP_Fig.png" alt="" style="background-color: white;"></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the current version of SpaMTP from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"GenomicsMachineLearning/SpaMTP"</span><span class="op">)</span></span></code></pre></div>
<p>For tutorials and more information please visit the <a href="https://genomicsmachinelearning.github.io/SpaMTP/">SpaMTP website</a></p>
<div class="section level3">
<h3 id="apptainer">Apptainer<a class="anchor" aria-label="anchor" href="#apptainer"></a>
</h3>
<p>To download the image:</p>
<pre><code>wget -c https://downloads.gmllab.com/SpaMTP/spamtp-1.1.sif</code></pre>
<p>Verify it download correctly:</p>
<pre><code>wget -q -O - https://downloads.gmllab.com/SpaMTP/spamtp-1.1.sif.md5 | md5sum -c -</code></pre>
<p>Or to create an Apptainer image from scratch (you need the environment-linux.yml, dependencies.R and spamtp.def files):</p>
<pre><code>apptainer build spamtp.sif spamtp.def</code></pre>
</div>
<div class="section level3">
<h3 id="contributing">Contributing<a class="anchor" aria-label="anchor" href="#contributing"></a>
</h3>
<p>To reproduce the dependencies and install SpaMTP into a conda environment:</p>
<pre><code>conda create --prefix [some-directory]/conda/spamtp python=3.10 r-base=4.4 imagemagick --y
conda activate [some-directory]/conda/spamtp
conda install -c conda-forge jupyter libconfig fontconfig freetype libtiff r-irkernel  --y
conda install --no-update-deps -c conda-forge r-here r-matrix r-Seurat r-SeuratObject r-hdf5r r-ggplot2 r-reshape2 r-patchwork fftw r-RCurl r-jpeg r-locfit r-ontologyindex --y
conda install --no-update-deps -c conda-forge r-fields r-ggdendro r-ggnewscale r-magick r-naturalsort r-pheatmap r-sf r-shinyjs r-zeallot r-r.utils --y
conda install --no-update-deps -c bioconda bioconductor-BiocGenerics bioconductor-enhancedvolcano bioconductor-biocparallel --y
conda install --no-update-deps -c bioconda bioconductor-protgenerics bioconductor-s4vectors bioconductor-biobase --y
conda install --no-update-deps -c bioconda bioconductor-edger bioconductor-fgsea bioconductor-limma bioconductor-scater bioconductor-singlecellexperiment --y
conda install --no-update-deps -c bioconda bioconductor-delayedmatrixstats bioconductor-hdf5array bioconductor-rhdf5 bioconductor-rhdf5lib --y
conda install --no-update-deps -c bioconda bioconductor-msnbase bioconductor-ebimage==4.48.0-0 --y
Rscript dependencies.R</code></pre>
</div>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/genomicsmachinelearning/SpaMTP/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/genomicsmachinelearning/SpaMTP/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing SpaMTP</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://github.com/agc888" class="external-link">Andrew Causer</a> <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0009-0001-7341-8366" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
<li>
<a href="https://github.com/BCRL-tylu" class="external-link">Tianyao Lu</a> <br><small class="roles"> Author </small>   </li>
<li>
<a href="https://github.com/ChemCharles" class="external-link">Christopher C.J. Fitzgerald</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-3935-4300" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
<li>
<a href="https://github.com/GenomicsMachineLearning" class="external-link"><img src="reference/figures/GML_invert.png" height="50" alt=""></a> <br><small class="roles"> Laboratory </small>   </li>
<li>
<a href="https://www.wehi.edu.au/laboratory/brain-cancer-research-lab/" class="external-link"><img src="https://thebraincancercentre.org.au/wp-content/uploads/2022/09/BCC-WEHI-logo-horiz-RGB3.svg" height="30" alt="WEHI Logo"></a> <br><small class="roles"> Laboratory </small>   </li>
<li>
<a href="https://imb.uq.edu.au/research-groups/nguyen" class="external-link"><img src="https://upload.wikimedia.org/wikipedia/en/8/8c/Logo_of_the_University_of_Queensland.svg" height="30" alt="University of Queensland Logo"></a> <br><small class="roles"> Funder </small>   </li>
<li>
<a href="https://www.qimrberghofer.edu.au/our-research/infection-and-inflammation/genomics-and-machine-learning-lab-2/" class="external-link"><img src="https://upload.wikimedia.org/wikipedia/commons/5/5a/Logo_of_the_QIMR.png" height="30" alt="QIMR Logo"></a> <br><small class="roles"> Funder </small>   </li>
<li>
<a href="https://www.wehi.edu.au/laboratory/brain-cancer-research-lab/" class="external-link"><img src="https://upload.wikimedia.org/wikipedia/en/7/7d/Wehi-logo-2020.png" height="30" alt="WEHI Logo"></a> <br><small class="roles"> Funder </small>   </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/agc888" class="external-link">Andrew Causer</a>, <a href="https://github.com/BCRL-tylu" class="external-link">Tianyao Lu</a>, <a href="https://github.com/ChemCharles" class="external-link">Christopher C.J. Fitzgerald</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
