---
title: "Multi-Omic Mouse Brain"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Multi-Omic Mouse Brain}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```



<br>

# SpaMTP: Multi-Omic Mouse Brain Vignette
<br>
This tutorial highlights the highlights the utility of *SpaMTP* on multi-omic data. This vignette will use mouse brain data, where both Spatial Transcriptomics (Visium) and Spatial Metabolomics (MALDI-MSI) were generated from the same tissue section. This data is publically avalible from [Vicari et al](https://doi.org/10.17632/w7nw4km7xd.1).

Using **SpaMTP** we will highlight:

* Loading Spatial Metabolomic Data
* Manually Aligning Spatial Metabolic and Transcriptomic datasets
* Mapping Aligned SM data to ST Spot Level
* Multi-Modal Data Integration of SM and ST Data
* Multi-Omic Data Visualisation
* Colocalisation of Metabolites and Genes
* Multi-Modality Pathway Analysis

Author: Andrew Causer

<br>
<br>

## Load Data

Here we will load in both our Visium and SM data! The Visium data is contained in 10X Genomic's standard format. The MALDI data is in a matrix format, where one table contains both x and y coordinates, and also intensity values for each m/z value. 


#### Install and Import *R* Libraries

First we need to import the required libraries for this analysis 

```{r warning=FALSE, message=FALSE}
#General Libraries
library(SpaMTP)
library(Cardinal)
library(Seurat)
library(dplyr)

#For plotting + DE plots
library(ggplot2)
library(EnhancedVolcano)

```

<br>

#### Load *SM* Data

As mentioned above, the format of the MSI data is in a mtx format. This is different to .ibd and .imzML files that were used in the [Mouse Bladder Vignette](https://github.com/BiomedicalMachineLearning/SpaMTP/blob/master/vignettes/Mouse_Bladder_Vignette.html). The layout of this dataset is demonstrated below:

<div style="margin-left: auto;
            margin-right: auto;
            width: 30%">

|  x  | y  | mz_1| mz_2| mz_3| ... | mz_n|        
| :--:| :---:|:---:|:---:|:---:|:---:|:---:|
 | 0 | 1 | 0|0|11|...|0|
  | 0 | 2 |0|0|0|...|0|
 | 0 | 3 |15|0|0|...|32|
  | 0 | 4 |20|0|0|...|32|
 | 0 | 5 |0|20|0|...|0|

</div>

Based on this we can read in the .csv file that contains our intensity and pixel coordinate values.


```{r, warning=FALSE, message=FALSE}
msi <- ReadSM_mtx("vignette_data_files/Mouse_Brain/SM_data/V11L12-109_B1.Visium.FMP.220826_smamsi.csv")
```

    Spliting matrix data from column 1 onwards .... 
  
    Adding Pixel Metadata ....
    
    Creating Centroids for Spatial Seurat Object ....


Great! Now lets expore our sample abit. First we will see the layout of the *SpaMTP* Seurat Object


```{r}
msi
```

 

We can see that we have about 1,538 different m/z masses across 4,914 pixels.

Lets look at what we have in our metadata!


```{r, echo=TRUE, results='hide'}
head(msi@meta.data)
```

```{r, echo=FALSE}
df <- head(msi@meta.data)
knitr::kable(df, format = "html", table.attr = "class='dataframe'",  align = "c")
```


We can see that our pixel names are a combination of x and y coordinates. Also we have 2 data columns nCount_Spatial and nFeature_Spatial were are generated when loading the data. Respectively, these represent the total intensity across all m/z per pixel, and the number of different m/z values which have an intensity > 0 per pixel. 

Lets have a look at the sample spatially -> we will just visualise the number of m/z per pixel


```{r fig.width=7, fig.height=5}
ImageFeaturePlot(msi, "nFeature_Spatial", size = 2, dark.background = FALSE) 
```


We have some pixels which dont contain any SM data, thus we will remove them first before starting any analysis.


```{r, warning=FALSE, message=FALSE, fig.width=10, fig.height=6}
msi <- subset(msi, subset = nCount_Spatial > 0)

ImageFeaturePlot(msi, "nCount_Spatial", size = 1.3, dark.background = FALSE) | ImageFeaturePlot(msi, "nFeature_Spatial", size = 1.3, dark.background = FALSE)
```

<br>

#### Load *ST* Data

Now that we have successfully loaded our metabolomic data lets load in the ST Visium data. This data can be loaded using *Seurat's* Load10X_Spatial() function.


```{r, warning=FALSE, message=FALSE}
vis <- Load10X_Spatial("vignette_data_files/Mouse_Brain/ST_data/outs/")
```


```{r}
vis 
```


The ST data has 32,285 genes and 3,120 spots. Next we will visualise the sample and compare to the MSI image above


```{r fig.width=10, fig.height=6}
SpatialFeaturePlot(vis, features = c("nCount_Spatial", "nFeature_Spatial"), pt.size.factor = 1.2)
```


In addition we can load in some metadata associated with the tissue morphology generated by the orginal publication authors. As mentioned previously the data we are analysing is from a Parkinson's mouse model, where the Dopamine neurons of one hemisphere of the brain have been lesioned. The metadata loaded below contains infomation surround tissue structures and the lession site. 

Lets now load this metadata and add it to our Visium Seurat object


```{r}
## Read in metadata files
lesion <- read.csv("vignette_data_files/Mouse_Brain/ST_data/outs/lesion.csv", row.names = 1)
region <- read.csv("vignette_data_files/Mouse_Brain/ST_data/outs/region.csv", row.names = 1)
region_loupe <- read.csv("vignette_data_files/Mouse_Brain/ST_data/outs/RegionLoupe.csv", row.names = 1)

## Combine into one dataframe
annotations <- cbind(lesion,region,region_loupe)

## Add data to metadata slot
vis@meta.data[colnames(annotations)] <- annotations[colnames(annotations)]
```

Now lets plot the results. We can use this information later for tissue alignment and removing poor quality spots sitting outside the tissue section


```{r, fig.width=20, fig.height=6}
SpatialDimPlot(vis, group.by = c("lesion","region", "RegionLoupe"))
```

<br>

## Data preprocessing

#### Annotate m/z masses

Now that we have succesffully loaded our ST and SM data we can now run some general preprocessing to remove any poor quality spots/pixel and genes/metabolites. 

In this case, the matrix run was an FMP10 matrix, meaning the masses of our metabolites have been previously annotated. We will load this file in an add the results to our SpaMTP MSI object.


```{r}
fmp10_metabolites <- read.csv("vignette_data_files/Mouse_Brain/SM_data/Public_data_FMP10_Annotations.csv")
```

```{r, echo=TRUE, results='hide'}
fmp10_metabolites
```

```{r, echo=FALSE}
df <- fmp10_metabolites
knitr::kable(df, format = "html", table.attr = "class='dataframe'",  align = "c")
```


```{r}
## This function subsets the data and adds the metabolite names to the provided seurat object
AddCustomMZAnnotations <- function(data, annotations, assay = "Spatial", return.only.annotated = FALSE, mass.threshold = 0.05){
    
    true_mzs <- c()
    for (mass in annotations$Mass){
        true_mz <- FindNearestMZ(data, mass)
        true_mzs <- c(true_mzs,true_mz)
    }

    annotations$true_mzs_name <- true_mzs
    annotations$true_mzs <- gsub("mz-", "", annotations$true_mzs_name)
    annotations$ppm_diff <- abs(annotations$Mass - as.numeric(annotations$true_mzs))

    if (!is.null(mass.threshold)){
        annotations <- annotations %>% filter(ppm_diff < mass.threshold)
    }
    
    if (return.only.annotated) {
        
        data_sub <- subsetMZFeatures(data,features = annotations$true_mzs_name, assay = assay)
        data_sub[[assay]]@meta.data$all_IsomerNames <- annotations$compound
    
    } else {
        metabolites <- lapply(rownames(data), function(x) {

            if (x %in% annotations$true_mzs_name) {
              annotations[annotations$true_mzs_name == x,][["compound"]]
            } else {
                x
            }      
        })
        
        data_sub <-  data
        data_sub[[assay]]@meta.data$all_IsomerNames <- unlist(metabolites)
    }


    return(data_sub)
}
```


```{r}
msi_annotated <- AddCustomMZAnnotations(msi, fmp10_metabolites)
```

The function above adds the annotations into the SpaMTP object's feature metadata slot. We can see this below:


```{r, echo=TRUE, results='hide'}
msi_annotated[["Spatial"]]@meta.data[410:415,]
```

```{r, echo=FALSE}
df <- msi_annotated[["Spatial"]]@meta.data[410:415,]
knitr::kable(df, format = "html", table.attr = "class='dataframe'",  align = "c")
```



We can see based on the FMP10-tagged value we are able to annotated some of our m/z values with their respective metabolites

Dopamine plays a critical part in Parkinson's disease, so lets visualise this annotations spatially!


```{r, fig.width=7, fig.height=6}
ImageMZAnnotationPlot(msi_annotated, "Dopamine (single derivatized)", slot = "counts", dark.background = FALSE, size = 1.5)
```


There is a strong localisation of dopamine in the non-lesioned hemisphere. 

<br>

## Align Multi-Omic Technologies 

Although we have two omic datasets from the same sample, in their current state the coordinates are not alined. We can see this in the first plot below. 


```{r, fig.width=15, fig.height=8}
df1 <- GetTissueCoordinates(vis)
df1$sample <- "RNA"
    
df2 <- GetTissueCoordinates(msi_annotated)
df2[,"sample"] <- "MSI"

p1 <- ggplot(df1, aes(x, y,color = sample)) + theme_classic() + geom_point(size = 0.5)  + scale_color_manual(values = "blue")
p2 <- ggplot(df2, aes(x, y,color = sample)) + theme_classic() + geom_point(size = 0.5) +  scale_color_manual(values = "red")

p1|p2
```


<br>


To achieve multi-omic alignment we must perform 2 steps:

**1. Align Coordinates:**

   * Here we will manually adjust the cooridnates of the SM data to match the H&E image of the visium data using a *Shiny app* generated through *ManualAlignImages()*
   * We can change whats data is plot/visualised to assist with correct alignment
   * This will output a SM *SpaMTP Seurat Class Object* with transformed cooridnates (transformed via an affine transformation)

**2. Map Omic Data to Common Spots:**

   * The final step is to map or bin the SM data to each ST spot using *MapSpatialOmics()*
   * In step works by finding all SM data points within a radius of a ST spot and asigning their respective data
   * Increasing the SM resolution in this function can imporve the mapping by reducing the distance between pixel and spot centroids.

<br>
Lets run these functions on our sample:


```{r, eval=FALSE}
msi_transformed <- AlignSpatialOmics(sm.data = msi_annotated, st.data = vis)
```

    Loading required package: shiny
    
    Listening on http://0.0.0.0:4698

    Sample: 2 
     1.232057 0.006562681 67.76325 
     0.02321728 -1.188214 569.6908 
     0 0 1 

```{r, include=FALSE}
msi_transformed <- readRDS("vignette_data_files/Mouse_Brain/SM_data/SpaMTP_msi_transformed.RDS")
```    


```{r, echo=TRUE, results='hide'}
head(vis)
```

```{r, echo=FALSE}
df <- head(vis)
knitr::kable(df, format = "html", table.attr = "class='dataframe'",  align = "c")
```



```{r, fig.width=10, fig.height=8}
CheckAlignment(msi_transformed, vis, image.res = "hires", size = 0.8)
```


we can also check that our SM cooridnates havent been distorted by plotting a m/z value. Let's compare before vs after


```{r, fig.width=15, fig.height=8}
ImageMZPlot(msi_annotated, mzs = 674.28, size = 2, slot = "counts") | 
  ImageMZPlot(msi_transformed, mzs = 674.28, size = 2, slot = "counts")
```


Great, now that we are confident that our SM and ST data are aligned to the same coordinates we can map each pixel and their relative m/z intensity values to their corresponding ST spot. 

NOTE: we have increased the resolution of the SM data by 4 so that each pixel is now split into 4 smaller pixels (this will help with matching pixel centroid to spot centroids).


```{r, warning=FALSE, message=FALSE, results='hide'}
combined.data <- MapSpatialOmics(SM.data = msi_transformed, ST.data = vis, res_increase = 4, annotations = TRUE)
```

    Increasing the resolution of MALDI Pixel Data ...
    Median Distance Between MALDI Spots: 11.8345060347659
    Generating psuedo-highres MALDI data: 
    
      |==================================================| 100%

    Assigning MALDI to Visium Spots ... 
    Merging MALDI counts ... 

      |==================================================| 100%
    
    Generating new MALDI Seurat Object ... 
 

Lets visualise how the mapping went and compare it to our orginal SM data object


```{r, fig.width=15,fig.height=8}
ImageMZPlot(msi_transformed, mzs = 674.28, size = 2, slot = "counts") | 
  SpatialMZPlot(combined.data, mzs = 674.28, assay = "SPM", slot = "counts", pt.size.factor = 2.2)
```


We can see that the expression of Dopamine (mz-674.28) displays the same expression pattern spatially in our visium spot mapped object compared to the original MSI-only data object. 

Our data looks almost ready to perform some downstream analyses on. However, based on the provided annotations displayed above, we should remove all spots that are not annotated as an atomical region (meaning they are located outside of the tissue).



```{r, warning=FALSE, message=FALSE}
combined.data <- subset(combined.data, subset = RegionLoupe == "", invert = T)
```


```{r, fig.width=20, fig.height=6}
SpatialDimPlot(combined.data, group.by = c("lesion", "RegionLoupe"))
```



We can see now that we have removed spots that were associated with the outer border of the tissue. For example the bottom portion of this section has a fold in the H&E and thus is technical noise, meaning these spots need to be removed. 

<br>

## Clustering/Downstream Analysis

Now that our spatial datasets are cleaned and fully aligned we can perform some downstream analyses to determine cell types, differentially expressed genes and metabolites, and also upregulated metabolic pathways. 

#### Cluster ST data

First we will run clustering on the ST data. Based on the two different modalities, mRNA data will perform much better at defining different cell types/structures within the tissue sample. We expect the clustering to be able to detect such cell types as the striatum.


```{r, warning=FALSE, message=FALSE}
DefaultAssay(combined.data) <- "SPT"

## Run ST clustering 
combined.data <- NormalizeData(combined.data, verbose = FALSE)
combined.data <- FindVariableFeatures(combined.data, verbose = FALSE)
combined.data <- ScaleData(combined.data, verbose = FALSE)
combined.data <- RunPCA(combined.data, npcs = 30, reduction.name = "spt.pca", verbose = FALSE)
combined.data <- FindNeighbors(combined.data, dims = 1:30, reduction = "spt.pca", verbose = FALSE)
combined.data <- RunUMAP(combined.data, dims = 1:30, reduction = "spt.pca", reduction.name = "spt.umap", verbose = FALSE)
combined.data <- FindClusters(combined.data, resolution = 0.5, cluster.name = "ST_clusters", verbose = FALSE)
```


```{r, fig.width=20, fig.height=8}
DimPlot(combined.data, group.by = "ST_clusters", reduction = "spt.umap") | 
  SpatialDimPlot(combined.data, group.by = "ST_clusters", pt.size.factor = 2.1)
```


The clustering results demonstrate clear groups of cells. Lets compare them to the provided annotations:


```{r, warning=FALSE, message=FALSE, fig.width=15, fig.height=5}
striatum <- subset(combined.data, subset = region == "striatum")

SpatialDimPlot(striatum, group.by = "ST_clusters", label = F, pt.size.factor = 2.4) | 
  SpatialDimPlot(striatum, group.by = "RegionLoupe", label = F, pt.size.factor = 2.4) 
```


Looking at these results we can see that our clustering using only genetic information can determine clear morphological brain regions within the Striatum, including the caudoputamen (CP) and nucleus accumbens (ACB). However, clustering using genetic infomation cannot identify lesioned regions of the Striatum, where the striatum cells of left hemisphere produce significantly higher levels of Dopamine.



```{r, warning=FALSE, message=FALSE}
Idents(combined.data) <- "ST_clusters"

DE <- FindAllMarkers(combined.data, only.pos = T)
```

 
```{r, echo=TRUE, results='hide'}
 ## prints top 5 genes for cluster 0 == Striatum
head(DE[DE$cluster == 0,] , n =5 )
```

```{r, echo=FALSE}
df <- head(DE[DE$cluster == 0,] , n =5 )
knitr::kable(df, format = "html", table.attr = "class='dataframe'",  align = "c")
```



Looking at the top 5 most over-expressed genes in cluster 0, we can see some key genes that define the striatum. These are plot below:


```{r, fig.width=15, fig.height=5}
SpatialFeaturePlot(combined.data, features = c("Ppp1r1b", "Gpr88", "Rgs9"), pt.size.factor = 2) 
```



Identified in the original publication, the gene 'Pcp4' was highlighted as differentially expressed between the in-tact vs lesioned Striatum. Because the gene expression alone cannot differentiate between the lesion vs non-lession cells this gene is not observed in the top most differentially expressed genes. We will need to look at the metabolic data to identify this gene

<br>

#### Cluster SM data

Clustering based on metabolic infomation will likely reveal different infomation compared to a transcriptomic approach. SM clustering will reveal more about the activity or functioning of different cell types. Lets see how the clustering differs:


```{r, warning=FALSE, message=FALSE}
DefaultAssay(combined.data) <- "SPM"

## Run ST clustering 
combined.data <- NormalizeSMData(combined.data)
combined.data <- FindVariableFeatures(combined.data, verbose = FALSE)
combined.data <- ScaleData(combined.data, verbose = FALSE)
combined.data <- RunPCA(combined.data, npcs = 30, reduction.name = "spm.pca", verbose = FALSE)
combined.data <- FindNeighbors(combined.data, dims = 1:30, reduction = "spm.pca", verbose = FALSE)
combined.data <- RunUMAP(combined.data, dims = 1:30, reduction = "spm.pca", reduction.name = "spm.umap", verbose = FALSE)
combined.data <- FindClusters(combined.data, resolution = 0.4,  cluster.name = "SM_clusters", verbose = FALSE)
```


```{r, fig.width=25, fig.height=8}
DimPlot(combined.data, group.by = "SM_clusters", reduction = "spm.umap") |
  SpatialDimPlot(combined.data, group.by = "SM_clusters", pt.size.factor = 2.1) | 
  SpatialDimPlot(combined.data, group.by = "lesion", pt.size.factor = 2.1)
```



Looking at the clustering results above we can clearly see <span style="color: purple">**(cluster 4)**</span> defines the cells of the intact Striatum which still produce dopamine. In addition, the spatial distribution of cluster 4 matches the expression pattern of dopamine displayed below:


```{r, fig.width=10, fig.height=8}
dopamine_mzs <- SearchAnnotations(combined.data, "Dopamine", assay = "SPM")
combined.data <- BinMetabolites(combined.data, mzs = dopamine_mzs$mz_names, slot = "counts", assay = "SPM",  bin_name = "Dopamine")

SpatialFeaturePlot(combined.data, features = "Dopamine", pt.size.factor = 2.1) + theme(legend.position = "right")
```

<br>

## Multi-Modal Integrative Analysis

As explained above transcriptomic and metabolic data reveal and define different aspects of the tissue. Based on this multi-modal integration can be used to combine these features into the one analysis. Below we will integrate the spatail transcriptomic and metabolomic data together to get a more refined and detailed clustering result.

The SpaMTP *MultiOmicIntegration* function allows for automated or manual specification of weights for each modality when generating the integrated clusters. 


```{r}
integrated.data <- MultiOmicIntegration(combined.data,return.intermediate = T, weight.list = list(0.4,0.6))
```


```{r}
integrated.data <- RunUMAP(integrated.data, nn.name = "weighted.nn", reduction.name = "wnn.umap", reduction.key = "wnnUMAP_",verbose = FALSE)
integrated.data <- FindClusters(integrated.data, graph.name = "wsnn", algorithm = 3, resolution = 1,  cluster.name = "integrated_clusters", verbose = FALSE)
```


```{r, fig.width=20, fig.height=8}
DimPlot(integrated.data, group.by = "integrated_clusters",  reduction = 'wnn.umap', label = F) | SpatialDimPlot(integrated.data, group.by = "integrated_clusters",  label = F, pt.size.factor = 2.1) 
```



Looking at the results above we can see the integrative analysis identifies a difference between the lesioned and in-tact striatum cells (<span style="color: #E68613">cluster 1</span> and <span style="color: #7CAE00">cluster 3</span> respectively)


#### Differential Metabolite and Gene Expression Analysis

Based on this we can look at which metabolites and genes are differentially expressed between these clusters. First we will identify which clusters display an up-regulate in dopamine:


```{r, warning=FALSE, message=FALSE}
# Performs pooling, pseudo-bulking and EdgeR Differential Expression Analysis
cluster_DEPs <- FindAllDEPs(data = integrated.data, ident = "integrated_clusters", n = 3, logFC_threshold = 1.2, 
                            DE_output_dir =NULL, return.individual = FALSE, assay = "SPM",
                            run_name = "FindAllDEPs", annotation.column = "all_IsomerNames" ) # DE results are returned in a data.frame
```

    Pooling one sample into 3 replicates...
    Running edgeR DE Analysis for  FindAllDEPs  -> with samples [ 3, 1, 2, 4, 7, 5, 6, 9, 0, 8, 10 ]
    Starting condition: 3
    Starting condition: 1
    Starting condition: 2
    Starting condition: 4
    Starting condition: 7
    Starting condition: 5
    Starting condition: 6
    Starting condition: 9
    Starting condition: 0
    Starting condition: 8
    Starting condition: 10
    
```{r, fig.width=6, fig.height=8}
results <- cluster_DEPs$DEPs[cluster_DEPs$DEPs$cluster == "1",]

metabolites <- c("Dopamine (double derivatized)", "Dopamine (single derivatized)")

### Sets up colouring for significant spots in volcano plot
keyvals <- ifelse(
    results$annotations %in% metabolites, "orange", ifelse(
    results$logFC < -1.2  & results$PValue < 10e-4, 'royalblue',
      ifelse(results$logFC > 1.2 & results$PValue < 10e-4, 'red',
        'black')))
  keyvals[is.na(keyvals)] <- 'black'
  names(keyvals)[keyvals == 'red'] <- 'Cluster 1 - UP'
  names(keyvals)[keyvals == 'black'] <- 'Non-Sig'
  names(keyvals)[keyvals == 'orange'] <- 'Dopamine'


### Plots volcano plot with DEP results
volc_plot <- EnhancedVolcano::EnhancedVolcano( results,
                                  selectLab = c(""), 
                                  lab = results$annotations,
                                  colCustom = keyvals,
                                  pCutoff = 10e-4,
                                  FCcutoff = 1.2,
                                  pointSize = 3,
                                  colAlpha = 1,
                                  x = 'logFC',
                                  y = 'PValue', 
                                  gridlines.major = FALSE,
                                  gridlines.minor = FALSE)


volc_plot
```



Analysing the DE results we can see that Dopamine (both the single and double FMP10-derivatized) is over expressed in <span style="color: #E68613">cluster 1</span>. 

Now we will look at the gene expression differences between cluster 1 and 3:


```{r, warning=FALSE, message=FALSE}
Idents(integrated.data) <- "integrated_clusters"
DefaultAssay(integrated.data) <- "SPT"

striatum <- subset(integrated.data, subset = integrated_clusters %in% c("1", "3"))

striatum_DEGs <- FindAllMarkers(striatum, only.pos = T)
```

```{r, echo=TRUE, results='hide'}
head(striatum_DEGs[striatum_DEGs$cluster == "1",], n = 5)
```

```{r, echo=FALSE}
df <- head(striatum_DEGs[striatum_DEGs$cluster == "1",], n = 5)
knitr::kable(df, format = "html", table.attr = "class='dataframe'",  align = "c")
```


Supporting the findings from the orginal publication we can see that "Pcp4" is up-regulated in cluster 1 (in-tact striatum cells).


```{r, fig.width=6, fig.height=8}
SpatialFeaturePlot(integrated.data, features = "Pcp4", pt.size.factor = 2)
```

<br>

## Multi-Omic Data Visualisation

Based on this analysis we can visualise this gene and metabolite of interest in the same 3D plot ("Pcp4" and "Dopamine")


```{r, fig.width=10, fig.height=30}
Plot3DFeature(combined.data, 
              features = c("Pcp4", "Dopamine"), 
              assays = c("SPT"), 
              between.layer.height = 10000, 
              #show.image = "slice1", 
              plot.height = 400, 
              plot.width = 800)
```

<br>

## Multi-Modality Pathway Analysis

TODO: Ty

```{r}
sessionInfo()
```
