---
title: "Single_Cell_MultiOmics"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Single_Cell_MultiOmics}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```


This vignette will use [***SpaMTP***](https://github.com/GenomicsMachineLearning/SpaMTP) to integrate single cell resolution spatial transcriptomics data with spatial metabolic data. This vignette will use simulated Xenium (ST) and simulated MALDI-TOF (SM) data.

Author: Andrew Causer

```{r load_packages, warning=FALSE, message=FALSE}
#General Libraries
library(SpaMTP)
library(Cardinal)
library(Seurat)
library(dplyr)
library(ggplot2)
```


##### Load Processed Data

Here, we will be using the simulated data which has synthetic clustering and cell type names. The gene and metabolite names are also arbitrary, this data is only used to highlight ***SpaMTP's*** functionally with single cell spatial datasets. 

```{r warning=FALSE, message=FALSE}
xenium <- readRDS("vignette_data_files/Simulated_data/Sim_Xenium.RDS")
MALDI <- readRDS("vignette_data_files/Simulated_data/Sim_MALDI.RDS")
```  


## Visualising Simulated Datasets

We can visualise the simulated datasets and observe the difference in resolution between our single cell ST and lower resolution SM data.

First, lets look at our Xenium data:

```{r, fig.width=11, fig.height=5, warning=FALSE, message=FALSE}
ImageDimPlot(xenium, group.by = "celltype", size = 1)
```

Now, lets look at our MALDI data:

```{r, fig.width=11, fig.height=5, warning=FALSE, message=FALSE}
ImageDimPlot(MALDI, group.by = "clusters", size = 1)
```



Based on these plots, we can see that our Xenium data is at a much higher resolution, compared to the MALDI data. We can zoom in and see this in more detail:

<details>
  <summary>*Setting Futures for FOV Creation*</summary>
```{r}
# Futures may be required to use Seurat's `Crop()` function

library(future)
plan("multicore", workers = 16) # Use all 16 cores

# Each core can have up to 16 GB of memory (16 GB / 16 cores)
options(future.globals.maxSize = 16000 * 1024^2) # 16 GB per core

```

</details> 

```{r, fig.width=11, fig.height=5, warning=FALSE, message=FALSE}
# Generate a smaller FOV to look at a zoomed in region of the ST data
cropped.coords.xenium <- Crop(xenium[["fov"]], y = c(2000, 3000), x = c(8500, 9500), coords = "tissue")
xenium[["zoom"]] <- cropped.coords.xenium

# Generate a smaller FOV to look at a zoomed in region of the SM data
cropped.coords.MALDI <- Crop(MALDI[["fov"]], y = c(2000, 3000), x = c(8500, 9500), coords = "tissue")
MALDI[["zoom"]] <- cropped.coords.MALDI
```

Lets visualise this zoomed area now:
```{r, fig.width=11, fig.height=5, warning=FALSE, message=FALSE}
# Set the boundary as segmentation for our Xenium data
DefaultBoundary(xenium[["zoom"]]) <- "segmentation"

ImageDimPlot(xenium, group.by = "celltype", fov = "zoom", size = 1)| ImageDimPlot(MALDI, group.by = "clusters", fov = "zoom", size = 2)
```
Looking at our zoomed in FOV the single cell resolution ST provides much higher levels of detail compared to the SM data. 


## Mapping Single Cell Resolution ST and SM

We want to eventually map our SM data to our single cell resolution ST data, meaning that for each single cell we will have both transcriptomic and metabolic information. First, we can check the alignment of our datasets. 

```{r, fig.width=11, fig.height=5, warning=FALSE, message=FALSE}
# Set the boundary as segmentation for our Xenium data
CheckAlignment(ST.data = xenium, SM.data = MALDI, image.slice = "fov") & coord_flip()
```

Based on this, we can see our datasets are aligned correctly. Next, we will map our SM and Xenium ST data to generate a *SpaMTP* *Seurat* object containing metabolite and transcriptomic information per individual cell. 

```{r, fig.width=11, fig.height=5, warning=FALSE, message=FALSE}
# Set the boundary as segmentation for our Xenium data
MO_data <- MapSpatialOmics(SM.data = MALDI, ST.data = xenium, ST.hires = TRUE, SM.assay = "Spatial",ST.assay = "Xenium", SM.fov = "fov", ST.image = "fov")
```

Our two datasets have been aligned to each individual cell. Lets visualise the results:

```{r}
MO_data
```
The multi-omic *SpaMTP* *Seurat* object now contains two assays, 'SPT' and 'SPM' which stores the transcriptomic and metabolomic data respectively. 

```{r, echo=TRUE, results='hide'}
head(MO_data, n = 3)
```

This object also has the combined metadata from both original objects. The metadata associated with our original SM object is marked by "_SPM".

```{r, echo=FALSE}
df <- head(MO_data, n = 3)
df %>%
    knitr::kable(format = "html", table.attr = "class='dataframe'", align = "c") %>%
    kableExtra::kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"),
                  full_width = F, position = "center") %>%
    kableExtra::row_spec(0, bold = TRUE)%>%
    kableExtra::row_spec(1:nrow(df), extra_css = "border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;")
```


Next, we can compare the data spatially:

```{r, fig.width=11, fig.height=5, warning=FALSE, message=FALSE}
# Set the boundary as segmentation for our Xenium data
DefaultBoundary(MO_data[["zoom"]]) <- "segmentation"

p1 <- ImageDimPlot(MALDI, group.by = "clusters", fov = "zoom", size = 1.5, dark.background = F)/ ImageDimPlot(MO_data, group.by = "clusters_SPM", fov = "zoom", dark.background = F)
p2 <- ImageFeaturePlot(MALDI, features = "mz-100", fov = "zoom", size = 1.5, dark.background = F)/ ImageFeaturePlot(MO_data, features = "mz-100", fov = "zoom", dark.background = F)

p1|p2
```

Comparing the mapped data to the original SM data we can see that the spatial pattern of both clusters and metabolite ('mz-100') match. Now that our data is mapped to single cell resolution we have a much more detailed visualisation of SM based clustering across our simulated tissue sample. 


SM data indicates the metabolic functioning of a cell, and based on this we can identify cell types with different metabolic states:

```{r, fig.width=11, fig.height=5, warning=FALSE, message=FALSE}
table(MO_data$clusters_SPM, MO_data$celltype)
```
Looking at Oligodendrocytes for example, we can identify different subtypes based on their metabolic activity. 


```{r, fig.width=11, fig.height=5, warning=FALSE, message=FALSE}
MO_data$Oligo_states <- ifelse(MO_data$celltype == "Oligodendrocyte", paste0("Oligo State: ", MO_data$clusters_SPM),  "Other")

ImageDimPlot(MO_data, group.by = "Oligo_states", fov = "zoom", dark.background = F, cols = c("grey", "red", "blue", "yellow", "green", "pink"))
```

```{r, fig.width=20, fig.height=5, warning=FALSE, message=FALSE}
DotPlot(MO_data, group.by = "Oligo_states", features = c(rownames(MO_data[["SPT"]]), "spm_mz-200"))
```
This dotplot may suggest that Oligo State 1 and 2 are different cell subtypes based on their differential expression of gene1/gene2 and mz-200 (Note: this is synthetic data and has no biological context - results only used as example).



An additional features of having multi-omic data mapped to the same coordinates is that we can visualise the expression of both metabolites and gene expression on a single plot. 

```{r, fig.width=11, fig.height=5, warning=FALSE, message=FALSE}

ImageFeaturePlot(MO_data, features = "mz-200", molecules = c("gene1", "gene2"), fov = "zoom", mols.cols = c("blue", "green"))
```
Analysing this spatial plot, there is a clear correlation between 'mz-100' and 'gene1' suggesting there interaction or similar functionality. 


## Session Info

```{r}
sessionInfo()
```


