---
title: "Benchmarking"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Benchmarking}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(SpaMTP)
library(UpSetR)
library(Seurat)
```


## Clustering Comparison to find dopamine neurons

```{r}
## use integrated.data from `Multi-Omic_Mouse_Brain.Rmd` tutorial

### split data based on spots identified as containing dopamine neurons 
annot <- subset(integrated.data, subset = annotations == "CP_intact")
sm <- subset(integrated.data, subset = SM_clusters=="3")
st <- subset(integrated.data, subset = ST_clusters == "0")
int <- subset(integrated.data, subset = integrated_clusters == "4")

list_names <- c("annot", "sm", "st", "int")
comp_data <- list("annot" = annot, "sm" = sm, "st" = st, "int" = int)


### Set up data.frame for storing results
df <- data.frame(matrix(ncol = 4, nrow = length(list_names)))
# Set row names
rownames(df) <- list_names
# Set column names
colnames(df) <- c("Mean", 
                  "Median", 
                  " % Dopamine Spots", 
                  "Mean Expression in Dopamine Positive Spots")


### Calculate statistics on dopamine expression across each clustering group
dopamine_mzs <- SearchAnnotations(integrated.data, "Dopamine", assay = "SPM")


for (x in names(comp_data)) {
  comp_data[[x]] <- BinMetabolites(comp_data[[x]], mzs = dopamine_mzs$mz_names, slot = "counts", assay = "SPM",  bin_name = "Dopamine")
  
  df[x, 1] <- mean(comp_data[[x]]$Dopamine)
  df[x, 2] <- median(comp_data[[x]]$Dopamine)
  
  data <- comp_data[[x]][["SPM"]]["counts"]["mz-674.2805",]
  comp_data[[x]]$contains_dop <- ifelse(data > 0, "pos", "neg")
  
  # Subset where contains_dop is "pos"
  sub <- subset(comp_data[[x]], subset = contains_dop == "pos")
  
  df[x, 3] <- sum(comp_data[[x]]$contains_dop == "pos") * 100 / (sum(comp_data[[x]]$contains_dop == "neg") + sum(comp_data[[x]]$contains_dop == "pos"))
  df[x, 4] <- mean(sub[["SPM"]]["counts"]["mz-674.2805",], na.rm = TRUE)
}



### Get Gene and metabolite correlation per cluster
integrated.data$annot <- ifelse(integrated.data$annotations == "CP_intact", "T","F")
integrated.data$sm <- ifelse(integrated.data$SM_clusters == "3", "T","F")
integrated.data$st <- ifelse(integrated.data$ST_clusters == "0", "T","F")
integrated.data$int <- ifelse(integrated.data$integrated_clusters == "4", "T","F")

striatum <- subset(integrated.data, subset = region == "striatum"& lesion == "intact")

correlation_df <- lapply(list_names, function(x){
  suppressWarnings({
  cor <- FindCorrelatedFeatures(striatum, ident = x, ST.assay = "SPT", SM.assay = "SPM", ST.slot = "counts", SM.slot = "counts", nfeatures = NULL)
  })
  cor[cor$ident == "T",]
})

names(correlation_df) <- list_names

lapply(names(correlation_df), function(x){
  correlation_df[[x]][correlation_df[[x]]$features == "mz-674.28833",]
})

df$Dopamine.Correlation <- do.call(rbind,lapply(names(correlation_df), function(x){
     correlation_df[[x]][correlation_df[[x]]$features == "mz-674.28833",]}))$correlation

df$Group <- rownames(df)
original_data <- df

# Scale the data (excluding Group)
scaled_data <- df %>%
    select(-Group) %>%
    scale() %>%
    as.data.frame()

# Add 'Group' back
scaled_data$Group <- df$Group

# Convert scaled data to long format
long_scaled <- scaled_data %>%
    pivot_longer(cols = -Group, names_to = "Metric", values_to = "Scaled_Value")

# Convert original data to long format
long_original <- original_data %>%
    pivot_longer(cols = -Group, names_to = "Metric", values_to = "Original_Value")

# Merge scaled and original data
long_data <- left_join(long_scaled, long_original, by = c("Group", "Metric"))

long_data$Metric <- factor(long_data$Metric, levels = rev(c("Mean", 
                                                        "Median",
                                                        "Dopamine.Correlation", 
                                                        " % Dopamine Spots", 
                                                        "Mean Expression in Dopamine Positive Spots"
                                                        )))

# Plot heatmap with original values displayed
ggplot(long_data, aes(x = Group, y = Metric, fill = Scaled_Value)) +
    geom_tile() +
    geom_text(aes(label = round(Original_Value, 1)), color = "black", size = 3) +  # Add original values as text
    scale_fill_gradientn(colors = c("blue", "white", "red")) +
    theme_minimal() +
    labs(title = "Scaled Heatmap with Original Values", x = "Group", y = "Metric") +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))


metric_levels <- unique(long_data$Metric)
background_df <- data.frame(
    Metric = metric_levels,
    Index = seq_along(metric_levels)  # Create an index to alternate colors
)


long_data$Metric <- factor(long_data$Metric, levels = c("Mean", 
                                                        "Median",
                                                        "Dopamine.Correlation", 
                                                        " % Dopamine Spots", 
                                                        "Mean Expression in Dopamine Positive Spots"
                                                        ))

ggplot(long_data, aes(x = Metric, y = Scaled_Value, group = Group, color = Group)) +
    # Add alternating grey bars
    geom_rect(data = background_df, 
              aes(xmin = Index - 0.5, xmax = Index + 0.5, ymin = -Inf, ymax = Inf), 
              fill = rep(c("grey90", "white"), length.out = nrow(background_df)), 
              inherit.aes = FALSE, alpha = 0.5) +
    
    geom_line(size = 1) +  # Line plot
    geom_point(size = 3) + # Add points
    theme_minimal() +
    labs(title = "Line Plot of Metrics by Group",
         x = "Metric", 
         y = "Scaled Value") +
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    theme_classic()

```




## Pathway Analysis Benchmarking

Load data, run DE and ***SpaMTP*** pathway analysis
```{r}
## Load Data
striatum.dhb.data <- readRDS("vignette_data_files/Mouse_Brain/DHB_data/striatum.dhb.data.RDS")

## Generate differentially expressed genes/metabolites
Idents(striatum.dhb.data) <- "striatum"
deg =  FindAllMarkers(striatum.dhb.data , test.use = "wilcox_limma", assay = "SPT",only.pos =F)
dem =  FindAllMarkers(striatum.dhb.data , assay = "SPM", only.pos =T)

DE.list = list("genes" = deg,"metabolites" = dem)

## Run multi-modal pathway analysis for SpaMTP
regpathway = FindRegionalPathways(
  striatum.dhb.data,
  analyte_types = c("genes", "metabolites"),
  SM_slot = "counts", ST_slot = "counts",
  ident = "striatum",
  DE.list = DE.list
)

## Get list of significant genes and metabolite HMDB ID's for MetaboAnalyst Benchmarking (Joint Pathway Analysis)
## https://www.metaboanalyst.ca/MetaboAnalyst/upload/PathUploadView.xhtml 
dem$ids <- striatum.dhb.data@assays$SPM@meta.data[striatum.dhb.data@assays$SPM@meta.data$mz_names %in% dem$gene,]$all_Isomers
dems <- dem[dem$cluster == "intact",]
dems <- dems[dems$p_val_adj < 0.05,]
dems <- dems[dems$ids != "No Annotation",]

dems <- dems %>%
    tidyr::separate_rows(ids, sep = "; ")

degs = deg[deg$cluster == "intact",][deg[deg$cluster == "intact",]$p_val_adj < 0.05,]
#write.csv(dems, "../Benchmarking_DE_Metabolite_IDs.csv")
#write.csv(degs, "../Benchmarking_DE_genes.csv")



upset_results <- c(
All.Pathways =   dim(regpathway[regpathway$Cluster_id == "intact",])[1] + 21 + 191,
Sig.Pathways = length(regpathway[regpathway$padj < 0.05 & regpathway$Cluster_id == "intact",]$pathwayName) + 0 + 44,
SpaMTP = dim(regpathway[regpathway$Cluster_id == "intact",])[1], 
MetaboAnalyst = 21,
IPA = 191, 
Detected.Dopamine.Pathways = 4,
"All.Pathways&SpaMTP" = dim(regpathway[regpathway$Cluster_id == "intact",])[1],
"Sig.Pathways&SpaMTP" = length(regpathway[regpathway$padj < 0.05 & regpathway$Cluster_id == "intact",]$pathwayName),
"All.Pathways&MetaboAnalyst" = 21,
"Sig.Pathways&MetaboAnalyst" = 0,
"All.Pathways&IPA" = 191,
"Sig.Pathways&IPA" = 44,
"Detected.Dopamine.Pathways&Sig.Pathways&SpaMTP" = 2,
"All.Pathways&Sig.Pathways" = length(regpathway[regpathway$padj < 0.05 & regpathway$Cluster_id == "intact",]$pathwayName) + 0 + 44
)


upset(fromExpression(upset_results),
nintersects = 40,
nsets = 6,
order.by = "freq",
decreasing = T,
mb.ratio = c(0.6, 0.4),
number.angles = 0,
text.scale = 1.1,
point.size = 2.8,
line.size = 1
)


#"Transport of bile salts and organic acids, metal ions and amine compounds"
bile <- regpathway[regpathway$pathwayName == "Transport of bile salts and organic acids, metal ions and amine compounds",]$adduct_info[1]

numbers <- str_extract_all(bile, "\\d+\\.\\d+")[[1]]

# Convert to numeric if needed
numbers <- as.numeric(numbers)
mzs <- list(unlist(lapply(numbers, function(x){FindNearestMZ(striatum.dhb.data, x)})))

striatum.dhb.data <- AddModuleScore(striatum.dhb.data, mzs, assay = "SPM", slot = "data", name = "bile_pathway", ctrl = 20)
SpatialFeaturePlot(striatum.dhb.data, features = "bile_pathway1", pt.size.factor = 2.5, image.alpha = 0.7) & scale_fill_gradient2(low = "blue", high = "red") 


PathwayNetworkPlots(striatum.dhb.data,ident = "striatum", regpathway =  regpathway, DE.list = list("genes" = deg, "metabolites" = dem), selected_pathways = c("Transport of bile salts and organic acids, metal ions and amine compounds"), path = "~../")


## benchmarking results provided in supplementary table in publication
#hm_benchmarking <- read.csv("../heatmap_benchmarking.csv")

df_long <- hm_benchmarking %>%
    pivot_longer(cols = -Pathway, names_to = "Tool", values_to = "Value")
df_long$Pathway <- factor(df_long$Pathway, hm_benchmarking$Pathway)

# Plot the heatmap
ggplot(df_long, aes(x = Tool, y = Pathway, fill = factor(Value))) +
    geom_tile(color = "white") +
    scale_fill_manual(values = c("-1" = "white", "0" = "grey", "1" = "red")) +
    labs(title = "Pathway Analysis Heatmap",
         x = "Analysis Tool",
         y = "Pathway",
         fill = "Value") +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))


```


## Benchmark Alignment
```{r}
library(bench)

msi_transformed <- readRDS("vignette_data_files/Mouse_Brain/SM_data/SpaMTP_msi_transformed.RDS")

vis <- Load10X_Spatial("vignette_data_files/Mouse_Brain/ST_data/outs/")
## Read in metadata files
lesion <- read.csv("vignette_data_files/Mouse_Brain/ST_data/outs/lesion.csv", row.names = 1)
region <- read.csv("vignette_data_files/Mouse_Brain/ST_data/outs/region.csv", row.names = 1)
region_loupe <- read.csv("vignette_data_files/Mouse_Brain/ST_data/outs/RegionLoupe.csv", row.names = 1)

## Combine into one dataframe
annotations <- cbind(lesion,region,region_loupe)

## Add data to metadata slot
vis@meta.data[colnames(annotations)] <- annotations[colnames(annotations)]
vis@meta.data <- vis@meta.data %>% mutate(annotations = ifelse(RegionLoupe %in% c("CP", "ACB"), paste0(RegionLoupe, "_", lesion), RegionLoupe))



 spot_mapping <- bench::mark(
    {
      combined.data <- MapSpatialOmics(SM.data = msi_transformed, ST.data = vis, overlap.threshold = 0.3, annotations = TRUE, ST.scale.factor = NULL)
    },
    iterations = 1, memory = TRUE, time_unit = "s", check = FALSE
  )


xenium <- readRDS("vignette_data_files/Simulated_data/Sim_Xenium.RDS")
MALDI <- readRDS("vignette_data_files/Simulated_data/Sim_MALDI.RDS")


library(future)
plan("multicore", workers = 16) # Use all 16 cores

# Each core can have up to 16 GB of memory (16 GB / 16 cores)
options(future.globals.maxSize = 16000 * 1024^2) # 16 GB per core



library(Seurat)

# Original object
original <- MALDI

# Get original count matrix
counts <- GetAssayData(original, assay = "Spatial", slot = "counts")

# Function to expand features to a given number
expand_features <- function(mat, target_n_features) {
  n_current <- nrow(mat)
  
  # How many times to repeat and how many extras
  reps <- target_n_features %/% n_current
  remainder <- target_n_features %% n_current
  
  # Duplicate rows
  mat_expanded <- mat[rep(1:n_current, times = reps), ]
  
  if (remainder > 0) {
    mat_expanded <- rbind(mat_expanded, mat[1:remainder, ])
  }
  
  # Assign new gene names
  rownames(mat_expanded) <- paste0("G", seq_len(target_n_features))
  
  return(mat_expanded)
}

# Define gene sizes
gene_counts <- c(10, 100, 1000, 5000)

# Create and store new Seurat objects
seurat_list <- lapply(gene_counts, function(n_genes) {
  expanded_counts <- expand_features(counts, n_genes)
  
  # Create new Seurat object
  seu <- CreateSeuratObject(counts = expanded_counts, assay = "Spatial")
  
  # Copy spatial info (e.g., images, coordinates) if needed
  seu@images <- original@images
  
  return(seu)
})

# Optionally name the list
names(seurat_list) <- paste0("genes_", gene_counts)

cell_mapping_results <- list()
for (i in 1:length(seurat_list)){
  xen <- xenium
  sm <- seurat_list[[i]]
  cell_mapping <- bench::mark(
    {
      MO_data <- MapSpatialOmics(SM.data = sm, ST.data = xenium, ST.hires = TRUE, SM.assay = "Spatial",ST.assay = "Xenium", SM.fov = "fov", ST.image = "fov")
    },
    iterations = 1, memory = TRUE, time_unit = "s", check = FALSE
  )
  cell_mapping_results[[i]] <- list(time = cell_mapping$median, memory = cell_mapping$mem_alloc)
  gc()
}



df <- bind_rows(lapply(cell_mapping_results, function(x) {

  
  tibble(
    time = x$time,
    memory_mb = x$memory
  )
}))

# Step 2: Add overlaps column
df$metabolites <- gene_counts





library(ggplot2)

# Assume df has: df$metabolites, df$time (in seconds), df$memory (in bytes)

# Convert memory from bytes to GB for readability
df$memory_gb <- df$memory_mb / 1024^3  # or keep in bytes if you prefer

# Scale memory to align with time for plotting
# You decide how you want the right axis to span (e.g., memory_gb from 65 to 70)
# We'll define a linear transformation: y_secondary = memory_gb * scale_factor

# Define desired y-axis ranges
range_time <- range(df$time)
range_mem <- c(70, 72.5)  # desired range in GB

# Scaling factor to match memory to time scale
scale_factor <- diff(range_time) / diff(range_mem)

# Intercept shift to align the minimums
intercept <- range_time[1] - range_mem[1] * scale_factor

# Plot
ggplot(df, aes(x = metabolites)) +
  geom_line(aes(y = time, color = "Time (s)")) +
  geom_point(aes(y = time, color = "Time (s)")) +
  geom_line(aes(y = memory_gb * scale_factor + intercept, color = "Memory (GB)")) +
  geom_point(aes(y = memory_gb * scale_factor + intercept, color = "Memory (GB)")) +
  scale_y_continuous(
    name = "Time (s)",
    sec.axis = sec_axis(~ (. - intercept) / scale_factor, name = "Memory (GB)")
  ) +
  scale_color_manual(values = c("Time (s)" = "blue", "Memory (GB)" = "red")) +
  labs(x = "Metabolites", title = "Runtime and Memory vs Metabolites", color = NULL) +
  theme_classic() +
  theme(
    axis.title.y.left = element_text(color = "blue"),
    axis.title.y.right = element_text(color = "red"),
    axis.text.y.left = element_text(color = "blue"),
    axis.text.y.right = element_text(color = "red")
  )



# Step 3: Plot
p1 <- ggplot(df, aes(x = metabolites, y = time)) +
  geom_line() +
  geom_point() +
  labs(title = "Runtime vs metabolites", x = "metabolites", y = "Time (s)") +
  theme_classic()

p2 <- ggplot(df, aes(x = metabolites, y = memory_mb)) +
  geom_line() +
  geom_point() +
  labs(title = "Memory vs metabolites", x = "metabolites", y = "Memory (MB)") +
  theme_classic()

# Show plots side by side (optional)
library(patchwork)
p1 /p2




# Extract spot mapping metrics
spot_time <- spot_mapping$total_time[[1]]           # in seconds
spot_memory <- as.numeric(spot_mapping$mem_alloc)   # already in bytes

# Convert to GB
spot_memory_gb <- spot_memory / 1024^3

# Extract cell mapping metrics
cell_time <- cell_mapping_results[[1]]$time
cell_memory_str <- cell_mapping_results[[1]]$memory

# Convert memory from string to GB
cell_memory_gb <- as.numeric(str_remove(cell_memory_str, "GB"))

# Create comparison dataframe
df <- tibble(
  method = c("Spot Mapping", "Cell Mapping"),
  time_s = c(spot_time, cell_time),
  memory_gb = c(spot_memory_gb, cell_memory_gb))

# Pivot longer for faceted bar plot
df_long <- df %>%
  pivot_longer(cols = c(time_s, memory_gb), names_to = "metric", values_to = "value")

# Pretty labels
df_long$metric <- recode(df_long$metric,
                         time_s = "Time (s)",
                         memory_gb = "Memory (GB)")

# Plot
ggplot(df_long, aes(x = method, y = value, fill = method)) +
  geom_col(width = 0.6, show.legend = FALSE) +
  facet_wrap(~metric, scales = "free_y") +
  labs(title = "Spot vs Cell Mapping: Time and Memory Comparison",
       x = NULL, y = NULL) +
  theme_classic(base_size = 14)




```
