<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Heatmap of Differentially Expressed Metabolites — DEMsHeatmap • SpaMTP</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Heatmap of Differentially Expressed Metabolites — DEMsHeatmap"><meta name="description" content="Generates a heatmap of DEMs generated from edgeR analysis run using FindAllDEMs().
This function uses pheatmap to plot data."><meta property="og:description" content="Generates a heatmap of DEMs generated from edgeR analysis run using FindAllDEMs().
This function uses pheatmap to plot data."><meta property="og:image" content="https://genomicsmachinelearning.github.io/SpaMTP/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SpaMTP</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/SpaMTP.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/Mouse_Urinary_Bladder.html">Spatial Metabolomics Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Multi-Omic_Mouse_Brain.html">Spatial Multi-Omics Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/SpaMTP_Additional_Features.html">Additional SpaMTP Features</a></li>
    <li><a class="dropdown-item" href="../articles/Single_Cell_MultiOmics.html">Simulated Single Cell Multi-Omics Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Big_Data.html">Handeling Big Datasets</a></li>
    <li><a class="dropdown-item" href="../articles/Annotation_Refinement.html">Refinning Metabolite Annotations</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Updates</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/GenomicsMachineLearning/SpaMTP" aria-label="GitHub"><span class="fa fa-github"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Heatmap of Differentially Expressed Metabolites</h1>
      <small class="dont-index">Source: <a href="https://github.com/genomicsmachinelearning/SpaMTP/blob/HEAD/R/DifferentialMZPeaks.R" class="external-link"><code>R/DifferentialMZPeaks.R</code></a></small>
      <div class="d-none name"><code>DEMsHeatmap.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Generates a heatmap of DEMs generated from edgeR analysis run using <code><a href="FindAllDEMs.html">FindAllDEMs()</a></code>.
This function uses <code>pheatmap</code> to plot data.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">DEMsHeatmap</span><span class="op">(</span></span>
<span>  <span class="va">edgeR_output</span>,</span>
<span>  n <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  only.pos <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  FDR.threshold <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  logfc.threshold <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  order.by <span class="op">=</span> <span class="st">"FDR"</span>,</span>
<span>  scale <span class="op">=</span> <span class="st">"row"</span>,</span>
<span>  color <span class="op">=</span> <span class="op">(</span><span class="fu">grDevices</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html" class="external-link">colorRampPalette</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"navy"</span>, <span class="st">"white"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fl">50</span><span class="op">)</span>,</span>
<span>  cluster_cols <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  cluster_rows <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  fontsize_row <span class="op">=</span> <span class="fl">15</span>,</span>
<span>  fontsize_col <span class="op">=</span> <span class="fl">15</span>,</span>
<span>  cutree_cols <span class="op">=</span> <span class="fl">9</span>,</span>
<span>  silent <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  plot_annotations_column <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  save_to_path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  plot.save.width <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  plot.save.height <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  nlabels.to.show <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  annotation_colors <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-edger-output">edgeR_output<a class="anchor" aria-label="anchor" href="#arg-edger-output"></a></dt>
<dd><p>A list containing outputs from edgeR analysis (from FindAllDEMs()). This includes pseudo-bulked counts and DEMs.</p></dd>


<dt id="arg-n">n<a class="anchor" aria-label="anchor" href="#arg-n"></a></dt>
<dd><p>A numeric integer that defines the number of UP and DOWN regulated peaks to plot (default = 25).</p></dd>


<dt id="arg-only-pos">only.pos<a class="anchor" aria-label="anchor" href="#arg-only-pos"></a></dt>
<dd><p>Boolean indicating if only positive markers should be returned (default = FALSE).</p></dd>


<dt id="arg-fdr-threshold">FDR.threshold<a class="anchor" aria-label="anchor" href="#arg-fdr-threshold"></a></dt>
<dd><p>Numeric value that defines the FDR threshold to use for defining most significant results (default = 0.05).</p></dd>


<dt id="arg-logfc-threshold">logfc.threshold<a class="anchor" aria-label="anchor" href="#arg-logfc-threshold"></a></dt>
<dd><p>Numeric value that defines the logFC threshold to use for filtering significant results (default = 0.5).</p></dd>


<dt id="arg-order-by">order.by<a class="anchor" aria-label="anchor" href="#arg-order-by"></a></dt>
<dd><p>Character string defining which parameter to order markers by, options are either 'FDR' or 'logFC' (default = "FDR").</p></dd>


<dt id="arg-scale">scale<a class="anchor" aria-label="anchor" href="#arg-scale"></a></dt>
<dd><p>A character string indicating if the values should be centered and scaled in either the row direction or the column direction, or none. Corresponding values are "row", "column" and "none"</p></dd>


<dt id="arg-color">color<a class="anchor" aria-label="anchor" href="#arg-color"></a></dt>
<dd><p>A vector of colors used in heatmap (default = grDevices::colorRampPalette(c("navy", "white", "red"))(50)).</p></dd>


<dt id="arg-cluster-cols">cluster_cols<a class="anchor" aria-label="anchor" href="#arg-cluster-cols"></a></dt>
<dd><p>Boolean value determining if columns should be clustered or hclust object (default = F).</p></dd>


<dt id="arg-cluster-rows">cluster_rows<a class="anchor" aria-label="anchor" href="#arg-cluster-rows"></a></dt>
<dd><p>Boolean value determining if rows should be clustered or hclust object (default = T).</p></dd>


<dt id="arg-fontsize-row">fontsize_row<a class="anchor" aria-label="anchor" href="#arg-fontsize-row"></a></dt>
<dd><p>A numeric value defining the fontsize of rownames (default = 15).</p></dd>


<dt id="arg-fontsize-col">fontsize_col<a class="anchor" aria-label="anchor" href="#arg-fontsize-col"></a></dt>
<dd><p>A numeric value defining the fontsize of colnames (default = 15).</p></dd>


<dt id="arg-cutree-cols">cutree_cols<a class="anchor" aria-label="anchor" href="#arg-cutree-cols"></a></dt>
<dd><p>A numeric value defining the number of clusters the columns are divided into, based on the hierarchical clustering(using cutree), if cols are not clustered, the argument is ignored (default = 9).</p></dd>


<dt id="arg-silent">silent<a class="anchor" aria-label="anchor" href="#arg-silent"></a></dt>
<dd><p>Boolean value indicating if the plot should not be draw (default = TRUE).</p></dd>


<dt id="arg-plot-annotations-column">plot_annotations_column<a class="anchor" aria-label="anchor" href="#arg-plot-annotations-column"></a></dt>
<dd><p>Character string indicating the column name that contains the metabolite annotations to plot. Annotations = TRUE must be used in FindAllDEMs() for edgeR output to include annotations. If plot_annotations_column = NULL, m/z vaues will be plotted (default = NULL).</p></dd>


<dt id="arg-save-to-path">save_to_path<a class="anchor" aria-label="anchor" href="#arg-save-to-path"></a></dt>
<dd><p>Character string defining the full filepath and name of the plot to be saved as.</p></dd>


<dt id="arg-plot-save-width">plot.save.width<a class="anchor" aria-label="anchor" href="#arg-plot-save-width"></a></dt>
<dd><p>Integer value representing the width of the saved pdf plot (default = 20).</p></dd>


<dt id="arg-plot-save-height">plot.save.height<a class="anchor" aria-label="anchor" href="#arg-plot-save-height"></a></dt>
<dd><p>Integer value representing the height of the saved pdf plot (default = 20).</p></dd>


<dt id="arg-nlabels-to-show">nlabels.to.show<a class="anchor" aria-label="anchor" href="#arg-nlabels-to-show"></a></dt>
<dd><p>Numeric value defining the number of annotations to show per m/z (default = NULL).</p></dd>


<dt id="arg-annotation-colors">annotation_colors<a class="anchor" aria-label="anchor" href="#arg-annotation-colors"></a></dt>
<dd><p>List for specifying annotation_row and annotation_col track colors manually. Check pheatmap R-Package documentation for details. If set to 'NA', default coloring will be used (default = NA).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A heatmap plot of significantly differentially expressed metabolites defined in the edgeR ouput object.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># DEMs &lt;- FindAllDEMs(SeuratObj, "sample")</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># DEMsHeatmap(DEMs)</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/agc888" class="external-link">Andrew Causer</a>, <a href="https://github.com/BCRL-tylu" class="external-link">Tianyao Lu</a>, <a href="https://github.com/ChemCharles" class="external-link">Christopher C.J. Fitzgerald</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

