<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Regional Pathway Enrichment — FindRegionalPathways • SpaMTP</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Regional Pathway Enrichment — FindRegionalPathways"><meta name="description" content="This the function used to compute the gene/metabolites set enrichment for multi-omics spatial data"><meta property="og:description" content="This the function used to compute the gene/metabolites set enrichment for multi-omics spatial data"><meta property="og:image" content="https://genomicsmachinelearning.github.io/SpaMTP/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SpaMTP</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/SpaMTP.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/Mouse_Urinary_Bladder.html">Spatial Metabolomics Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Multi-Omic_Mouse_Brain.html">Spatial Multi-Omics Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/human_brain_tissue_analysis.html">Single Cell Multi-Omics Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/SpaMTP_Additional_Features.html">Additional SpaMTP Features</a></li>
    <li><a class="dropdown-item" href="../articles/Big_Data.html">Handling Big Datasets</a></li>
    <li><a class="dropdown-item" href="../articles/Annotation_Refinement.html">Refinning Metabolite Annotations</a></li>
    <li><a class="dropdown-item" href="../articles/Single_Cell_MultiOmics.html">Simulated scMulti-Omics Analysis</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Updates</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/GenomicsMachineLearning/SpaMTP" aria-label="GitHub"><span class="fa fa-github"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Regional Pathway Enrichment</h1>
      <small class="dont-index">Source: <a href="https://github.com/genomicsmachinelearning/SpaMTP/blob/HEAD/R/PathwayAnalysis.R" class="external-link"><code>R/PathwayAnalysis.R</code></a></small>
      <div class="d-none name"><code>FindRegionalPathways.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This the function used to compute the gene/metabolites set enrichment for multi-omics spatial data</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">FindRegionalPathways</span><span class="op">(</span></span>
<span>  <span class="va">SpaMTP</span>,</span>
<span>  <span class="va">ident</span>,</span>
<span>  <span class="va">DE.list</span>,</span>
<span>  analyte_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"genes"</span>, <span class="st">"metabolites"</span><span class="op">)</span>,</span>
<span>  SM_assay <span class="op">=</span> <span class="st">"SPM"</span>,</span>
<span>  ST_assay <span class="op">=</span> <span class="st">"SPT"</span>,</span>
<span>  SM_slot <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>  ST_slot <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>  min_path_size <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  max_path_size <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  pval_cutoff_mets <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  pval_cutoff_genes <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-spamtp">SpaMTP<a class="anchor" aria-label="anchor" href="#arg-spamtp"></a></dt>
<dd><p>A SpaMTP Seurat object contains spatial metabolomics(SM)/transcriptomics(ST) data or both, if contains SM data, it should be annotated via SpaMTP::AnnotateSM function.</p></dd>


<dt id="arg-ident">ident<a class="anchor" aria-label="anchor" href="#arg-ident"></a></dt>
<dd><p>A name character to specific the cluster vector for regions in <code>SpaMTP@meta.data</code> slot.</p></dd>


<dt id="arg-de-list">DE.list<a class="anchor" aria-label="anchor" href="#arg-de-list"></a></dt>
<dd><p>A list consisting of differential expression data.frames for each input modality. Within each data.frame column names MUST include 'cluster', 'gene', ('avg_log2FC' or 'logFC') and ('p_val_adj' or 'FDR').</p></dd>


<dt id="arg-analyte-types">analyte_types<a class="anchor" aria-label="anchor" href="#arg-analyte-types"></a></dt>
<dd><p>Vector of character strings defining which analyte types to use. Options can be c("genes"), c("metabolites") or both (default = c("genes", "metabolites")).</p></dd>


<dt id="arg-sm-assay">SM_assay<a class="anchor" aria-label="anchor" href="#arg-sm-assay"></a></dt>
<dd><p>A Character string defining describing slot name for spatial metabolomics data in SpaMTP to extract intensity values from (default = "SPM").</p></dd>


<dt id="arg-st-assay">ST_assay<a class="anchor" aria-label="anchor" href="#arg-st-assay"></a></dt>
<dd><p>A Character string defining describing slot name for spatial transcriptomics data in SpaMTP to extract RNA count values from (default = "SPT").</p></dd>


<dt id="arg-sm-slot">SM_slot<a class="anchor" aria-label="anchor" href="#arg-sm-slot"></a></dt>
<dd><p>The slot name containing the SM assay matrix data (default = "counts").</p></dd>


<dt id="arg-st-slot">ST_slot<a class="anchor" aria-label="anchor" href="#arg-st-slot"></a></dt>
<dd><p>The slot name containing the ST assay matrix data (default = "counts").</p></dd>


<dt id="arg-min-path-size">min_path_size<a class="anchor" aria-label="anchor" href="#arg-min-path-size"></a></dt>
<dd><p>The min number of metabolites in a specific pathway (default = 5).</p></dd>


<dt id="arg-max-path-size">max_path_size<a class="anchor" aria-label="anchor" href="#arg-max-path-size"></a></dt>
<dd><p>The max number of metabolites in a specific pathway (default = 500).</p></dd>


<dt id="arg-pval-cutoff-mets">pval_cutoff_mets<a class="anchor" aria-label="anchor" href="#arg-pval-cutoff-mets"></a></dt>
<dd><p>A numerical value defining the adjusted p value cutoff for significant differentially expressed metabolites. If <code>NULL</code> cutoff = <code>0.05</code> (default = 0.05).</p></dd>


<dt id="arg-pval-cutoff-genes">pval_cutoff_genes<a class="anchor" aria-label="anchor" href="#arg-pval-cutoff-genes"></a></dt>
<dd><p>A numerical value defining the adjusted p value cutoff for significant differentially expressed genes. If <code>NULL</code> cutoff = <code>0.05</code> (default = 0.05).</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Boolean indicating whether to show informative messages. If FALSE these messages will be suppressed (default = TRUE).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A SpaMTP object with set enrichment on given analyte types.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># SpaMTP = FindRegionalPathways(SpaMTP, polarity = "positive")</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/agc888" class="external-link">Andrew Causer</a>, <a href="https://github.com/BCRL-tylu" class="external-link">Tianyao Lu</a>, <a href="https://github.com/ChemCharles" class="external-link">Christopher C.J. Fitzgerald</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

