<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index • SpaMTP</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><meta property="og:image" content="https://genomicsmachinelearning.github.io/SpaMTP/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SpaMTP</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/SpaMTP.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/Mouse_Urinary_Bladder.html">Spatial Metabolomics Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Multi-Omic_Mouse_Brain.html">Spatial Multi-Omics Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/SpaMTP_Additional_Features.html">Additional SpaMTP Features</a></li>
    <li><a class="dropdown-item" href="../articles/Single_Cell_MultiOmics.html">Simulated Single Cell Multi-Omics Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Big_Data.html">Handling Big Datasets</a></li>
    <li><a class="dropdown-item" href="../articles/Annotation_Refinement.html">Refinning Metabolite Annotations</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Updates</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/GenomicsMachineLearning/SpaMTP" aria-label="GitHub"><span class="fa fa-github"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="loading-spatial-metabolic-data-into-a-spamtp-seurat-object">Loading Spatial Metabolic Data into a SpaMTP Seurat Object<a class="anchor" aria-label="anchor" href="#loading-spatial-metabolic-data-into-a-spamtp-seurat-object"></a></h2>

      <div class="section-desc"><p>Functions that allow the user to load in SM data in different formats</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="LoadSM.html">LoadSM()</a></code>

        </dt>
        <dd>Loads spatial metabolic data into a SpaMTP Seurat Object</dd>
      </dl><dl><dt>

          <code><a href="ReadSM_mtx.html">ReadSM_mtx()</a></code>

        </dt>
        <dd>Read Spatial Metabolomics matrix file (.csv format)</dd>
      </dl></div><div class="section level2">
      <h2 id="converting-between-data-objects">Converting Between Data Objects<a class="anchor" aria-label="anchor" href="#converting-between-data-objects"></a></h2>

      <div class="section-desc"><p>Functions that allow the user to convert between SpaMTP Seurat Objects and Cardinal Objects</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="CardinalToSeurat.html">CardinalToSeurat()</a></code>

        </dt>
        <dd>Converts a Cardinal Object into a SpaMTP Seurat Object</dd>
      </dl><dl><dt>

          <code><a href="ConvertSeuratToCardinal.html">ConvertSeuratToCardinal()</a></code>

        </dt>
        <dd>Converts a SpaMTP Seurat object to a Cardinal object, including annotations and metadata</dd>
      </dl></div><div class="section level2">
      <h2 id="binning-spatial-metabolomic-data">Binning Spatial Metabolomic Data<a class="anchor" aria-label="anchor" href="#binning-spatial-metabolomic-data"></a></h2>

      <div class="section-desc"><p>Functions that bin m/z values into a lower resolution/wider peak.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="BinSpaMTP.html">BinSpaMTP()</a></code>

        </dt>
        <dd>Bin SpaMTP Object</dd>
      </dl></div><div class="section level2">
      <h2 id="annotating-m-z-masses">Annotating m/z Masses<a class="anchor" aria-label="anchor" href="#annotating-m-z-masses"></a></h2>

      <div class="section-desc"><p>Functions required for performing and handling m/z annotation using a reference metabolic database</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="AnnotateSM.html">AnnotateSM()</a></code>

        </dt>
        <dd>Annotates m/z values stored in a SpaMTP Object</dd>
      </dl><dl><dt>

          <code><a href="AddCustomMZAnnotations.html">AddCustomMZAnnotations()</a></code>

        </dt>
        <dd>Assign custom annotations to m/z values</dd>
      </dl><dl><dt>

          <code><a href="AddFMP10Annotations.html">AddFMP10Annotations()</a></code>

        </dt>
        <dd>Annotates FMP10 matrix data</dd>
      </dl><dl><dt>

          <code><a href="SearchAnnotations.html">SearchAnnotations()</a></code>

        </dt>
        <dd>Find Annotation</dd>
      </dl><dl><dt>

          <code><a href="GetMZMetadata.html">GetMZMetadata()</a></code>

        </dt>
        <dd>Gets values from a single metadata column for a respective m/z value.</dd>
      </dl><dl><dt>

          <code><a href="FindDuplicateAnnotations.html">FindDuplicateAnnotations()</a></code>

        </dt>
        <dd>Finds if any metabolite is duplicated across multiple m/z values.</dd>
      </dl><dl><dt>

          <code><a href="SubsetMZFeatures.html">SubsetMZFeatures()</a></code>

        </dt>
        <dd>Subset a SpaMTP Seurat Spatial Metabolomic object by a list of m/z's</dd>
      </dl><dl><dt>

          <code><a href="AnnotateBigData.html">AnnotateBigData()</a></code>

        </dt>
        <dd>Annotates vector of m/z values</dd>
      </dl><dl><dt>

          <code><a href="getRefinedAnnotations.html">getRefinedAnnotations()</a></code>

        </dt>
        <dd>Refines and reduces m/z annotations</dd>
      </dl></div><div class="section level2">
      <h2 id="simplify-annotations">Simplify Annotations<a class="anchor" aria-label="anchor" href="#simplify-annotations"></a></h2>

      <div class="section-desc"><p>Functions required for simplifying m/z annotation results</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="CalculateAnnotationStatistics.html">CalculateAnnotationStatistics()</a></code>

        </dt>
        <dd>Calculate annotation statistics for all m/z value suggesting the most likely metabolite based on correlated pathway expression.</dd>
      </dl><dl><dt>

          <code><a href="CalculateSingleAnnotationStatistics.html">CalculateSingleAnnotationStatistics()</a></code>

        </dt>
        <dd>Calculate annotation statistics for a single m/z value suggesting the most likely metabolite based on correlated pathway expression.</dd>
      </dl><dl><dt>

          <code><a href="Pseudo_msms.html">Pseudo_msms()</a></code>

        </dt>
        <dd>Find all library spectra10 above a cosine threshold for each imaging peak, restricting to a global precursor-mz range.</dd>
      </dl><dl><dt>

          <code><a href="CreatePathwayObject.html">CreatePathwayObject()</a></code>

        </dt>
        <dd>Create a SpaMTP Seurat Object containing expression values for all present pathways</dd>
      </dl></div><div class="section level2">
      <h2 id="simplifying-lipid-nomenclature">Simplifying Lipid Nomenclature<a class="anchor" aria-label="anchor" href="#simplifying-lipid-nomenclature"></a></h2>

      <div class="section-desc"><p>Function used to simplify lipid names into general lipid categories, classes, and more</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="RefineLipids.html">RefineLipids()</a></code>

        </dt>
        <dd>Uses common lipid nomenclature to simplify lipid annotations</dd>
      </dl></div><div class="section level2">
      <h2 id="analysis-of-differentially-expressed-peaks">Analysis of Differentially Expressed Peaks<a class="anchor" aria-label="anchor" href="#analysis-of-differentially-expressed-peaks"></a></h2>

      <div class="section-desc"><p>Functions required for performing pseudo-bulking differential expression analysis</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="FindAllDEMs.html">FindAllDEMs()</a></code>

        </dt>
        <dd>Finds differentially expressed m/z values/metabolites between all comparison groups.</dd>
      </dl></div><div class="section level2">
      <h2 id="visualising-deps-analysis">Visualising DEPs Analysis<a class="anchor" aria-label="anchor" href="#visualising-deps-analysis"></a></h2>

      <div class="section-desc"><p>Functions used to generate a heatmap from DEP results</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="DEMsHeatmap.html">DEMsHeatmap()</a></code>

        </dt>
        <dd>Heatmap of Differentially Expressed Metabolites</dd>
      </dl></div><div class="section level2">
      <h2 id="metabolic-and-transcriptomic-pathway-analysis">Metabolic and Transcriptomic Pathway Analysis<a class="anchor" aria-label="anchor" href="#metabolic-and-transcriptomic-pathway-analysis"></a></h2>

      <div class="section-desc"><p>Functions used to perform pathway analysis, both PCA and metabolite/gene set-based (GSEA)</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="FishersPathwayAnalysis.html">FishersPathwayAnalysis()</a></code>

        </dt>
        <dd>Calculates Significant Metabolic Pathways using a Fisher Exact Test</dd>
      </dl><dl><dt>

          <code><a href="FindRegionalPathways.html">FindRegionalPathways()</a></code>

        </dt>
        <dd>Regional Pathway Enrichment</dd>
      </dl><dl><dt>

          <code><a href="RunRAMPgeseca.html">RunRAMPgeseca()</a></code>

        </dt>
        <dd>Runs multilevel Monte-Carlo variant for performing gene sets co-regulation analysis using the RAMP_DB metabolite/gene database.</dd>
      </dl><dl><dt>

          <code><a href="CreatePathwayAssay.html">CreatePathwayAssay()</a></code>

        </dt>
        <dd>Create a Pathway Assay from Gene or Metabolite Data</dd>
      </dl></div><div class="section level2">
      <h2 id="metabolic-and-transcriptomic-pathway-visualisation">Metabolic and Transcriptomic Pathway Visualisation<a class="anchor" aria-label="anchor" href="#metabolic-and-transcriptomic-pathway-visualisation"></a></h2>

      <div class="section-desc"><p>Functions used to visualise pathway analysis results</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="VisualisePathways.html">VisualisePathways()</a></code>

        </dt>
        <dd>Visualise Significant Pathways</dd>
      </dl><dl><dt>

          <code><a href="PlotRegionalPathways.html">PlotRegionalPathways()</a></code>

        </dt>
        <dd>Plot significantly enriched pathways per region</dd>
      </dl><dl><dt>

          <code><a href="PlotPathways.html">PlotPathways()</a></code>

        </dt>
        <dd>Plots the expression profile of a feature set corresponding to specified pathways onto a 2D scatter plot based on a dimensionality reduction technique.</dd>
      </dl><dl><dt>

          <code><a href="PlotPathwaysSpatially.html">PlotPathwaysSpatially()</a></code>

        </dt>
        <dd>Plots the spatial expression profile of a feature set corresponding to specified pathways</dd>
      </dl></div><div class="section level2">
      <h2 id="pathway-network-plotting">Pathway Network Plotting<a class="anchor" aria-label="anchor" href="#pathway-network-plotting"></a></h2>

      <div class="section-desc"><p>Functions used to generate network plots for specified patheays</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="PathwayNetworkPlots.html">PathwayNetworkPlots()</a></code>

        </dt>
        <dd>Constructs an interactive network for exploring spatial metabolomics and transcriptomics data.</dd>
      </dl></div><div class="section level2">
      <h2 id="dimentionality-reduction-analysis">Dimentionality Reduction Analysis<a class="anchor" aria-label="anchor" href="#dimentionality-reduction-analysis"></a></h2>

      <div class="section-desc"><p>Functions that are used for calculating PCA embeddings and projections based on SM and/or ST data</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="RunMetabolicPCA.html">RunMetabolicPCA()</a></code>

        </dt>
        <dd>Generates PCA analysis results for a SpaMTP Seurat Object</dd>
      </dl><dl><dt>

          <code><a href="RunSpatialGraphPCA.html">RunSpatialGraphPCA()</a></code>

        </dt>
        <dd>Perform Dimensionality Reduction using Graph-Regularised PCA on Spatial Data</dd>
      </dl><dl><dt>

          <code><a href="GetKmeanClusters.html">GetKmeanClusters()</a></code>

        </dt>
        <dd>Perform K-means clustering on a specified reduction</dd>
      </dl></div><div class="section level2">
      <h2 id="spamtp-metabolic-data-visualisation">SpaMTP Metabolic Data Visualisation<a class="anchor" aria-label="anchor" href="#spamtp-metabolic-data-visualisation"></a></h2>

      <div class="section-desc"><p>Functions that can be used to visualize data from SpaMTP object</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="ImageMZPlot.html">ImageMZPlot()</a></code>

        </dt>
        <dd>Plot expression of m/z values spatially</dd>
      </dl><dl><dt>

          <code><a href="ImageMZAnnotationPlot.html">ImageMZAnnotationPlot()</a></code>

        </dt>
        <dd>Plot expression of annotated metabolites spatially</dd>
      </dl><dl><dt>

          <code><a href="SpatialMZPlot.html">SpatialMZPlot()</a></code>

        </dt>
        <dd>Plot expression of m/z values spatially for a Spatial SpaMTP Seurat Objects.</dd>
      </dl><dl><dt>

          <code><a href="SpatialMZAnnotationPlot.html">SpatialMZAnnotationPlot()</a></code>

        </dt>
        <dd>Plot expression of metabolites in spatially from a Spatial SpaMTP Objects.</dd>
      </dl><dl><dt>

          <code><a href="Plot3DFeature.html">Plot3DFeature()</a></code>

        </dt>
        <dd>Generates a 3D spatial feature plot from a SpaMTP object</dd>
      </dl><dl><dt>

          <code><a href="MassIntensityPlot.html">MassIntensityPlot()</a></code>

        </dt>
        <dd>Plot mass intensity spectra</dd>
      </dl><dl><dt>

          <code><a href="DensityMap.html">DensityMap()</a></code>

        </dt>
        <dd>Generates interactive 3D spatial density plot for m/z values</dd>
      </dl><dl><dt>

          <code><a href="CheckAlignment.html">CheckAlignment()</a></code>

        </dt>
        <dd>Check multi-modal coordinate alignment</dd>
      </dl></div><div class="section level2">
      <h2 id="interactive-spatial-binning-visualisation">Interactive Spatial Binning Visualisation<a class="anchor" aria-label="anchor" href="#interactive-spatial-binning-visualisation"></a></h2>

      <div class="section-desc"><p>Interactive plot that displays spatial changes to m/z intensity values based on changes to bin size.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="InteractiveSpatialPlot.html">InteractiveSpatialPlot()</a></code>

        </dt>
        <dd>Interactive Spatial Plot for visualising different m/z bin sizes</dd>
      </dl></div><div class="section level2">
      <h2 id="additional-spamtp-functions">Additional SpaMTP Functions<a class="anchor" aria-label="anchor" href="#additional-spamtp-functions"></a></h2>

      <div class="section-desc"><p>Functions that can be used to find the closest metabolite and bin the expression of multiple metabolites into one</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="FindNearestMZ.html">FindNearestMZ()</a></code>

        </dt>
        <dd>Finds the nearest m/z peak to a given value in a SpaMTP Object</dd>
      </dl><dl><dt>

          <code><a href="BinMetabolites.html">BinMetabolites()</a></code>

        </dt>
        <dd>Sums the intensity values of multiple m/z values into one</dd>
      </dl></div><div class="section level2">
      <h2 id="spatial-analysis-of-metabolomic-data">Spatial Analysis of Metabolomic Data<a class="anchor" aria-label="anchor" href="#spatial-analysis-of-metabolomic-data"></a></h2>

      <div class="section-desc"><p>Functions used to identify spatially correlated features (metabolites/genes)</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="FindCorrelatedFeatures.html">FindCorrelatedFeatures()</a></code>

        </dt>
        <dd>Find top features and metabolites that are strongly correlated with a given feature</dd>
      </dl><dl><dt>

          <code><a href="FindSpatiallyVariableMetabolites.html">FindSpatiallyVariableMetabolites()</a></code>

        </dt>
        <dd>Find Spatially Variable Metabolites</dd>
      </dl><dl><dt>

          <code><a href="GetSpatiallyVariableMetabolites.html">GetSpatiallyVariableMetabolites()</a></code>

        </dt>
        <dd>Get top spatially variable metabolites</dd>
      </dl><dl><dt>

          <code><a href="RowVar.html">RowVar()</a></code>

        </dt>
        <dd>Compute the row variances for each m/z value</dd>
      </dl></div><div class="section level2">
      <h2 id="multi-omic-data-integration">Multi-Omic Data Integration<a class="anchor" aria-label="anchor" href="#multi-omic-data-integration"></a></h2>

      <div class="section-desc"><p>Functions used to Align, Map and Integrate Spatial Metabolomic and Transcriptomic data</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="MapSpatialOmics.html">MapSpatialOmics()</a></code>

        </dt>
        <dd>Maps Spatial Metabolomic (MALDI) data to corresponding Spatial Transcriptomics data and coordinates.</dd>
      </dl><dl><dt>

          <code><a href="AlignSpatialOmics.html">AlignSpatialOmics()</a></code>

        </dt>
        <dd>Interactive app for SM and ST coordinate alignment</dd>
      </dl><dl><dt>

          <code><a href="MultiOmicIntegration.html">MultiOmicIntegration()</a></code>

        </dt>
        <dd>Mult-Omic data integration</dd>
      </dl><dl><dt>

          <code><a href="CreateMergedModalityAssay.html">CreateMergedModalityAssay()</a></code>

        </dt>
        <dd>Create a singular multiomics assay by merging data from multiple assays.</dd>
      </dl></div><div class="section level2">
      <h2 id="adding-image-data">Adding Image Data<a class="anchor" aria-label="anchor" href="#adding-image-data"></a></h2>

      <div class="section-desc"><p>Functions used to align and add an image to a spatial metabolic SpaMTP object.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="AddSMImage.html">AddSMImage()</a></code>

        </dt>
        <dd>Manually align an image (e.g. H&amp;E, Immuno) to a SM SpaMTP dataset</dd>
      </dl></div><div class="section level2">
      <h2 id="roi-selection">ROI Selection<a class="anchor" aria-label="anchor" href="#roi-selection"></a></h2>

      <div class="section-desc"><p>Functions used subset or select specific spatial regions of interest.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="SelectROIs.html">SelectROIs()</a></code>

        </dt>
        <dd>Launch an Interactive ROI Annotation App for Seurat Spatial Data</dd>
      </dl></div><div class="section level2">
      <h2 id="pre-processing-spamtp-metabolic-data">Pre-Processing SpaMTP Metabolic Data<a class="anchor" aria-label="anchor" href="#pre-processing-spamtp-metabolic-data"></a></h2>

      <div class="section-desc"><p>Functions for normalising and visualising the pre-processing of SpaMTP datasets</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="NormalizeSMData.html">NormalizeSMData()</a></code>

        </dt>
        <dd>Normalizes m/z intensity data stored in a SpaMTP Seurat Object</dd>
      </dl><dl><dt>

          <code><a href="TMMNormalize.html">TMMNormalize()</a></code>

        </dt>
        <dd>Performs TMM normalization between categories based on a specified ident</dd>
      </dl><dl><dt>

          <code><a href="MZRidgePlot.html">MZRidgePlot()</a></code>

        </dt>
        <dd>Generates a ridge plot of spatial metabolic intensity data</dd>
      </dl><dl><dt>

          <code><a href="MZVlnPlot.html">MZVlnPlot()</a></code>

        </dt>
        <dd>Generates a violin plot of spatial metabolic intensity data</dd>
      </dl><dl><dt>

          <code><a href="MZBoxPlot.html">MZBoxPlot()</a></code>

        </dt>
        <dd>Generates a Boxplot of spatial metabolic intensity data</dd>
      </dl></div><div class="section level2">
      <h2 id="exporting-spamtp-data">Exporting SpaMTP Data<a class="anchor" aria-label="anchor" href="#exporting-spamtp-data"></a></h2>

      <div class="section-desc"><p>Function to export SpaMTP data in .mtx, barcodes.csv, features.csv, metadata.csv, and feature.metadata.csv files</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="SaveSpaMTPData.html">SaveSpaMTPData()</a></code>

        </dt>
        <dd>Saves SpaMTP Object</dd>
      </dl></div><div class="section level2">
      <h2 id="reference-metabolite-datasets">Reference Metabolite Datasets<a class="anchor" aria-label="anchor" href="#reference-metabolite-datasets"></a></h2>

      <div class="section-desc"><p>Metabolite datasets used for annotating m/z masses</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="HMDB_db.html">HMDB_db</a></code>

        </dt>
        <dd>HMDB_db: A cleaned version of the reference metabolomics dataset from the Human Metabolome Database (HMDB)</dd>
      </dl><dl><dt>

          <code><a href="Lipidmaps_db.html">Lipidmaps_db</a></code>

        </dt>
        <dd>Lipidmaps_db: A cleaned version of the lipid database from LIPID MAPS</dd>
      </dl><dl><dt>

          <code><a href="Chebi_db.html">Chebi_db</a></code>

        </dt>
        <dd>Chebi_db: Cleaned ChEBI <code>(Chemical entities of biological interest)</code> reference dataset</dd>
      </dl><dl><dt>

          <code><a href="GNPS_db.html">GNPS_db</a></code>

        </dt>
        <dd>GNPS_db: A cleaned database of metabolites from GNPS</dd>
      </dl><dl><dt>

          <code><a href="filtered_fmp10.html">filtered_fmp10</a></code>

        </dt>
        <dd>filtered_fmp10: data.frame containing FMP10+ metabolite mappings</dd>
      </dl></div><div class="section level2">
      <h2 id="metabolic-pathway-datasets">Metabolic Pathway Datasets<a class="anchor" aria-label="anchor" href="#metabolic-pathway-datasets"></a></h2>

      <div class="section-desc"><p>Various datasets required for pathway analysis</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="adduct_file.html">adduct_file</a></code>

        </dt>
        <dd>adduct_file: A dataframe containing possible adducts used for pathway analysis</dd>
      </dl><dl><dt>

          <code><a href="analyte.html">analyte</a></code>

        </dt>
        <dd>analyte: A dataframe containing ID's of possible RAMP analytes</dd>
      </dl><dl><dt>

          <code><a href="analytehaspathway.html">analytehaspathway</a></code>

        </dt>
        <dd>analytehaspathway: A dataframe containing RAMP_pathway ID's</dd>
      </dl><dl><dt>

          <code><a href="chem_props.html">chem_props</a></code>

        </dt>
        <dd>chem_props: A database containing the chemical properties and metadata of each RAMP_DB analyte</dd>
      </dl><dl><dt>

          <code><a href="pathway.html">pathway</a></code>

        </dt>
        <dd>pathway: A dataframe containing RAMP_DB pathways and their relative metadata</dd>
      </dl><dl><dt>

          <code><a href="source_df.html">source_df</a></code>

        </dt>
        <dd>source_df: A dataframe containing source information about RAMP_ID analyte used for analysis</dd>
      </dl></div><div class="section level2">
      <h2 id="pathway-network-datasets">Pathway Network Datasets<a class="anchor" aria-label="anchor" href="#pathway-network-datasets"></a></h2>

      <div class="section-desc"><p>Various datasets required for generated pathway network plots</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="RAMP_hmdb.html">RAMP_hmdb</a></code>

        </dt>
        <dd>RAMP_hmdb: A list containing network plot information about pathways from the HMDB database</dd>
      </dl><dl><dt>

          <code><a href="RAMP_Reactome.html">RAMP_Reactome</a></code>

        </dt>
        <dd>RAMP_Reactome: A list containing network plot information about pathways from the Reactome database</dd>
      </dl><dl><dt>

          <code><a href="RAMP_kegg.html">RAMP_kegg</a></code>

        </dt>
        <dd>RAMP_kegg: A list containing network plot information about pathways from the KEGG database</dd>
      </dl><dl><dt>

          <code><a href="RAMP_wikipathway.html">RAMP_wikipathway</a></code>

        </dt>
        <dd>RAMP_wikipathway: A list containing network plot information about pathways from the Wiki database</dd>
      </dl><dl><dt>

          <code><a href="reaction_type.html">reaction_type</a></code>

        </dt>
        <dd>reaction_type: data.frame containing reaction type mappings</dd>
      </dl></div><div class="section level2">
      <h2 id="cardinal-wrapper-functions">Cardinal Wrapper Functions<a class="anchor" aria-label="anchor" href="#cardinal-wrapper-functions"></a></h2>

      <div class="section-desc"><p>Functions used to alter Cardinal Objects.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="add_ssc_annotation.html">add_ssc_annotation()</a></code>

        </dt>
        <dd>Adds Cardinal ssc segmentation annotation to m/z count data object</dd>
      </dl></div><div class="section level2">
      <h2 id="additional-worker-functions">Additional Worker Functions<a class="anchor" aria-label="anchor" href="#additional-worker-functions"></a></h2>

      <div class="section-desc"><p>Various helper functions used by SpaMTP</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="verbose_message.html">verbose_message()</a></code>

        </dt>
        <dd>Helper function for suppressing function progress messages</dd>
      </dl><dl><dt>

          <code><a href="subset_SPM.html">subset_SPM()</a></code>

        </dt>
        <dd>Subsets SpaMTP Seurat Object containing FOVs</dd>
      </dl><dl><dt>

          <code><a href="check_cardinal_version.html">check_cardinal_version()</a></code>

        </dt>
        <dd>Check Cardinal Version</dd>
      </dl></div><div class="section level2">
      <h2 id="utility-functions">Utility Functions<a class="anchor" aria-label="anchor" href="#utility-functions"></a></h2>




    </div><div class="section level2">

      <h3 id="binning-helper-functions">Binning Helper Functions<a class="anchor" aria-label="anchor" href="#binning-helper-functions"></a></h3>
      <div class="section-desc"><p>Helper functions for binning m/z intensity spectra for SpaMTP objects</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="BinnedCardinalToSeurat.html">BinnedCardinalToSeurat()</a></code>

        </dt>
        <dd>Converts a SpaMTP binned Cardinal Object into a SpaMTP Seurat Object</dd>
      </dl><dl><dt>

          <code><a href="spectral_binning.html">spectral_binning()</a></code>

        </dt>
        <dd>Spectral binning of intensity values stored in a Matrix object, converted from matter.</dd>
      </dl></div><div class="section level2">

      <h3 id="plotting-helper-functions">Plotting Helper Functions<a class="anchor" aria-label="anchor" href="#plotting-helper-functions"></a></h3>
      <div class="section-desc"><p>Helper functions used for SpaMTP data visualisation</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="bin.mz.html">bin.mz()</a></code>

        </dt>
        <dd>Bins multiple m/z values into one.</dd>
      </dl><dl><dt>

          <code><a href="plusminus.html">plusminus()</a></code>

        </dt>
        <dd>Identifies all mz peaks within a plus-minus range of the target_mz</dd>
      </dl><dl><dt>

          <code><a href="plot_plus_minus.html">plot_plus_minus()</a></code>

        </dt>
        <dd>Helper Function to generate merged counts within the plus minus range provided</dd>
      </dl><dl><dt>

          <code><a href="check_column_type.html">check_column_type()</a></code>

        </dt>
        <dd>Helper function to determine if a column contains categorical or continuous Data</dd>
      </dl><dl><dt>

          <code><a href="pixelPlot.html">pixelPlot()</a></code>

        </dt>
        <dd>Helper function for converting Seurat Class ggplots from spot to pixel layout</dd>
      </dl></div><div class="section level2">

      <h3 id="annotation-helper-functions">Annotation Helper Functions<a class="anchor" aria-label="anchor" href="#annotation-helper-functions"></a></h3>
      <div class="section-desc"><p>Helper functions required for running ‘AnnotateSM()’</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="annotateTable.html">annotateTable()</a></code>

        </dt>
        <dd>Annotates m/z values sotred in a data.frame based on reference metabolite dataset</dd>
      </dl><dl><dt>

          <code><a href="labels_to_show.html">labels_to_show()</a></code>

        </dt>
        <dd>Filters the annotation list to only include the first n number of annotations per m/z</dd>
      </dl><dl><dt>

          <code><a href="add_backslashes_to_specialfeatures.html">add_backslashes_to_specialfeatures()</a></code>

        </dt>
        <dd>Adds in backslashes required to take into account special using grepl such as brackets and +</dd>
      </dl><dl><dt>

          <code><a href="check_and_truncate_adduct_vector.html">check_and_truncate_adduct_vector()</a></code>

        </dt>
        <dd>Checks if the complete adduct is in the data base, else returns a truncated adduct</dd>
      </dl><dl><dt>

          <code><a href="db_adduct_filter.html">db_adduct_filter()</a></code>

        </dt>
        <dd>Filters the provided metabolomic database by polarity and adducts</dd>
      </dl><dl><dt>

          <code><a href="is_formula_valid.html">is_formula_valid()</a></code>

        </dt>
        <dd>Checks if a formula contains only the allowed elements</dd>
      </dl><dl><dt>

          <code><a href="formula_filter.html">formula_filter()</a></code>

        </dt>
        <dd>Filters reference Database to only select natural elements</dd>
      </dl><dl><dt>

          <code><a href="calculate_bounds.html">calculate_bounds()</a></code>

        </dt>
        <dd>Calculates the mz range of the observed_df</dd>
      </dl><dl><dt>

          <code><a href="ppm_error.html">ppm_error()</a></code>

        </dt>
        <dd>Calculates the ppm error as a valve</dd>
      </dl><dl><dt>

          <code><a href="ppm_range_match.html">ppm_range_match()</a></code>

        </dt>
        <dd>Calculates the ppm range and check if mz values are within the range <ul><li><p>Returns TRUE if match is found and false if no match.</p></li> </ul></dd>
      </dl><dl><dt>

          <code><a href="proc_db.html">proc_db()</a></code>

        </dt>
        <dd>Searches observed mz values against the data base list and returns matching annotations</dd>
      </dl></div><div class="section level2">

      <h3 id="differential-abundance-helper-functions">Differential Abundance Helper Functions<a class="anchor" aria-label="anchor" href="#differential-abundance-helper-functions"></a></h3>
      <div class="section-desc"><p>Helper functions for calculating and plotting differentially expressed metabolites</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="run_pooling.html">run_pooling()</a></code>

        </dt>
        <dd>Pools SpaMTP Seurat object into random pools for pseudo-bulking.</dd>
      </dl><dl><dt>

          <code><a href="run_DE.html">run_DE()</a></code>

        </dt>
        <dd>Runs EdgeR analysis for pooled data</dd>
      </dl><dl><dt>

          <code><a href="save_pheatmap_as_pdf.html">save_pheatmap_as_pdf()</a></code>

        </dt>
        <dd>Saves a DEMsHeatmap as a PDF</dd>
      </dl></div><div class="section level2">

      <h3 id="pathway-analysis-helper-functions">Pathway Analysis Helper Functions<a class="anchor" aria-label="anchor" href="#pathway-analysis-helper-functions"></a></h3>
      <div class="section-desc"><p>Helper functions used for running pathway analysis</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="PlotSinglePathway.html">PlotSinglePathway()</a></code>

        </dt>
        <dd>Plots the expression profile of a feature set corresponding to specified pathways onto a 2D scatter plot based on a dimensionality reduction technique.</dd>
      </dl><dl><dt>

          <code><a href="PlotSinglePathwaySpatially.html">PlotSinglePathwaySpatially()</a></code>

        </dt>
        <dd>Plot expression profile of a single RAMP pathway spatially</dd>
      </dl><dl><dt>

          <code><a href="addGesecaScores.html">addGesecaScores()</a></code>

        </dt>
        <dd>Add GESECA Scores to SpaMTP Object</dd>
      </dl><dl><dt>

          <code><a href="get_analytes_db.html">get_analytes_db()</a></code>

        </dt>
        <dd>Helper function for building a pathway db based on detected</dd>
      </dl><dl><dt>

          <code><a href="list_to_pprcomp.html">list_to_pprcomp()</a></code>

        </dt>
        <dd>Creates a pprcomp object based on an input list</dd>
      </dl></div><div class="section level2">

      <h3 id="multi-omic-helper-functions">Multi-Omic Helper Functions<a class="anchor" aria-label="anchor" href="#multi-omic-helper-functions"></a></h3>
      <div class="section-desc"><p>Helper functions used in multi-omic analysis functions</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="kneighbors_graph.html">kneighbors_graph()</a></code>

        </dt>
        <dd>Construct a k-Nearest Neighbour Graph from Spatial Coordinates.</dd>
      </dl><dl><dt>

          <code><a href="get_square_coordinates.html">get_square_coordinates()</a></code>

        </dt>
        <dd>This function computes the coordinates of a square's four corners based on a given center point and width.</dd>
      </dl><dl><dt>

          <code><a href="lowresMapping.html">lowresMapping()</a></code>

        </dt>
        <dd>Maps SM pixels to low resolution ST data</dd>
      </dl><dl><dt>

          <code><a href="hiresMapping.html">hiresMapping()</a></code>

        </dt>
        <dd>Maps SM pixels to high resolution ST data</dd>
      </dl><dl><dt>

          <code><a href="translate.html">translate()</a></code>

        </dt>
        <dd>Creates a transformation matrix that translates an object in 2D</dd>
      </dl><dl><dt>

          <code><a href="mirror.html">mirror()</a></code>

        </dt>
        <dd>Creates a transformation matrix that mirrors an object in 2D along either the x axis or y axis around its center of mass</dd>
      </dl><dl><dt>

          <code><a href="stretch.html">stretch()</a></code>

        </dt>
        <dd>Stretch along angle</dd>
      </dl><dl><dt>

          <code><a href="rigid.rot.html">rigid.rot()</a></code>

        </dt>
        <dd>Creates a transformation matrix for rotation</dd>
      </dl><dl><dt>

          <code><a href="rigid.transf.html">rigid.transf()</a></code>

        </dt>
        <dd>Creates a transformation matrix for rotation and translation</dd>
      </dl><dl><dt>

          <code><a href="rigid.transl.html">rigid.transl()</a></code>

        </dt>
        <dd>Creates a transformation matrix for translation with an offset of (h, k)</dd>
      </dl><dl><dt>

          <code><a href="rigid.refl.html">rigid.refl()</a></code>

        </dt>
        <dd>Creates a transformation matrix for reflection</dd>
      </dl><dl><dt>

          <code><a href="rigid.stretch.html">rigid.stretch()</a></code>

        </dt>
        <dd>Creates a transformation matrix for stretching</dd>
      </dl><dl><dt>

          <code><a href="combine.tr.html">combine.tr()</a></code>

        </dt>
        <dd>Combines rigid tranformation matrices</dd>
      </dl></div><div class="section level2">

      <h3 id="pre-processing-helper-functions">Pre-Processing Helper Functions<a class="anchor" aria-label="anchor" href="#pre-processing-helper-functions"></a></h3>
      <div class="section-desc"><p>Helper functions for generating pre-processing plots</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="statPlot.html">statPlot()</a></code>

        </dt>
        <dd>Helper function for QC plots by generating intensity count data</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/agc888" class="external-link">Andrew Causer</a>, <a href="https://github.com/BCRL-tylu" class="external-link">Tianyao Lu</a>, <a href="https://github.com/ChemCharles" class="external-link">Christopher C.J. Fitzgerald</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

