<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Maps Spatial Metabolomic (MALDI) data to corresponding Spatial Transcriptomics data and coordinates. — MapSpatialOmics • SpaMTP</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Maps Spatial Metabolomic (MALDI) data to corresponding Spatial Transcriptomics data and coordinates. — MapSpatialOmics"><meta name="description" content="Maps Spatial Metabolomic (MALDI) data to corresponding Spatial Transcriptomics data and coordinates."><meta property="og:description" content="Maps Spatial Metabolomic (MALDI) data to corresponding Spatial Transcriptomics data and coordinates."><meta property="og:image" content="https://genomicsmachinelearning.github.io/SpaMTP/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SpaMTP</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/SpaMTP.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/Mouse_Urinary_Bladder.html">Spatial Metabolomics Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Multi-Omic_Mouse_Brain.html">Spatial Multi-Omics Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/SpaMTP_Additional_Features.html">Additional SpaMTP Features</a></li>
    <li><a class="dropdown-item" href="../articles/Single_Cell_MultiOmics.html">Simulated Single Cell Multi-Omics Analysis</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Updates</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/GenomicsMachineLearning/SpaMTP" aria-label="GitHub"><span class="fa fa-github"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Maps Spatial Metabolomic (MALDI) data to corresponding Spatial Transcriptomics data and coordinates.</h1>
      <small class="dont-index">Source: <a href="https://github.com/genomicsmachinelearning/SpaMTP/blob/HEAD/R/SpatialMultiomicAlignment.R" class="external-link"><code>R/SpatialMultiomicAlignment.R</code></a></small>
      <div class="d-none name"><code>MapSpatialOmics.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Maps Spatial Metabolomic (MALDI) data to corresponding Spatial Transcriptomics data and coordinates.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">MapSpatialOmics</span><span class="op">(</span></span>
<span>  <span class="va">SM.data</span>,</span>
<span>  <span class="va">ST.data</span>,</span>
<span>  ST.hires <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  SM.assay <span class="op">=</span> <span class="st">"Spatial"</span>,</span>
<span>  ST.assay <span class="op">=</span> <span class="st">"Spatial"</span>,</span>
<span>  SM.fov <span class="op">=</span> <span class="st">"fov"</span>,</span>
<span>  ST.image <span class="op">=</span> <span class="st">"slice1"</span>,</span>
<span>  ST.scale.factor <span class="op">=</span> <span class="st">"hires"</span>,</span>
<span>  SM.pixel.width <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  overlap.threshold <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  annotations <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  add.metadata <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  merge.unique.metadata <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  map.data <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  new_SPT.assay <span class="op">=</span> <span class="st">"SPT"</span>,</span>
<span>  new_SPM.assay <span class="op">=</span> <span class="st">"SPM"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-sm-data">SM.data<a class="anchor" aria-label="anchor" href="#arg-sm-data"></a></dt>
<dd><p>A SpaMTP Seurat object representing the Spatial Metabolomics data.</p></dd>


<dt id="arg-st-data">ST.data<a class="anchor" aria-label="anchor" href="#arg-st-data"></a></dt>
<dd><p>A Seurat object representing the Spatial Transcriptomics data.</p></dd>


<dt id="arg-st-hires">ST.hires<a class="anchor" aria-label="anchor" href="#arg-st-hires"></a></dt>
<dd><p>Boolean string defining if the ST data is at a higher resolution compared to the SM pixel data. For example, generally Visium data will be lower res whereas Xenium/single-cell resolution spatial data will be a higher resolution (default = FALSE).</p></dd>


<dt id="arg-sm-assay">SM.assay<a class="anchor" aria-label="anchor" href="#arg-sm-assay"></a></dt>
<dd><p>Character string defining the Seurat assay that contains the annotated counts and metadata corresponding to the m/z values (default = "Spatial").</p></dd>


<dt id="arg-st-assay">ST.assay<a class="anchor" aria-label="anchor" href="#arg-st-assay"></a></dt>
<dd><p>Character string specifying the current assay to use to extract transcriptional data from (default = "Spatial").</p></dd>


<dt id="arg-sm-fov">SM.fov<a class="anchor" aria-label="anchor" href="#arg-sm-fov"></a></dt>
<dd><p>Character string of the image fov associated with the spatial metabolomic data (default = "fov").</p></dd>


<dt id="arg-st-image">ST.image<a class="anchor" aria-label="anchor" href="#arg-st-image"></a></dt>
<dd><p>Character string matching the image name associated with the ST data such as 'fov' or 'slice1' object (default = "slice1").</p></dd>


<dt id="arg-st-scale-factor">ST.scale.factor<a class="anchor" aria-label="anchor" href="#arg-st-scale-factor"></a></dt>
<dd><p>Character string defining the image resolution associated with the Visium image pixel data. If <code>NULL</code> the full-res coordinates will be used and no scaling will be performed. Note: This parameter is only required for aligning lowres ST data (default = "hires").</p></dd>


<dt id="arg-sm-pixel-width">SM.pixel.width<a class="anchor" aria-label="anchor" href="#arg-sm-pixel-width"></a></dt>
<dd><p>Numeric value defining the width of each SM pixel. If set to <code>NULL</code>, the median pixel width will be calculated based on the distance between each pixel (default = NULL).</p></dd>


<dt id="arg-overlap-threshold">overlap.threshold<a class="anchor" aria-label="anchor" href="#arg-overlap-threshold"></a></dt>
<dd><p>Numeric value defining the overlap proportion threshold for a SM pixel to be associated with a ST spot. For example, if res_increase = 0.2 then pixels that have at least 20% area overlap with the respective visium spot will be assigned a match. Note: This parameter is only required for aligning lowres ST data (default = 0.2).</p></dd>


<dt id="arg-annotations">annotations<a class="anchor" aria-label="anchor" href="#arg-annotations"></a></dt>
<dd><p>Boolean value indicating if the Spatial Metabolomics (MALDI) Seurat object contains annotations assigned to m/z values (default = TRUE).</p></dd>


<dt id="arg-add-metadata">add.metadata<a class="anchor" aria-label="anchor" href="#arg-add-metadata"></a></dt>
<dd><p>Boolean defining whether to add the current metadata stored in the SM object to the new mapped multi-omic SpaMTP object (default = TRUE)</p></dd>


<dt id="arg-merge-unique-metadata">merge.unique.metadata<a class="anchor" aria-label="anchor" href="#arg-merge-unique-metadata"></a></dt>
<dd><p>Boolean indicating whether to summaries duplicated metadata terms to only store unique values in the metadata. Note: This parameter is only required for aligning lowres ST data, and <code>add.metadata</code> must be set to <code>TRUE</code> for this functionality to be implemented (default = TRUE).</p></dd>


<dt id="arg-map-data">map.data<a class="anchor" aria-label="anchor" href="#arg-map-data"></a></dt>
<dd><p>Boolean indicating whether to map normalised/additional data stored in the <code>data</code> slot of the SpaMTP assay. Note: this process is computationally expensive with large datasets (default = FALSE).</p></dd>


<dt id="arg-new-spt-assay">new_SPT.assay<a class="anchor" aria-label="anchor" href="#arg-new-spt-assay"></a></dt>
<dd><p>Character string defining the assay name of the new overlaid SpaMTP Seurat object containing all updated transcriptomics data (default = "SPT").</p></dd>


<dt id="arg-new-spm-assay">new_SPM.assay<a class="anchor" aria-label="anchor" href="#arg-new-spm-assay"></a></dt>
<dd><p>Character string defining the assay name of the new overlaid SpaMTP Seurat object containing all updated metabolomic data (default = "SPM").</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Boolean value indicating whether to print informative progression update messages and progress bars (default = TRUE).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A SpaMTP Seurat object with Spatial Metabolomic data mapped to equivalent Spatial Transcripomics coordinates (Visium spots/Xenium cells).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Mapping MALDI data to equivalent Visium spots</span></span></span>
<span class="r-in"><span><span class="co"># MapSpatialOmics(VisiumObj, SeuratObj, ST.scale.factor = "hires", SM.assay = "Spatial", ST.assay = "Spatial")</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#' ## Mapping MALDI data to equivalent Xenium cells</span></span></span>
<span class="r-in"><span><span class="co"># MapSpatialOmics(VisiumObj, SeuratObj, SM.assay = "Spatial", ST.assay = "Xenium")</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/agc888" class="external-link">Andrew Causer</a>, <a href="https://github.com/BCRL-tylu" class="external-link">Tianyao Lu</a>, <a href="https://github.com/ChemCharles" class="external-link">Christopher C.J. Fitzgerald</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

