<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>SpaMTP: Simulated Single Cell Multi-Omics Analysis • SpaMTP</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="SpaMTP: Simulated Single Cell Multi-Omics Analysis">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SpaMTP</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/SpaMTP.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/Mouse_Urinary_Bladder.html">Spatial Metabolomics Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Multi-Omic_Mouse_Brain.html">Spatial Multi-Omics Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/SpaMTP_Additional_Features.html">Additional SpaMTP Features</a></li>
    <li><a class="dropdown-item" href="../articles/Single_Cell_MultiOmics.html">Simulated Single Cell Multi-Omics Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Big_Data.html">Handling Big Datasets</a></li>
    <li><a class="dropdown-item" href="../articles/Annotation_Refinement.html">Refinning Metabolite Annotations</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Updates</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/GenomicsMachineLearning/SpaMTP" aria-label="GitHub"><span class="fa fa-github"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<script src="Single_Cell_MultiOmics_files/kePrint-0.0.1/kePrint.js"></script><link href="Single_Cell_MultiOmics_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>SpaMTP: Simulated Single Cell Multi-Omics Analysis</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/genomicsmachinelearning/SpaMTP/blob/HEAD/vignettes/Single_Cell_MultiOmics.Rmd" class="external-link"><code>vignettes/Single_Cell_MultiOmics.Rmd</code></a></small>
      <div class="d-none name"><code>Single_Cell_MultiOmics.Rmd</code></div>
    </div>

    
    
<p>This vignette will use <a href="https://github.com/GenomicsMachineLearning/SpaMTP" class="external-link"><strong><em>SpaMTP</em></strong></a>
to integrate single cell resolution spatial transcriptomics data with
spatial metabolic data. This vignette will use simulated Xenium (ST) and
simulated MALDI-TOF (SM) data.</p>
<p>Author: Andrew Causer</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Install SpaMTP if not previously installed</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/genomicsmachinelearning/SpaMTP" class="external-link">"SpaMTP"</a></span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"GenomicsMachineLearning/SpaMTP"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#General Libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/genomicsmachinelearning/SpaMTP" class="external-link">SpaMTP</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://www.cardinalmsi.org" class="external-link">Cardinal</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level5">
<h5 id="load-processed-data">Load Processed Data<a class="anchor" aria-label="anchor" href="#load-processed-data"></a>
</h5>
<p>Here, we will be using the simulated data which has synthetic
clustering and cell type names. The gene and metabolite names are also
arbitrary, this data is only used to highlight
<strong><em>SpaMTP’s</em></strong> functionally with single cell spatial
datasets.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xenium</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://zenodo.org/records/17247007/files/Sim_Xenium.RDS?download=1"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">MALDI</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://zenodo.org/records/17247007/files/Sim_MALDI.RDS?download=1"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">MALDI</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/UpdateSeuratObject.html" class="external-link">UpdateSeuratObject</a></span><span class="op">(</span><span class="va">MALDI</span><span class="op">)</span> <span class="co"># Only Required if SeuratObjects =&gt; 5.0.2</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualising-simulated-datasets">Visualising Simulated Datasets<a class="anchor" aria-label="anchor" href="#visualising-simulated-datasets"></a>
</h2>
<p>We can visualise the simulated datasets and observe the difference in
resolution between our single cell ST and lower resolution SM data.</p>
<p>First, lets look at our Xenium data:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/ImageDimPlot.html" class="external-link">ImageDimPlot</a></span><span class="op">(</span><span class="va">xenium</span>, group.by <span class="op">=</span> <span class="st">"celltype"</span>, size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="Single_Cell_MultiOmics_files/figure-html/unnamed-chunk-4-1.png" width="1056" style="display: block; margin: auto;"></p>
<p>Now, lets look at our MALDI data:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/ImageDimPlot.html" class="external-link">ImageDimPlot</a></span><span class="op">(</span><span class="va">MALDI</span>, group.by <span class="op">=</span> <span class="st">"clusters"</span>, size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="Single_Cell_MultiOmics_files/figure-html/unnamed-chunk-5-1.png" width="1056" style="display: block; margin: auto;"></p>
<p>Based on these plots, we can see that our Xenium data is at a much
higher resolution, compared to the MALDI data. We can zoom in and see
this in more detail:</p>
<details><summary><em>Setting Futures for FOV Creation</em>
</summary><div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Futures may be required to use Seurat's `Crop()` function</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://future.futureverse.org" class="external-link">future</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'future'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:Cardinal':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     reset, run</span></span>
<span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="st">"multicore"</span>, workers <span class="op">=</span> <span class="fl">16</span><span class="op">)</span> <span class="co"># Use all 16 cores</span></span>
<span><span class="co">#&gt; Warning in supportsMulticoreAndRStudio(...): [ONE-TIME WARNING] Forked</span></span>
<span><span class="co">#&gt; processing ('multicore') is not supported when running R from RStudio because</span></span>
<span><span class="co">#&gt; it is considered unstable. For more details, how to control forked processing</span></span>
<span><span class="co">#&gt; or not, and how to silence this warning in future R sessions, see</span></span>
<span><span class="co">#&gt; ?parallelly::supportsMulticore</span></span>
<span></span>
<span><span class="co"># Each core can have up to 16 GB of memory (16 GB / 16 cores)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>future.globals.maxSize <span class="op">=</span> <span class="fl">16000</span> <span class="op">*</span> <span class="fl">1024</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span> <span class="co"># 16 GB per core</span></span></code></pre></div>
</details><div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate a smaller FOV to look at a zoomed in region of the ST data</span></span>
<span><span class="va">cropped.coords.xenium</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/Crop.html" class="external-link">Crop</a></span><span class="op">(</span><span class="va">xenium</span><span class="op">[[</span><span class="st">"segmentations"</span><span class="op">]</span><span class="op">]</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2000</span>, <span class="fl">3000</span><span class="op">)</span>, x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">8500</span>, <span class="fl">9500</span><span class="op">)</span>, coords <span class="op">=</span> <span class="st">"tissue"</span><span class="op">)</span></span>
<span><span class="va">xenium</span><span class="op">[[</span><span class="st">"zoom"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">cropped.coords.xenium</span></span>
<span></span>
<span><span class="co"># Generate a smaller FOV to look at a zoomed in region of the SM data</span></span>
<span><span class="va">cropped.coords.MALDI</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/Crop.html" class="external-link">Crop</a></span><span class="op">(</span><span class="va">MALDI</span><span class="op">[[</span><span class="st">"fov"</span><span class="op">]</span><span class="op">]</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2000</span>, <span class="fl">3000</span><span class="op">)</span>, x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">8500</span>, <span class="fl">9500</span><span class="op">)</span>, coords <span class="op">=</span> <span class="st">"tissue"</span><span class="op">)</span></span>
<span><span class="va">MALDI</span><span class="op">[[</span><span class="st">"zoom"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">cropped.coords.MALDI</span></span></code></pre></div>
<p>Lets visualise this zoomed area now:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set the boundary as segmentation for our Xenium data</span></span>
<span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/Boundaries.html" class="external-link">DefaultBoundary</a></span><span class="op">(</span><span class="va">xenium</span><span class="op">[[</span><span class="st">"zoom"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"segmentation"</span></span>
<span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/ImageDimPlot.html" class="external-link">ImageDimPlot</a></span><span class="op">(</span><span class="va">xenium</span>, group.by <span class="op">=</span> <span class="st">"celltype"</span>, fov <span class="op">=</span> <span class="st">"zoom"</span>, size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">|</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ImageDimPlot.html" class="external-link">ImageDimPlot</a></span><span class="op">(</span><span class="va">MALDI</span>, group.by <span class="op">=</span> <span class="st">"clusters"</span>, fov <span class="op">=</span> <span class="st">"zoom"</span>, size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="Single_Cell_MultiOmics_files/figure-html/unnamed-chunk-8-1.png" width="1056" style="display: block; margin: auto;">
Looking at our zoomed in FOV the single cell resolution ST provides much
higher levels of detail compared to the SM data.</p>
</div>
<div class="section level2">
<h2 id="mapping-single-cell-resolution-st-and-sm">Mapping Single Cell Resolution ST and SM<a class="anchor" aria-label="anchor" href="#mapping-single-cell-resolution-st-and-sm"></a>
</h2>
<p>We want to eventually map our SM data to our single cell resolution
ST data, meaning that for each single cell we will have both
transcriptomic and metabolic information. First, we can check the
alignment of our datasets.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set the boundary as segmentation for our Xenium data</span></span>
<span><span class="fu"><a href="../reference/CheckAlignment.html">CheckAlignment</a></span><span class="op">(</span>ST.data <span class="op">=</span> <span class="va">xenium</span>, SM.data <span class="op">=</span> <span class="va">MALDI</span>, image.slice <span class="op">=</span> <span class="st">"fov"</span>,size <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="Single_Cell_MultiOmics_files/figure-html/unnamed-chunk-9-1.png" width="672" style="display: block; margin: auto;"></p>
<p>Based on this, we can see our datasets are aligned correctly. Next,
we will map our SM and Xenium ST data to generate a <em>SpaMTP</em>
<em>Seurat</em> object containing metabolite and transcriptomic
information per individual cell.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set the boundary as segmentation for our Xenium data</span></span>
<span><span class="va">MO_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MapSpatialOmics.html">MapSpatialOmics</a></span><span class="op">(</span>SM.data <span class="op">=</span> <span class="va">MALDI</span>, ST.data <span class="op">=</span> <span class="va">xenium</span>, ST.hires <span class="op">=</span> <span class="cn">TRUE</span>, SM.assay <span class="op">=</span> <span class="st">"Spatial"</span>,ST.assay <span class="op">=</span> <span class="st">"Xenium"</span>, SM.fov <span class="op">=</span> <span class="st">"fov"</span>, ST.image <span class="op">=</span> <span class="st">"centroids"</span><span class="op">)</span></span></code></pre></div>
<p>Our two datasets have been aligned to each individual cell. Lets
visualise the results:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MO_data</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 8 features across 15448 samples within 2 assays </span></span>
<span><span class="co">#&gt; Active assay: SPT (4 features, 0 variable features)</span></span>
<span><span class="co">#&gt;  1 layer present: counts</span></span>
<span><span class="co">#&gt;  1 other assay present: SPM</span></span>
<span><span class="co">#&gt;  3 spatial fields of view present: centroids segmentations zoom</span></span></code></pre></div>
<p>The multi-omic <em>SpaMTP</em> <em>Seurat</em> object now contains
two assays, ‘SPT’ and ‘SPM’ which stores the transcriptomic and
metabolomic data respectively.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">MO_data</span>, n <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p>This object also has the combined metadata from both original
objects. The metadata associated with our original SM object is marked
by “_SPM”.</p>
<table class="table dataframe table table-striped table-hover table-condensed table-responsive" style="width: auto !important; margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;font-weight: bold;">
</th>
<th style="text-align:center;font-weight: bold;">
orig.ident
</th>
<th style="text-align:center;font-weight: bold;">
nCount_Xenium
</th>
<th style="text-align:center;font-weight: bold;">
nFeature_Xenium
</th>
<th style="text-align:center;font-weight: bold;">
celltype
</th>
<th style="text-align:center;font-weight: bold;">
SPM_pixels
</th>
<th style="text-align:center;font-weight: bold;">
nCount_SPM
</th>
<th style="text-align:center;font-weight: bold;">
nFeature_SPM
</th>
<th style="text-align:center;font-weight: bold;">
orig.ident_SPM
</th>
<th style="text-align:center;font-weight: bold;">
nCount_Spatial_SPM
</th>
<th style="text-align:center;font-weight: bold;">
nFeature_Spatial_SPM
</th>
<th style="text-align:center;font-weight: bold;">
clusters_SPM
</th>
<th style="text-align:center;font-weight: bold;">
nCount_SPT
</th>
<th style="text-align:center;font-weight: bold;">
nFeature_SPT
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
acficmpm-1
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
SeuratProject
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
0
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
0
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
Astrocyte
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
9850_3370
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
80
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
1
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
SeuratProject
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
80
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
1
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
2
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
0
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
0
</td>
</tr>
<tr>
<td style="text-align:left;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
acfiiaog-1
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
SeuratProject
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
0
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
0
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
Astrocyte
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
9870_3390
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
0
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
0
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
SeuratProject
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
0
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
0
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
2
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
0
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
0
</td>
</tr>
<tr>
<td style="text-align:left;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
acfimhih-1
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
SeuratProject
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
0
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
0
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
Astrocyte
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
9870_3370
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
58
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
1
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
SeuratProject
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
58
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
1
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
2
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
0
</td>
<td style="text-align:center;border-bottom: 1px solid #ddd; border-top: 1px solid #ddd;">
0
</td>
</tr>
</tbody>
</table>
<p>Next, we can compare the data spatially:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set the boundary as segmentation for our Xenium data</span></span>
<span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/Boundaries.html" class="external-link">DefaultBoundary</a></span><span class="op">(</span><span class="va">MO_data</span><span class="op">[[</span><span class="st">"zoom"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"segmentation"</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ImageDimPlot.html" class="external-link">ImageDimPlot</a></span><span class="op">(</span><span class="va">MALDI</span>, group.by <span class="op">=</span> <span class="st">"clusters"</span>, fov <span class="op">=</span> <span class="st">"zoom"</span>, size <span class="op">=</span> <span class="fl">1.5</span>, dark.background <span class="op">=</span> <span class="cn">F</span><span class="op">)</span><span class="op">/</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ImageDimPlot.html" class="external-link">ImageDimPlot</a></span><span class="op">(</span><span class="va">MO_data</span>, group.by <span class="op">=</span> <span class="st">"clusters_SPM"</span>, fov <span class="op">=</span> <span class="st">"zoom"</span>, dark.background <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ImageFeaturePlot.html" class="external-link">ImageFeaturePlot</a></span><span class="op">(</span><span class="va">MALDI</span>, features <span class="op">=</span> <span class="st">"mz-100"</span>, fov <span class="op">=</span> <span class="st">"zoom"</span>, size <span class="op">=</span> <span class="fl">1.5</span>, dark.background <span class="op">=</span> <span class="cn">F</span><span class="op">)</span><span class="op">/</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ImageFeaturePlot.html" class="external-link">ImageFeaturePlot</a></span><span class="op">(</span><span class="va">MO_data</span>, features <span class="op">=</span> <span class="st">"mz-100"</span>, fov <span class="op">=</span> <span class="st">"zoom"</span>, dark.background <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span><span class="op">|</span><span class="va">p2</span></span></code></pre></div>
<p><img src="Single_Cell_MultiOmics_files/figure-html/unnamed-chunk-14-1.png" width="1056" style="display: block; margin: auto;"></p>
<p>Comparing the mapped data to the original SM data we can see that the
spatial pattern of both clusters and metabolite (‘mz-100’) match. Now
that our data is mapped to single cell resolution we have a much more
detailed visualisation of SM based clustering across our simulated
tissue sample.</p>
<p>SM data indicates the metabolic functioning of a cell, and based on
this we can identify cell types with different metabolic states:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">MO_data</span><span class="op">$</span><span class="va">clusters_SPM</span>, <span class="va">MO_data</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;     Astrocyte Microglia Oligodendrocyte Proliferation</span></span>
<span><span class="co">#&gt;   2      3551        63             596          3472</span></span>
<span><span class="co">#&gt;   3       232         2              26          2329</span></span>
<span><span class="co">#&gt;   1      2743        34            1677           560</span></span>
<span><span class="co">#&gt;   4         8         0               1            14</span></span>
<span><span class="co">#&gt;   0         2         0               1           137</span></span></code></pre></div>
<p>Looking at Oligodendrocytes for example, we can identify different
subtypes based on their metabolic activity.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MO_data</span><span class="op">$</span><span class="va">Oligo_states</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">MO_data</span><span class="op">$</span><span class="va">celltype</span> <span class="op">==</span> <span class="st">"Oligodendrocyte"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Oligo State: "</span>, <span class="va">MO_data</span><span class="op">$</span><span class="va">clusters_SPM</span><span class="op">)</span>,  <span class="st">"Other"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/ImageDimPlot.html" class="external-link">ImageDimPlot</a></span><span class="op">(</span><span class="va">MO_data</span>, group.by <span class="op">=</span> <span class="st">"Oligo_states"</span>, fov <span class="op">=</span> <span class="st">"zoom"</span>, dark.background <span class="op">=</span> <span class="cn">F</span>, cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"grey"</span>, <span class="st">"red"</span>, <span class="st">"blue"</span>, <span class="st">"yellow"</span>, <span class="st">"green"</span>, <span class="st">"pink"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Single_Cell_MultiOmics_files/figure-html/unnamed-chunk-16-1.png" width="1056" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DotPlot.html" class="external-link">DotPlot</a></span><span class="op">(</span><span class="va">MO_data</span>, group.by <span class="op">=</span> <span class="st">"Oligo_states"</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">MO_data</span><span class="op">[[</span><span class="st">"SPT"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, <span class="st">"spm_mz-200"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Single_Cell_MultiOmics_files/figure-html/unnamed-chunk-17-1.png" width="1920" style="display: block; margin: auto;">
This dotplot may suggest that Oligo State 1 and 2 are different cell
subtypes based on their differential expression of gene1/gene2 and
mz-200 (Note: this is synthetic data and has no biological context -
results only used as example).</p>
<p>An additional features of having multi-omic data mapped to the same
coordinates is that we can visualise the expression of both metabolites
and gene expression on a single plot.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/ImageFeaturePlot.html" class="external-link">ImageFeaturePlot</a></span><span class="op">(</span><span class="va">MO_data</span>, features <span class="op">=</span> <span class="st">"mz-200"</span>, molecules <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene1"</span>, <span class="st">"gene2"</span><span class="op">)</span>, fov <span class="op">=</span> <span class="st">"zoom"</span>, mols.cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"green"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Single_Cell_MultiOmics_files/figure-html/unnamed-chunk-18-1.png" width="1056" style="display: block; margin: auto;">
Analysing this spatial plot, there is a clear correlation between
‘mz-100’ and ‘gene1’ suggesting there interaction or similar
functionality.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">#&gt; Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">#&gt; Running under: macOS 15.5</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Europe/Dublin</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] future_1.67.0       ggplot2_4.0.0       dplyr_1.1.4        </span></span>
<span><span class="co">#&gt;  [4] Seurat_5.3.0        SeuratObject_5.2.0  sp_2.2-0           </span></span>
<span><span class="co">#&gt;  [7] Cardinal_3.8.3      S4Vectors_0.44.0    ProtGenerics_1.38.0</span></span>
<span><span class="co">#&gt; [10] BiocGenerics_0.52.0 BiocParallel_1.40.2 SpaMTP_1.1.0       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RColorBrewer_1.1-3     rstudioapi_0.17.1      jsonlite_2.0.0        </span></span>
<span><span class="co">#&gt;   [4] magrittr_2.0.4         spatstat.utils_3.2-0   farver_2.1.2          </span></span>
<span><span class="co">#&gt;   [7] rmarkdown_2.29         fs_1.6.6               ragg_1.5.0            </span></span>
<span><span class="co">#&gt;  [10] vctrs_0.6.5            ROCR_1.0-11            spatstat.explore_3.5-3</span></span>
<span><span class="co">#&gt;  [13] RCurl_1.98-1.17        htmltools_0.5.8.1      sass_0.4.10           </span></span>
<span><span class="co">#&gt;  [16] sctransform_0.4.2      parallelly_1.45.1      KernSmooth_2.23-26    </span></span>
<span><span class="co">#&gt;  [19] bslib_0.9.0            htmlwidgets_1.6.4      desc_1.4.3            </span></span>
<span><span class="co">#&gt;  [22] ica_1.0-3              plyr_1.8.9             plotly_4.11.0         </span></span>
<span><span class="co">#&gt;  [25] zoo_1.8-14             cachem_1.1.0           igraph_2.1.4          </span></span>
<span><span class="co">#&gt;  [28] mime_0.13              lifecycle_1.0.4        pkgconfig_2.0.3       </span></span>
<span><span class="co">#&gt;  [31] Matrix_1.7-4           R6_2.6.1               fastmap_1.2.0         </span></span>
<span><span class="co">#&gt;  [34] fitdistrplus_1.2-4     shiny_1.11.1           digest_0.6.37         </span></span>
<span><span class="co">#&gt;  [37] patchwork_1.3.2        tensor_1.5.1           RSpectra_0.16-2       </span></span>
<span><span class="co">#&gt;  [40] irlba_2.3.5.1          textshaping_1.0.3      labeling_0.4.3        </span></span>
<span><span class="co">#&gt;  [43] progressr_0.16.0       spatstat.sparse_3.1-0  httr_1.4.7            </span></span>
<span><span class="co">#&gt;  [46] polyclip_1.10-7        abind_1.4-8            compiler_4.4.1        </span></span>
<span><span class="co">#&gt;  [49] proxy_0.4-27           withr_3.0.2            S7_0.2.0              </span></span>
<span><span class="co">#&gt;  [52] tiff_0.1-12            DBI_1.2.3              fastDummies_1.7.5     </span></span>
<span><span class="co">#&gt;  [55] MASS_7.3-65            classInt_0.4-11        units_0.8-7           </span></span>
<span><span class="co">#&gt;  [58] tools_4.4.1            lmtest_0.9-40          httpuv_1.6.16         </span></span>
<span><span class="co">#&gt;  [61] future.apply_1.20.0    goftest_1.2-3          glue_1.8.0            </span></span>
<span><span class="co">#&gt;  [64] nlme_3.1-168           EBImage_4.48.0         promises_1.3.3        </span></span>
<span><span class="co">#&gt;  [67] sf_1.0-21              grid_4.4.1             Rtsne_0.17            </span></span>
<span><span class="co">#&gt;  [70] cluster_2.1.8.1        reshape2_1.4.4         generics_0.1.4        </span></span>
<span><span class="co">#&gt;  [73] gtable_0.3.6           spatstat.data_3.1-8    class_7.3-23          </span></span>
<span><span class="co">#&gt;  [76] tidyr_1.3.1            data.table_1.17.8      xml2_1.4.0            </span></span>
<span><span class="co">#&gt;  [79] spatstat.geom_3.6-0    RcppAnnoy_0.0.22       ggrepel_0.9.6         </span></span>
<span><span class="co">#&gt;  [82] RANN_2.6.2             pillar_1.11.1          stringr_1.5.2         </span></span>
<span><span class="co">#&gt;  [85] spam_2.11-1            RcppHNSW_0.6.0         limma_3.62.2          </span></span>
<span><span class="co">#&gt;  [88] later_1.4.4            splines_4.4.1          lattice_0.22-7        </span></span>
<span><span class="co">#&gt;  [91] survival_3.8-3         deldir_2.0-4           tidyselect_1.2.1      </span></span>
<span><span class="co">#&gt;  [94] CardinalIO_1.4.0       locfit_1.5-9.12        miniUI_0.1.2          </span></span>
<span><span class="co">#&gt;  [97] pbapply_1.7-4          knitr_1.50             gridExtra_2.3         </span></span>
<span><span class="co">#&gt; [100] matter_2.8.0           svglite_2.2.1          scattermore_1.2       </span></span>
<span><span class="co">#&gt; [103] xfun_0.53              Biobase_2.66.0         statmod_1.5.0         </span></span>
<span><span class="co">#&gt; [106] matrixStats_1.5.0      fftwtools_0.9-11       stringi_1.8.7         </span></span>
<span><span class="co">#&gt; [109] lazyeval_0.2.2         yaml_2.3.10            kableExtra_1.4.0      </span></span>
<span><span class="co">#&gt; [112] evaluate_1.0.5         codetools_0.2-20       tibble_3.3.0          </span></span>
<span><span class="co">#&gt; [115] cli_3.6.5              ontologyIndex_2.12     uwot_0.2.3            </span></span>
<span><span class="co">#&gt; [118] xtable_1.8-4           reticulate_1.43.0      systemfonts_1.2.3     </span></span>
<span><span class="co">#&gt; [121] jquerylib_0.1.4        Rcpp_1.1.0             globals_0.18.0        </span></span>
<span><span class="co">#&gt; [124] spatstat.random_3.4-2  zeallot_0.2.0          png_0.1-8             </span></span>
<span><span class="co">#&gt; [127] spatstat.univar_3.1-4  parallel_4.4.1         pkgdown_2.1.3         </span></span>
<span><span class="co">#&gt; [130] dotCall64_1.2          jpeg_0.1-11            bitops_1.0-9          </span></span>
<span><span class="co">#&gt; [133] listenv_0.9.1          viridisLite_0.4.2      e1071_1.7-16          </span></span>
<span><span class="co">#&gt; [136] scales_1.4.0           ggridges_0.5.7         purrr_1.1.0           </span></span>
<span><span class="co">#&gt; [139] rlang_1.1.6            cowplot_1.2.0          shinyjs_2.1.0</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/agc888" class="external-link">Andrew Causer</a>, <a href="https://github.com/BCRL-tylu" class="external-link">Tianyao Lu</a>, <a href="https://github.com/ChemCharles" class="external-link">Christopher C.J. Fitzgerald</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
