<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Getting Started • SpaMTP</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Getting Started">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SpaMTP</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/SpaMTP.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/Mouse_Urinary_Bladder.html">Spatial Metabolomics Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Multi-Omic_Mouse_Brain.html">Spatial Multi-Omics Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/SpaMTP_Additional_Features.html">Additional SpaMTP Features</a></li>
    <li><a class="dropdown-item" href="../articles/Single_Cell_MultiOmics.html">Simulated Single Cell Multi-Omics Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Big_Data.html">Handling Big Datasets</a></li>
    <li><a class="dropdown-item" href="../articles/Annotation_Refinement.html">Refinning Metabolite Annotations</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Updates</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/GenomicsMachineLearning/SpaMTP" aria-label="GitHub"><span class="fa fa-github"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Getting Started</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/genomicsmachinelearning/SpaMTP/blob/HEAD/vignettes/SpaMTP.Rmd" class="external-link"><code>vignettes/SpaMTP.Rmd</code></a></small>
      <div class="d-none name"><code>SpaMTP.Rmd</code></div>
    </div>

    
    
<style>
.two-column-htmltable {
  width: 66.666%;
  background-color: white;
}
.one-column-htmltable {
  width: 33.333%;
  background-color: white;
}

</style>
<p>Listed below are a number of useful tutorials demonstrating how to
use SpaMTP when analysing your spatial metabolomic datasets. For more
documentation of each function used in these tutorials, please visit our
<a href="../reference/index.html">Reference page</a>.</p>
<div class="section level2">
<h2 id="useful-spamtp-vignettes">Useful SpaMTP Vignettes<a class="anchor" aria-label="anchor" href="#useful-spamtp-vignettes"></a>
</h2>
<p>For a general introduction in importing data, annotating metabolites
and running key analysis pipelines on your spatial metabolomic data, we
suggest starting with the Spatial Metabolomic Analysis Tutorial. This
vignette uses mouse bladder data and demonstrates how to perform general
tasks such as plotting, subsetting, and manipulating <em>SpaMTP</em>
Seurat objects. For those with paired multi-modal datasets, we suggest
working through the Multi-Modal Integration Tutorial. Here, we provide
methods to align your two datasets to the same coordinates, map spatial
metabolomic pixels to spatial transcriptomic spots and then run various
analysis using this integrated data.</p>
<table class="table gmisc_table" style="border-collapse: collapse; margin-top: 1em; margin-bottom: 1em;"><tbody>
<tr style="border-top: 2px solid grey;">
<td style="padding-left: .75em; width: 30%; background-color: white; border-color: rgba(0,0,0,0); color: black; border-top: 2px solid grey; border-left: 1px solid black; border-right: 1px solid black; text-align: left;">
<div style="text-align:center; font-size:large; font-weight:bold">
<a href="Mouse_Urinary_Bladder.html">Spatial Metabolomics Analysis using
SpaMTP</a>
</div>
</td>
<td style="padding-left: .75em; width: 30%; background-color: white; border-color: rgba(0,0,0,0); color: black; border-top: 2px solid grey; border-right: 1px solid black; text-align: left;">
<div style="text-align:center; font-size:large; font-weight:bold">
<a href="Multi-Omic_Mouse_Brain.html">Multi-Modal Integration of Spatial
Data using SpaMTP</a>
</div>
</td>
<td style="padding-left: .75em; width: 30%; background-color: white; border-color: rgba(0,0,0,0); color: black; border-top: 2px solid grey; border-right: 1px solid black; text-align: left;">
<div style="text-align:center; font-size:large; font-weight:bold">
<a href="SpaMTP_Additional_Features.html">Additional SpaMTP
Functionality</a>
</div>
</td>
</tr>
<tr>
<td style="padding-left: .75em; width: 30%; background-color: white; border-color: rgba(0,0,0,0); color: black; border-left: 1px solid black; border-right: 1px solid black; text-align: left;">
<img src="assets/SM_Analysis.png" width="3000">
</td>
<td style="padding-left: .75em; width: 30%; background-color: white; border-color: rgba(0,0,0,0); color: black; border-right: 1px solid black; text-align: left;">
<img src="assets/mm_integration.png" width="3000">
</td>
<td style="padding-left: .75em; width: 30%; background-color: white; border-color: rgba(0,0,0,0); color: black; border-right: 1px solid black; text-align: left;">
<img src="assets/AdditionalFeatures.png" width="3000">
</td>
</tr>
<tr>
<td style="padding-left: .75em; width: 30%; background-color: white; border-color: rgba(0,0,0,0); color: black; border-left: 1px solid black; border-right: 1px solid black; text-align: left;">
An introduction to analysing spatial metabolomic datasets using SpaMTP.
</td>
<td style="padding-left: .75em; width: 30%; background-color: white; border-color: rgba(0,0,0,0); color: black; border-right: 1px solid black; text-align: left;">
Align, map and integrate spatial metabolomic and transcriptomics
datasets together using SpaMTP.
</td>
<td style="padding-left: .75em; width: 30%; background-color: white; border-color: rgba(0,0,0,0); color: black; border-right: 1px solid black; text-align: left;">
Examples of additional SpaMTP visualisation and analysis functions that
are useful for SM analysis pipelines.
</td>
</tr>
<tr>
<td style="padding-left: .75em; width: 30%; background-color: white; border-color: rgba(0,0,0,0); color: black; border-bottom: 2px solid grey; border-left: 1px solid black; border-right: 1px solid black; text-align: left;">
<a class="btn btn-primary" href="Mouse_Urinary_Bladder.html" role="button" style="width:100%">GO</a>
</td>
<td style="padding-left: .75em; width: 30%; background-color: white; border-color: rgba(0,0,0,0); color: black; border-bottom: 2px solid grey; border-right: 1px solid black; text-align: left;">
<a class="btn btn-primary" href="Multi-Omic_Mouse_Brain.html" role="button" style="width:100%">GO</a>
</td>
<td style="padding-left: .75em; width: 30%; background-color: white; border-color: rgba(0,0,0,0); color: black; border-bottom: 2px solid grey; border-right: 1px solid black; text-align: left;">
<a class="btn btn-primary" href="SpaMTP_Additional_Features.html" role="button" style="width:100%">GO</a>
</td>
</tr>
</tbody></table>
</div>
<div class="section level2">
<h2 id="spamtp-public-datasets">SpaMTP Public Datasets<a class="anchor" aria-label="anchor" href="#spamtp-public-datasets"></a>
</h2>
<p>To access and download all our datasets used in the vignettes you can
vist the <a href="https://zenodo.org/communities/spamtp/records" class="external-link"><em>SpaMTP</em>
zenodo page</a>.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the current version of SpaMTP from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"GenomicsMachineLearning/SpaMTP"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<div class="section level3">
<h3 id="installing-with-apptainer">Installing with <em>Apptainer</em><a class="anchor" aria-label="anchor" href="#installing-with-apptainer"></a>
</h3>
<p>To download the image:</p>
<pre class="console"><code>wget -c https://downloads.gmllab.com/SpaMTP/spamtp-1.1.sif</code></pre>
<p>Verify it download correctly:</p>
<pre class="console"><code>wget -q -O - https://downloads.gmllab.com/SpaMTP/spamtp-1.1.sif.md5 | md5sum -c -</code></pre>
<p>Or to create an Apptainer image from scratch (you need the
environment-linux.yml, dependencies.R and spamtp.def files):</p>
<pre class="console"><code>apptainer build spamtp.sif spamtp.def</code></pre>
</div>
<div class="section level3">
<h3 id="installing-with-conda-environment">Installing with <em>conda</em> environment<a class="anchor" aria-label="anchor" href="#installing-with-conda-environment"></a>
</h3>
<p>Below is an example of how to set up SpaMTP using a conda
environment:</p>
<pre class="console"><code>conda create -n SpaMTP -c conda-forge r-base=4.3.3 r-essentials r-devtools r-cairo python=3.9 r-reticulate r-leiden r-imager r-magick r-hdf5r r-sf bioconda::bioconductor-ebimage bioconda::bioconductor-scater bioconda::bioconductor-dropletutils bioconda::bioconductor-fgsea bioconda::bioconductor-rgoslin bioconda::bioconductor-glmgampoi
conda activate SpaMTP</code></pre>
<p>Within the SpaMTP environment open <em>R</em> and install
<strong><em>SpaMTP</em></strong>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Install Seurat version 5.3.0</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"satijalab/seurat"</span>, <span class="st">"seurat5"</span>, quiet <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Install Cardinal</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"Cardinal"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Install SpaMTP</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"GenomicsMachineLearning/SpaMTP"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level3">
<h3 id="possible-installiation-errors">Possible Installiation Errors<a class="anchor" aria-label="anchor" href="#possible-installiation-errors"></a>
</h3>
<div class="section level4">
<h4 id="cairo-failed-to-install">Cairo Failed to Install<a class="anchor" aria-label="anchor" href="#cairo-failed-to-install"></a>
</h4>
<pre class="console"><code>checking if R was compiled with the RConn patch... no
checking for cairo.h... no
configure: error: Cannot find cairo.h! Please install cairo (http://www.cairographics.org/) and/or set CAIRO_CFLAGS/LIBS correspondingly.
ERROR: configuration failed for package ‘Cairo’</code></pre>
<p>Check these libraries exist:</p>
<pre class="console"><code>conda install conda-forge::r-cairo</code></pre>
<p>This should resolve any issues. Rerun:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"scater"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="cardinal-failed-to-install">Cardinal Failed to Install<a class="anchor" aria-label="anchor" href="#cardinal-failed-to-install"></a>
</h4>
<pre class="console"><code>ERROR: dependency ‘EBImage’ is not available for package ‘Cardinal’

BiocManager::install("EBImage")

fftwtools.c:28:9: fatal error: fftw3.h: No such file or directory
   28 | #include&lt;fftw3.h&gt;
      |         ^~~~~~~~~</code></pre>
<p>If EBImage failed to installed, it is most likely due to an issue
with the installation of <em>fftwtools</em>.</p>
<p>Try:</p>
<pre class="console"><code>conda config --add channels conda-forge
conda config --set channel_priority strict

conda install r-fftwtools</code></pre>
<p>Else try:</p>
<pre class="console"><code>conda install bioconda::r-fftwtools
conda install conda-forge::fftw</code></pre>
<p>Try in <em>R</em>:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="co">#Check if BiocManager is installed</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"fftwtools"</span><span class="op">)</span></span></code></pre></div>
<p>If none of the above methods resolve instiallation issues, install
<em>Cardinal</em> directly through <em>conda</em>:</p>
<pre class="console"><code>conda install bioconda::bioconductor-cardinal  </code></pre>
</div>
<div class="section level4">
<h4 id="rgoslin-failed-to-install">
<em>rgoslin</em> Failed to Install<a class="anchor" aria-label="anchor" href="#rgoslin-failed-to-install"></a>
</h4>
<p>If rgoslin fails to install please head to their <a href="https://github.com/lifs-tools/rgoslin" class="external-link">github</a>. Alternative,
you can try installing through BiocManager or mamba shown below:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"rgoslin"</span><span class="op">)</span></span></code></pre></div>
<p>or</p>
<pre class="console"><code>mamba install bioconductor-rgoslin</code></pre>
</div>
<div class="section level4">
<h4 id="matrix-version-incompatible-with-seurat">
<em>Matrix</em> version incompatible with <em>Seurat</em><a class="anchor" aria-label="anchor" href="#matrix-version-incompatible-with-seurat"></a>
</h4>
<p>If the <em>Matrix</em> package version installed is not compatible
with <em>Seurat</em> then try installing a different version and
re-installing <em>Seurat</em></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Install Correct version of Matrix package</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_version.html" class="external-link">install_version</a></span><span class="op">(</span><span class="st">"Matrix"</span>, version <span class="op">=</span> <span class="st">"1.6-4"</span>, repos <span class="op">=</span> <span class="st">"https://cran.r-project.org"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Install Seurat version 5.1.0</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"Seruat"</span><span class="op">)</span></span></code></pre></div>
<p>For other issue please flag on github under <em>Issues</em>.</p>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/agc888" class="external-link">Andrew Causer</a>, <a href="https://github.com/BCRL-tylu" class="external-link">Tianyao Lu</a>, <a href="https://github.com/ChemCharles" class="external-link">Christopher C.J. Fitzgerald</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
