<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>SpaMTP: Additional Analysis Features • SpaMTP</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="SpaMTP: Additional Analysis Features">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SpaMTP</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/SpaMTP.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/Mouse_Urinary_Bladder.html">Spatial Metabolomics Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Multi-Omic_Mouse_Brain.html">Spatial Multi-Omics Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/SpaMTP_Additional_Features.html">Additional SpaMTP Features</a></li>
    <li><a class="dropdown-item" href="../articles/Single_Cell_MultiOmics.html">Simulated Single Cell Multi-Omics Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Big_Data.html">Handling Big Datasets</a></li>
    <li><a class="dropdown-item" href="../articles/Annotation_Refinement.html">Refinning Metabolite Annotations</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Updates</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/GenomicsMachineLearning/SpaMTP" aria-label="GitHub"><span class="fa fa-github"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>SpaMTP: Additional Analysis Features</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/genomicsmachinelearning/SpaMTP/blob/HEAD/vignettes/SpaMTP_Additional_Features.Rmd" class="external-link"><code>vignettes/SpaMTP_Additional_Features.Rmd</code></a></small>
      <div class="d-none name"><code>SpaMTP_Additional_Features.Rmd</code></div>
    </div>

    
    
<p><br></p>
<p>This vignette will highlights some additional features of <a href="https://github.com/GenomicsMachineLearning/SpaMTP" class="external-link"><strong><em>SpaMTP</em></strong></a>
that can be useful for analysing and visualising your SM data. This
vignette will again use the public mouse urinary bladder dataset
analysed extensively <a href="https://genomicsmachinelearning.github.io/SpaMTP/articles/Mouse_Urinary_Bladder.html">here</a>.</p>
<p><em>Author: Andrew Causer</em></p>
<div class="sourceCode" id="section-cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Install SpaMTP if not previously installed</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/genomicsmachinelearning/SpaMTP" class="external-link">"SpaMTP"</a></span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"GenomicsMachineLearning/SpaMTP"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#General Libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/genomicsmachinelearning/SpaMTP" class="external-link">SpaMTP</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://www.cardinalmsi.org" class="external-link">Cardinal</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">#For plotting + DE plots</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/kevinblighe/EnhancedVolcano" class="external-link">EnhancedVolcano</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level5">
<h5 id="section-load-processed-data">Load Processed Data<a class="anchor" aria-label="anchor" href="#section-load-processed-data"></a>
</h5>
<p>Here, we will be using already processed, clustered (via
spatially-aware Shrunken Centroid) and annotated data. For more
information about how the data was processed, please visit <a href="https://genomicsmachinelearning.github.io/SpaMTP/articles/Mouse_Urinary_Bladder.html">here</a>.</p>
<div class="sourceCode" id="section-cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bladder_annotated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://zenodo.org/records/17246684/files/bladder_annotated.RDS?download=1"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">bladder_annotated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/UpdateSeuratObject.html" class="external-link">UpdateSeuratObject</a></span><span class="op">(</span><span class="va">bladder_annotated</span><span class="op">)</span> <span class="co"># Only Required if SeuratObjects =&gt; 5.0.2</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="section-querying-an-spamtp-dataset">Querying an <strong><em>SpaMTP</em></strong> Dataset<a class="anchor" aria-label="anchor" href="#section-querying-an-spamtp-dataset"></a>
</h2>
<p><strong><em>SpaMTP</em></strong> stores the relative m/z value
annotations in the objects feature metadata slot. We can see this
below:</p>
<div class="sourceCode" id="section-cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">bladder_annotated</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">Spatial</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span> </span></code></pre></div>
<table class="table dataframe">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:center;">
raw_mz
</th>
<th style="text-align:center;">
mz_names
</th>
<th style="text-align:center;">
observed_mz
</th>
<th style="text-align:center;">
all_IsomerNames
</th>
<th style="text-align:center;">
all_Isomers
</th>
<th style="text-align:center;">
all_Adducts
</th>
<th style="text-align:center;">
all_Formulas
</th>
<th style="text-align:center;">
all_Errors
</th>
<th style="text-align:center;">
Lipid.Maps.Category
</th>
<th style="text-align:center;">
Lipid.Maps.Main.Class
</th>
<th style="text-align:center;">
Species.Name
</th>
<th style="text-align:center;">
Species.Name.Simple
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
16
</td>
<td style="text-align:center;">
426.9289
</td>
<td style="text-align:center;">
mz-426.928863525391
</td>
<td style="text-align:center;">
426.928863525391
</td>
<td style="text-align:center;">
Laurenenyne A; Laurenenyne B; Katsuurenyne A
</td>
<td style="text-align:center;">
LMPK02000053; LMPK02000054; LMPK02000055
</td>
<td style="text-align:center;">
M+K
</td>
<td style="text-align:center;">
C15H18O2Br2
</td>
<td style="text-align:center;">
3.8565
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
NA
</td>
</tr>
<tr>
<td style="text-align:left;">
17
</td>
<td style="text-align:center;">
431.0381
</td>
<td style="text-align:center;">
mz-431.038116455078
</td>
<td style="text-align:center;">
431.038116455078
</td>
<td style="text-align:center;">
5,7,3’,4’,5’-Pentahydroxy-3,6,8-trimethoxyflavone
</td>
<td style="text-align:center;">
LMPK12113357
</td>
<td style="text-align:center;">
M+K
</td>
<td style="text-align:center;">
C18H16O10
</td>
<td style="text-align:center;">
1.4116
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
NA
</td>
</tr>
<tr>
<td style="text-align:left;">
37
</td>
<td style="text-align:center;">
465.0228
</td>
<td style="text-align:center;">
mz-465.022796630859
</td>
<td style="text-align:center;">
465.022796630859
</td>
<td style="text-align:center;">
Aspergillusether D
</td>
<td style="text-align:center;">
LMPK13090053
</td>
<td style="text-align:center;">
M+K
</td>
<td style="text-align:center;">
C20H20O6Cl2
</td>
<td style="text-align:center;">
8.725
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
NA
</td>
</tr>
<tr>
<td style="text-align:left;">
39
</td>
<td style="text-align:center;">
467.1028
</td>
<td style="text-align:center;">
mz-467.102752685547
</td>
<td style="text-align:center;">
467.102752685547
</td>
<td style="text-align:center;">
Emeguisin B
</td>
<td style="text-align:center;">
LMPK13080006
</td>
<td style="text-align:center;">
M+K
</td>
<td style="text-align:center;">
C24H25O5Cl
</td>
<td style="text-align:center;">
1.1597
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
NA
</td>
</tr>
<tr>
<td style="text-align:left;">
40
</td>
<td style="text-align:center;">
468.1027
</td>
<td style="text-align:center;">
mz-468.102722167969
</td>
<td style="text-align:center;">
468.102722167969
</td>
<td style="text-align:center;">
14-carboxy-15,16,17,18,19,20-hexanor-N-acetyl-leukotriene E3
</td>
<td style="text-align:center;">
LMFA03020084
</td>
<td style="text-align:center;">
M+K
</td>
<td style="text-align:center;">
C19H27NO8S
</td>
<td style="text-align:center;">
13.1953
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
NA
</td>
</tr>
<tr>
<td style="text-align:left;">
41
</td>
<td style="text-align:center;">
469.1054
</td>
<td style="text-align:center;">
mz-469.105377197266
</td>
<td style="text-align:center;">
469.105377197266
</td>
<td style="text-align:center;">
6-iodo-5-hydroxy-eicosa-8Z,11Z,14Z-trienoic acid gamma lactone
</td>
<td style="text-align:center;">
LMFA03000015
</td>
<td style="text-align:center;">
M+K
</td>
<td style="text-align:center;">
C20H31O2I
</td>
<td style="text-align:center;">
11.3775
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
NA
</td>
</tr>
</tbody>
</table>
<p>However, there is are some useful functions that lets the user search
through their <strong>SpaMTP</strong> dataset to find either exact m/z
values or metabolite names that they are interested in. Lets see how
each works:</p>
<p>The first is <em>FindNearestMZ()</em>, this function lets users find
the m/z value in their dataset that is closest to the value being
queried.</p>
<div class="sourceCode" id="section-cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/FindNearestMZ.html">FindNearestMZ</a></span><span class="op">(</span><span class="va">bladder_annotated</span>, <span class="fl">741.5</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "mz-741.530578613281"</span></span></code></pre>
<p>This is useful if the metabolite of interest has a known m/z mass, or
if the user wants to use <em>Seurat’s</em> feature plotting
functions.</p>
<p><br></p>
<p>The second is <em>SearchAnnotations()</em>, this function lets users
query a metabolite to find if it is present within their dataset.</p>
<div class="sourceCode" id="section-cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/SearchAnnotations.html">SearchAnnotations</a></span><span class="op">(</span><span class="va">bladder_annotated</span>, metabolite <span class="op">=</span> <span class="st">"Isorhamnetin 3-(6''-malonylglucoside)"</span>, search.exact <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<table class="table dataframe">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:center;">
raw_mz
</th>
<th style="text-align:center;">
mz_names
</th>
<th style="text-align:center;">
observed_mz
</th>
<th style="text-align:center;">
all_IsomerNames
</th>
<th style="text-align:center;">
all_Isomers
</th>
<th style="text-align:center;">
all_Adducts
</th>
<th style="text-align:center;">
all_Formulas
</th>
<th style="text-align:center;">
all_Errors
</th>
<th style="text-align:center;">
Lipid.Maps.Category
</th>
<th style="text-align:center;">
Lipid.Maps.Main.Class
</th>
<th style="text-align:center;">
Species.Name
</th>
<th style="text-align:center;">
Species.Name.Simple
</th>
</tr></thead>
<tbody><tr>
<td style="text-align:left;">
130
</td>
<td style="text-align:center;">
603.0773
</td>
<td style="text-align:center;">
mz-603.077331542969
</td>
<td style="text-align:center;">
603.077331542969
</td>
<td style="text-align:center;">
Isorhamnetin 3-(6’‘-malonylglucoside); Larycitrin
3-(4’’-malonylrhamnoside)
</td>
<td style="text-align:center;">
LMPK12110589; LMPK12112481
</td>
<td style="text-align:center;">
M+K
</td>
<td style="text-align:center;">
C25H24O15
</td>
<td style="text-align:center;">
4.3917
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
NA
</td>
</tr></tbody>
</table>
<p>The <code>search.exact</code> parameter specifies whether there needs
to be an exact match to the metabolite name. If set to
<code>FALSE</code>, users can find all metabolites with similar names.
For example, we want to see all the different types of
<code>(3'-sulfo)Galbeta-Cer</code> lipids in our dataset.</p>
<div class="sourceCode" id="section-cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/SearchAnnotations.html">SearchAnnotations</a></span><span class="op">(</span><span class="va">bladder_annotated</span>, metabolite <span class="op">=</span> <span class="st">"(3'-sulfo)Galbeta-Cer"</span>, search.exact <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<table class="table dataframe">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:center;">
raw_mz
</th>
<th style="text-align:center;">
mz_names
</th>
<th style="text-align:center;">
observed_mz
</th>
<th style="text-align:center;">
all_IsomerNames
</th>
<th style="text-align:center;">
all_Isomers
</th>
<th style="text-align:center;">
all_Adducts
</th>
<th style="text-align:center;">
all_Formulas
</th>
<th style="text-align:center;">
all_Errors
</th>
<th style="text-align:center;">
Lipid.Maps.Category
</th>
<th style="text-align:center;">
Lipid.Maps.Main.Class
</th>
<th style="text-align:center;">
Species.Name
</th>
<th style="text-align:center;">
Species.Name.Simple
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
287
</td>
<td style="text-align:center;">
918.5832
</td>
<td style="text-align:center;">
mz-918.583190917969
</td>
<td style="text-align:center;">
918.583190917969
</td>
<td style="text-align:center;">
(3’-sulfo)Galbeta-Cer(d18:1/22:0(2OH))
</td>
<td style="text-align:center;">
LMSP06020011
</td>
<td style="text-align:center;">
M+K
</td>
<td style="text-align:center;">
C46H89NO12S
</td>
<td style="text-align:center;">
10.3224
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
NA
</td>
</tr>
<tr>
<td style="text-align:left;">
295
</td>
<td style="text-align:center;">
946.6147
</td>
<td style="text-align:center;">
mz-946.61474609375
</td>
<td style="text-align:center;">
946.61474609375
</td>
<td style="text-align:center;">
(3’-sulfo)Galbeta-Cer(d18:1/24:0(2OH))
</td>
<td style="text-align:center;">
LMSP06020014
</td>
<td style="text-align:center;">
M+K
</td>
<td style="text-align:center;">
C48H93NO12S
</td>
<td style="text-align:center;">
10.2863
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
NA
</td>
</tr>
</tbody>
</table>
<p>This function also lets users select the feature metadata column they
want to query. For example, if <em>RefineLipis()</em> has been run,
users can search for all lipids of the Class “GlcCer”.</p>
<div class="sourceCode" id="section-cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/SearchAnnotations.html">SearchAnnotations</a></span><span class="op">(</span><span class="va">bladder_annotated</span>, metabolite <span class="op">=</span> <span class="st">"GlcCer"</span>, search.exact <span class="op">=</span> <span class="cn">FALSE</span>, column.name <span class="op">=</span> <span class="st">"Lipid.Maps.Main.Class"</span><span class="op">)</span></span></code></pre></div>
<table class="table dataframe">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:center;">
raw_mz
</th>
<th style="text-align:center;">
mz_names
</th>
<th style="text-align:center;">
observed_mz
</th>
<th style="text-align:center;">
all_IsomerNames
</th>
<th style="text-align:center;">
all_Isomers
</th>
<th style="text-align:center;">
all_Adducts
</th>
<th style="text-align:center;">
all_Formulas
</th>
<th style="text-align:center;">
all_Errors
</th>
<th style="text-align:center;">
Lipid.Maps.Category
</th>
<th style="text-align:center;">
Lipid.Maps.Main.Class
</th>
<th style="text-align:center;">
Species.Name
</th>
<th style="text-align:center;">
Species.Name.Simple
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
168
</td>
<td style="text-align:center;">
682.4571
</td>
<td style="text-align:center;">
mz-682.457092285156
</td>
<td style="text-align:center;">
682.457092285156
</td>
<td style="text-align:center;">
GlcCer(d18:1/12:0); GlcCer(d14:1/16:0)
</td>
<td style="text-align:center;">
LMSP0501AA01; LMSP0501AA39
</td>
<td style="text-align:center;">
M+K
</td>
<td style="text-align:center;">
C36H69NO8
</td>
<td style="text-align:center;">
12.2859
</td>
<td style="text-align:center;">
SP
</td>
<td style="text-align:center;">
GlcCer
</td>
<td style="text-align:center;">
GlcCer 30:1;O2
</td>
<td style="text-align:center;">
GlcCer(30:1)
</td>
</tr>
<tr>
<td style="text-align:left;">
236
</td>
<td style="text-align:center;">
808.5806
</td>
<td style="text-align:center;">
mz-808.580627441406
</td>
<td style="text-align:center;">
808.580627441406
</td>
<td style="text-align:center;">
GlcCer(d18:2(4E,8Z)/20:0(2OH[R])); GlcCer(d18:2(4E,8E)/20:0(2OH[R]));
GlcCer(d14:1(4E)/24:1(15Z)(2OH)); GlcCer(d14:2(4E,6E)/24:0(2OH));
GlcCer(d16:2(4E,6E)/22:0(2OH))
</td>
<td style="text-align:center;">
LMSP05010052; LMSP05010065; LMSP0501AA69; LMSP0501AA73; LMSP0501AA80
</td>
<td style="text-align:center;">
M+K
</td>
<td style="text-align:center;">
C44H83NO9
</td>
<td style="text-align:center;">
13.2152
</td>
<td style="text-align:center;">
SP
</td>
<td style="text-align:center;">
GlcCer
</td>
<td style="text-align:center;">
GlcCer 38:2;O3
</td>
<td style="text-align:center;">
GlcCer(38:2)
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="section-subsetting-a-spamtp-dataset-by-metabolites">Subsetting a <strong><em>SpaMTP</em></strong> Dataset by
Metabolites<a class="anchor" aria-label="anchor" href="#section-subsetting-a-spamtp-dataset-by-metabolites"></a>
</h2>
<p>In some cases, we may want to reduce the number of features present
in our dataset. Often there are thousands or millions of metabolites in
one dataset, removing unwanted metabolites can often reduce the
computational load. Below, we demonstrate how to subset our bladder
dataset to only include Glycerophospholipids.</p>
<div class="sourceCode" id="section-cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GPs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SearchAnnotations.html">SearchAnnotations</a></span><span class="op">(</span><span class="va">bladder_annotated</span>, metabolite <span class="op">=</span> <span class="st">"GP"</span>, search.exact <span class="op">=</span> <span class="cn">FALSE</span>, column.name <span class="op">=</span> <span class="st">"Lipid.Maps.Category"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">bladder_GPs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SubsetMZFeatures.html">SubsetMZFeatures</a></span><span class="op">(</span><span class="va">bladder_annotated</span>, features <span class="op">=</span> <span class="va">GPs</span><span class="op">$</span><span class="va">mz_names</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="section-cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bladder_annotated</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 79 features across 34840 samples within 1 assay </span></span>
<span><span class="co">## Active assay: Spatial (79 features, 0 variable features)</span></span>
<span><span class="co">##  1 layer present: counts</span></span>
<span><span class="co">##  1 spatial field of view present: fov</span></span></code></pre>
<div class="sourceCode" id="section-cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bladder_GPs</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 33 features across 34840 samples within 1 assay </span></span>
<span><span class="co">## Active assay: Spatial (33 features, 0 variable features)</span></span>
<span><span class="co">##  1 layer present: counts</span></span>
<span><span class="co">##  1 spatial field of view present: fov</span></span></code></pre>
<p>We now only have 33 features, compared to the 79 in the original
dataset.</p>
</div>
<div class="section level2">
<h2 id="section-visualising-sm-data-using-spamtp">Visualising SM data using <strong><em>SpaMTP</em></strong><a class="anchor" aria-label="anchor" href="#section-visualising-sm-data-using-spamtp"></a>
</h2>
<p>An important part of any analysis pipeline is to generate informative
and clear visualisations of the results.
<strong><em>SpaMTP</em></strong> has a range of visualisation functions
and features, which we will step through below.</p>
<div class="section level3">
<h3 id="section-plotting-a-mz-value">Plotting a m/z Value<a class="anchor" aria-label="anchor" href="#section-plotting-a-mz-value"></a>
</h3>
<p>Plotting in <strong><em>SpaMTP</em></strong> builds on
<em>ggplot2</em> allowing for easy organisation and plot manipulation.
Below, we will plot the expression of the m/z 741.530578613281 using
<em>Seurat’s</em> plotting function <em>ImageFeaturePlot()</em>. Below
we use ggplot helper functions to orientate the image correctly.</p>
<div class="sourceCode" id="section-cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/ImageFeaturePlot.html" class="external-link">ImageFeaturePlot</a></span><span class="op">(</span><span class="va">bladder_annotated</span>, features <span class="op">=</span> <span class="st">"mz-741.530578613281"</span>, size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_reverse</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="SpaMTP_Additional_Features_files/figure-html/unnamed-chunk-15-1.png" width="792" style="display: block; margin: auto;"></p>
<p>Comparatively, the <strong><em>SpaMTP</em></strong> function
<em>ImageMZPlot()</em> provides more SM-based functionality,
including:</p>
<ul>
<li>Plotting the closest m/z value to the numeric value provided.</li>
<li>Plotting the combined expression of ± the mz value in ppm</li>
<li>Plotting the spots as either circles (<em>ST</em>) or pixels
(<em>SM</em>)</li>
</ul>
<div class="sourceCode" id="section-cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ImageMZPlot.html">ImageMZPlot</a></span><span class="op">(</span><span class="va">bladder_annotated</span>, mz <span class="op">=</span> <span class="fl">741.53057</span>, plusminus <span class="op">=</span> <span class="fl">1</span>, size <span class="op">=</span> <span class="fl">1</span>, plot.pixel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_reverse</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="SpaMTP_Additional_Features_files/figure-html/unnamed-chunk-16-1.png" width="80%" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="section-plotting-a-metabolite-name">Plotting a Metabolite Name<a class="anchor" aria-label="anchor" href="#section-plotting-a-metabolite-name"></a>
</h3>
<p>In addition, <strong><em>SpaMTP</em></strong> also has the function
<em>ImageMZAnnotationPlot()</em>, which allows users to plot the spatial
expression of named metabolites. In addition to the features above, this
function also lets users:</p>
<ul>
<li>Plot from a specific column in the feature metadata containing the
m/z metabolite annotations</li>
<li>Combine the expression of all metabolites that match the specified
name (as described above using plot.exact = <code>FALSE</code>).</li>
</ul>
<p>We will use the “Galbeta-Cer” metabolites as an example:</p>
<div class="sourceCode" id="section-cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mb</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="../reference/SearchAnnotations.html">SearchAnnotations</a></span><span class="op">(</span><span class="va">bladder_annotated</span>, metabolite <span class="op">=</span> <span class="st">"Galbeta-Cer"</span>, search.exact <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">all_IsomerNames</span></span>
<span></span>
<span><span class="va">mb</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "(3'-sulfo)Galbeta-Cer(d18:1/22:0(2OH))"</span></span>
<span><span class="co">## [2] "(3'-sulfo)Galbeta-Cer(d18:1/24:0(2OH))"</span></span></code></pre>
<p>Lets plot them individually:</p>
<div class="sourceCode" id="section-cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ImageMZAnnotationPlot.html">ImageMZAnnotationPlot</a></span><span class="op">(</span><span class="va">bladder_annotated</span>, metabolites <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"(3'-sulfo)Galbeta-Cer(d18:1/22:0(2OH))"</span>, <span class="st">"(3'-sulfo)Galbeta-Cer(d18:1/24:0(2OH))"</span><span class="op">)</span>, slot <span class="op">=</span> <span class="st">"counts"</span>, plot.exact <span class="op">=</span> <span class="cn">FALSE</span>, size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_reverse</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="SpaMTP_Additional_Features_files/figure-html/unnamed-chunk-18-1.png" width="2112" style="display: block; margin: auto;"></p>
<p>Lets plot them combined:</p>
<div class="sourceCode" id="section-cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ImageMZAnnotationPlot.html">ImageMZAnnotationPlot</a></span><span class="op">(</span><span class="va">bladder_annotated</span>, metabolites <span class="op">=</span> <span class="st">"Galbeta-Cer"</span>, slot <span class="op">=</span> <span class="st">"counts"</span>, plot.exact <span class="op">=</span> <span class="cn">FALSE</span>, size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&amp;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_reverse</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="SpaMTP_Additional_Features_files/figure-html/unnamed-chunk-19-1.png" width="80%" style="display: block; margin: auto;"></p>
<p>There are equivalent functions for <strong><em>SpaMTP</em></strong>
<em>Seurat</em> objects that contain images these being
<em>SpatialMZPlot()</em> and <em>SpatialMZAnnotationPlot()</em>.
Examples of these are shown below.</p>
</div>
<div class="section level3">
<h3 id="section-plotting-mass-intensity-spectra-plots">Plotting Mass Intensity Spectra Plots<a class="anchor" aria-label="anchor" href="#section-plotting-mass-intensity-spectra-plots"></a>
</h3>
<p>Often we are interested in different patterns of metabolite
expression between groups such as clusters. Mass intensity plots can be
useful at displaying the overall patters of each tissue region.
<strong><em>SpaMTP</em></strong> provides functionality to generate
these plots and also label peaks with selected metabolites.</p>
<div class="sourceCode" id="section-cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ROI</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">bladder_annotated</span>, subset <span class="op">=</span> <span class="va">ssc</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"2"</span>, <span class="st">"5"</span>, <span class="st">"6"</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="fu"><a href="../reference/MassIntensityPlot.html">MassIntensityPlot</a></span><span class="op">(</span><span class="va">ROI</span>, group.by <span class="op">=</span> <span class="st">"ssc"</span>, label.annotations <span class="op">=</span> <span class="cn">TRUE</span>, metabolite.labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PC(34:1)"</span><span class="op">)</span>, annotation.column <span class="op">=</span> <span class="st">"Species.Name.Simple"</span>,labelAdj <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  labelOffset <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="SpaMTP_Additional_Features_files/figure-html/unnamed-chunk-20-1.png" width="1920" style="display: block; margin: auto;"></p>
<p>Based on this plot, we can see that our metabolite “PC(34:1)” has
been annotated on the plot. The plot looks quite crowded though. We can
either change the mass range to only show the specific region on the
spectrum, or we could generate individual plots for each cluster. We
will show both approaches below:</p>
<p>Change the mass range:</p>
<div class="sourceCode" id="section-cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/MassIntensityPlot.html">MassIntensityPlot</a></span><span class="op">(</span><span class="va">ROI</span>, group.by <span class="op">=</span> <span class="st">"ssc"</span>, mass.range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">750</span>, <span class="fl">850</span><span class="op">)</span>, label.annotations <span class="op">=</span> <span class="cn">TRUE</span>, metabolite.labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PC(34:1)"</span><span class="op">)</span>, annotation.column <span class="op">=</span> <span class="st">"Species.Name.Simple"</span>,labelAdj <span class="op">=</span> <span class="fl">0</span>, labelOffset <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="SpaMTP_Additional_Features_files/figure-html/unnamed-chunk-21-1.png" width="1920" style="display: block; margin: auto;"></p>
<p>Generate individual plots per cluster:</p>
<div class="sourceCode" id="section-cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/MassIntensityPlot.html">MassIntensityPlot</a></span><span class="op">(</span><span class="va">ROI</span>, split.by  <span class="op">=</span> <span class="st">"ssc"</span>, label.annotations <span class="op">=</span> <span class="cn">TRUE</span>, metabolite.labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PC(34:1)"</span><span class="op">)</span>, annotation.column <span class="op">=</span> <span class="st">"Species.Name.Simple"</span>,labelAdj <span class="op">=</span> <span class="fl">2</span>, labelOffset <span class="op">=</span> <span class="fl">0</span>,labelCex <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="SpaMTP_Additional_Features_files/figure-html/unnamed-chunk-22-1.png" width="1920" style="display: block; margin: auto;"></p>
<p>Now we can clearly see that this metabolite is highly expressed by
cluster 6!</p>
<p>There are various additional visualisation techniques such as 3D
Density Kernel plotting and various DE and pathway analysis plotting
functions. These can all be seen in either the <a href="https://genomicsmachinelearning.github.io/SpaMTP/articles/Mouse_Urinary_Bladder.html">SM
data analysis</a> or <a href="https://genomicsmachinelearning.github.io/SpaMTP/articles/Multi-Omic_Mouse_Brain.html">Multi-Omic
data</a> analysis vignettes.</p>
</div>
</div>
<div class="section level2">
<h2 id="section-binning-metabolites">Binning Metabolites<a class="anchor" aria-label="anchor" href="#section-binning-metabolites"></a>
</h2>
<p>Expression of specific metabolites on their own can sometime be
difficult to detect, however when combining the expression of similar
metabolites, these patterns can become more observable.
<strong><em>SpaMTP</em></strong> has the function
<em>BinMetabolites()</em>, which lets users specify the metabolites they
want to combine the expression of. These values are then stored in the
<code>@meta.data</code> slot of the <strong><em>SpaMTP</em></strong>
<em>Seurat</em> object and can be plotted or used for further analysis.
Below, we will demonstrate how to bin the expression of all
Glycerophospholipids.</p>
<div class="sourceCode" id="section-cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mzs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SearchAnnotations.html">SearchAnnotations</a></span><span class="op">(</span><span class="va">bladder_annotated</span>, metabolite <span class="op">=</span> <span class="st">"GP"</span>, search.exact <span class="op">=</span> <span class="cn">FALSE</span>, column.name <span class="op">=</span> <span class="st">"Lipid.Maps.Category"</span><span class="op">)</span><span class="op">$</span><span class="va">mz_names</span></span>
<span></span>
<span><span class="va">bladder_annotated</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/BinMetabolites.html">BinMetabolites</a></span><span class="op">(</span><span class="va">bladder_annotated</span>, mzs <span class="op">=</span> <span class="va">mzs</span>, assay <span class="op">=</span> <span class="st">"Spatial"</span>, slot <span class="op">=</span> <span class="st">"counts"</span>, bin_name <span class="op">=</span> <span class="st">"Binned_GPs"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">bladder_annotated</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span></code></pre></div>
<table class="table dataframe">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:center;">
orig.ident
</th>
<th style="text-align:center;">
nCount_Spatial
</th>
<th style="text-align:center;">
nFeature_Spatial
</th>
<th style="text-align:center;">
sample
</th>
<th style="text-align:center;">
x_coord
</th>
<th style="text-align:center;">
y_coord
</th>
<th style="text-align:center;">
ssc
</th>
<th style="text-align:center;">
Binned_GPs
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
1_1
</td>
<td style="text-align:center;">
SeuratProject
</td>
<td style="text-align:center;">
54142.66
</td>
<td style="text-align:center;">
32
</td>
<td style="text-align:center;">
mouse-bladder-peaks
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
4
</td>
<td style="text-align:center;">
0.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
2_1
</td>
<td style="text-align:center;">
SeuratProject
</td>
<td style="text-align:center;">
33354.74
</td>
<td style="text-align:center;">
12
</td>
<td style="text-align:center;">
mouse-bladder-peaks
</td>
<td style="text-align:center;">
2
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
4
</td>
<td style="text-align:center;">
0.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
3_1
</td>
<td style="text-align:center;">
SeuratProject
</td>
<td style="text-align:center;">
62429.26
</td>
<td style="text-align:center;">
59
</td>
<td style="text-align:center;">
mouse-bladder-peaks
</td>
<td style="text-align:center;">
3
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
4
</td>
<td style="text-align:center;">
415.6588
</td>
</tr>
<tr>
<td style="text-align:left;">
4_1
</td>
<td style="text-align:center;">
SeuratProject
</td>
<td style="text-align:center;">
75365.22
</td>
<td style="text-align:center;">
89
</td>
<td style="text-align:center;">
mouse-bladder-peaks
</td>
<td style="text-align:center;">
4
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
4
</td>
<td style="text-align:center;">
0.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
5_1
</td>
<td style="text-align:center;">
SeuratProject
</td>
<td style="text-align:center;">
75335.54
</td>
<td style="text-align:center;">
80
</td>
<td style="text-align:center;">
mouse-bladder-peaks
</td>
<td style="text-align:center;">
5
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
4
</td>
<td style="text-align:center;">
0.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
6_1
</td>
<td style="text-align:center;">
SeuratProject
</td>
<td style="text-align:center;">
63557.09
</td>
<td style="text-align:center;">
61
</td>
<td style="text-align:center;">
mouse-bladder-peaks
</td>
<td style="text-align:center;">
6
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
4
</td>
<td style="text-align:center;">
425.8180
</td>
</tr>
</tbody>
</table>
<p>We can see the expression values have been stored in the
<code>@meta.data</code> column “Binnde_GPs”. Now lets plot the
results:</p>
<div class="sourceCode" id="section-cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/ImageFeaturePlot.html" class="external-link">ImageFeaturePlot</a></span><span class="op">(</span><span class="va">bladder_annotated</span>, features <span class="op">=</span> <span class="st">"Binned_GPs"</span>, size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_reverse</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_fill_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="SpaMTP_Additional_Features_files/figure-html/unnamed-chunk-25-1.png" width="80%" style="display: block; margin: auto;"></p>
<p>We can see that there is high expression of GP’s in the urothelium
and adventitia layers of the bladder. This feature is especially useful
to combine metabolites that are all up expressed in a specific
condition/cluster (following differential expression analysis).</p>
</div>
<div class="section level2">
<h2 id="section-adding-he-image-to-spatial-metabolomic-data">Adding H&amp;E Image to Spatial Metabolomic Data<a class="anchor" aria-label="anchor" href="#section-adding-he-image-to-spatial-metabolomic-data"></a>
</h2>
<p>One useful feature of <strong><em>SpaMPT</em></strong> is the ability
to add a H&amp;E or tissue image to the <em>SpaMTP</em> object. This can
allow the user to visualise key structural metabolic patterns that match
to the tissue morphology. Below, we will demonstrate how to manually
align your image with your SM dataset.</p>
<div class="sourceCode" id="section-cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bladder_HE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddSMImage.html">AddSMImage</a></span><span class="op">(</span><span class="st">"./mouse_urinary_bladder_HE.tif"</span>, <span class="va">bladder_annotated</span>, msi.pixel.multiplier <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p>This function will activate an interactive window where you can
change the MSI coordinates to align to the image provided. This includes
rotating, adjusting x and y coordinates, flipping around the x or y
axis, and also scaling. The user can also select any column in the
object’s metadata to plot. This can help assist the alignment process,
in particular by using clustering results.</p>
<p>Here is an example of what the alignment GUI looks like, and also the
perimeters we used to align the H&amp;E image:</p>
<p><img src="vignette_data_files/Mouse_Urinary_Bladder/H%26E_Alignment.png" width="80%" style="display: block; margin: auto;"></p>
<p>After clicking the return aligned data button a
<strong><em>SpaMTP</em></strong> <em>Seurat</em> object with the H&amp;E
image added to the <code>@image$slice1</code> slot will be returned.
Here “fov” is the original SM plotting coordinates and “slice1” has the
adjusted coordinates to match the H&amp;E image.</p>
<div class="sourceCode" id="section-cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bladder_HE</span><span class="op">@</span><span class="va">images</span></span></code></pre></div>
<pre><code><span><span class="co">## $fov</span></span>
<span><span class="co">## Spatial coordinates for 34840 cells</span></span>
<span><span class="co">## Default segmentation boundary: centroids </span></span>
<span><span class="co">## Associated assay: Spatial </span></span>
<span><span class="co">## Key: Spatial_ </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $slice1</span></span>
<span><span class="co">## Spatial coordinates for 34840 cells</span></span>
<span><span class="co">## Default segmentation boundary: centroids </span></span>
<span><span class="co">## Associated assay: Spatial </span></span>
<span><span class="co">## Key: slice_</span></span></code></pre>
<p>Lets now visualise this:</p>
<div class="sourceCode" id="section-cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="fu"><a href="https://satijalab.org/seurat/reference/ImageDimPlot.html" class="external-link">ImageDimPlot</a></span><span class="op">(</span><span class="va">bladder_HE</span>, group.by <span class="op">=</span> <span class="st">"ssc"</span>, fov <span class="op">=</span> <span class="st">"fov"</span>, size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_reverse</a></span><span class="op">(</span><span class="op">)</span> <span class="op">)</span><span class="op">/</span> </span>
<span>  <span class="op">(</span><span class="fu"><a href="https://satijalab.org/seurat/reference/SpatialPlot.html" class="external-link">SpatialDimPlot</a></span><span class="op">(</span><span class="va">bladder_HE</span>, group.by <span class="op">=</span> <span class="st">"ssc"</span>, images <span class="op">=</span> <span class="st">"slice1"</span>, pt.size.factor <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">&amp;</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="SpaMTP_Additional_Features_files/figure-html/unnamed-chunk-30-1.png" width="80%" style="display: block; margin: auto;"></p>
<p>This also means that we can now visualise our data using the
<em>Plot3D()</em> function with a H&amp;E image:</p>
<div class="sourceCode" id="section-cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Identified the m/z values that are to be plot</span></span>
<span><span class="va">mzs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">798.5411</span>, <span class="fl">741.5306</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="../reference/FindNearestMZ.html">FindNearestMZ</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">bladder_annotated</span>, target_mz <span class="op">=</span> <span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ROI_HE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">bladder_HE</span>, subset <span class="op">=</span> <span class="va">ssc</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"2"</span>, <span class="st">"5"</span>, <span class="st">"6"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="section-cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Plot3DFeature.html">Plot3DFeature</a></span><span class="op">(</span><span class="va">ROI_HE</span>, features <span class="op">=</span> <span class="va">mzs</span>, show.image <span class="op">=</span> <span class="st">"slice1"</span>, assays <span class="op">=</span> <span class="st">"Spatial"</span>, between.layer.height <span class="op">=</span> <span class="fl">300</span>, names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PC(34:1)"</span>,<span class="st">"SM(34:1)"</span><span class="op">)</span>, plot.height <span class="op">=</span> <span class="fl">400</span>, plot.width <span class="op">=</span> <span class="fl">700</span>, downscale.image <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<iframe src="vignette_data_files/Mouse_Urinary_Bladder/3d_plot1.html" width="100%" height="400px">
</iframe>
<p>This function is designed for plotting two features simultaneously.
If we only want to visualise one of the features, we can use a work
around to only show a single trace, where trace 1 = feature1, trace 2 =
feature2 and trace 3 = image. Here, we just provide one feature to plot.
To stop this one feature being replicated, we can add an additional line
of code shown below:</p>
<div class="sourceCode" id="section-cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Plot3DFeature.html">Plot3DFeature</a></span><span class="op">(</span><span class="va">ROI_HE</span>, features <span class="op">=</span> <span class="va">mzs</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, show.image <span class="op">=</span> <span class="st">"slice1"</span>, assays <span class="op">=</span> <span class="st">"Spatial"</span>, between.layer.height <span class="op">=</span> <span class="fl">300</span>, names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SM(34:1)"</span>, <span class="st">"SM(34:1)"</span><span class="op">)</span>, plot.height <span class="op">=</span> <span class="fl">400</span>, plot.width <span class="op">=</span> <span class="fl">700</span>, downscale.image <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plotly</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/plotly/man/style.html" class="external-link">style</a></span><span class="op">(</span><span class="va">p</span>, visible <span class="op">=</span> <span class="cn">FALSE</span>, traces <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<iframe src="vignette_data_files/Mouse_Urinary_Bladder/3d_plot2.html" width="100%" height="400px">
</iframe>
</div>
<div class="section level2">
<h2 id="section-d-ploting-of-features-and-metadata">3D Ploting of features and metadata<a class="anchor" aria-label="anchor" href="#section-d-ploting-of-features-and-metadata"></a>
</h2>
<p>Although the Plot3DFeatures function is mainly used for plotting
features from different modalities (i.e. genes and metabolites). It can
also be used to plot metadata with feature or metadata between
modalities. Lets visualise this below:</p>
<div class="sourceCode" id="section-cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot3DFeatures can either take in a colour palette (i.e. "Reds", "Blues", etc.) or a list of RGB values</span></span>
<span><span class="co"># In this case we want our feature to be plotted using "Reds" and our clustering results to be coloured in categorical colours </span></span>
<span></span>
<span><span class="va">col.palette</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="st">"Reds"</span>,  <span class="co"># Use built-in Plotly colorscale</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>     <span class="co"># Define custom colorscale manually</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">0</span>, <span class="st">"rgb(0,0,255)"</span><span class="op">)</span>,   <span class="co"># Blue at 0</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="st">"rgb(0,255,0)"</span><span class="op">)</span>, <span class="co"># Green at 0.5</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">1</span>, <span class="st">"rgb(255,0,0)"</span><span class="op">)</span>    <span class="co"># Red at 1</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="../reference/Plot3DFeature.html">Plot3DFeature</a></span><span class="op">(</span><span class="va">ROI_HE</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">mzs</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="st">"ssc"</span><span class="op">)</span>, show.image <span class="op">=</span> <span class="st">"slice1"</span>, assays <span class="op">=</span> <span class="st">"Spatial"</span>, between.layer.height <span class="op">=</span> <span class="fl">300</span>, names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Metabolite"</span>, <span class="st">"Clusters"</span><span class="op">)</span>, plot.height <span class="op">=</span> <span class="fl">400</span>, plot.width <span class="op">=</span> <span class="fl">700</span>, downscale.image <span class="op">=</span> <span class="fl">2</span>, col.palette <span class="op">=</span> <span class="va">col.palette</span><span class="op">)</span></span></code></pre></div>
<iframe src="vignette_data_files/Mouse_Urinary_Bladder/3d_plot3.html" width="100%" height="400px">
</iframe>
<p>If we want to specify the exact colours for our clusters we can use
the code below to convert colours/hex values to RGB:</p>
<div class="sourceCode" id="section-cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># We will use the same colour palette as used in the "Mouse Urinary Bladder" tutorial</span></span>
<span><span class="co"># Make sure that you only include values that are present in your plot</span></span>
<span><span class="va">hex_list</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"2"</span> <span class="op">=</span> <span class="st">"#C2B03B"</span>,</span>
<span>                   <span class="st">"5"</span> <span class="op">=</span> <span class="st">"#0074B0"</span>,  </span>
<span>                   <span class="st">"6"</span> <span class="op">=</span> <span class="st">"#DE4D6C"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Convert the names to numeric values</span></span>
<span><span class="va">values</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">hex_list</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Normalize values between 0 and 1</span></span>
<span><span class="va">normalized_values</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">values</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">values</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">values</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">values</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Convert colors to RGB (supports both hex and named colors)</span></span>
<span><span class="va">custom_colorscale</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">values</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">rgb_values</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/col2rgb.html" class="external-link">col2rgb</a></span><span class="op">(</span><span class="va">hex_list</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">values</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>  <span class="co"># Convert to RGB</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">normalized_values</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"rgb("</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">rgb_values</span>, collapse<span class="op">=</span><span class="st">","</span><span class="op">)</span>, <span class="st">")"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="../reference/Plot3DFeature.html">Plot3DFeature</a></span><span class="op">(</span><span class="va">ROI_HE</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">mzs</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="st">"ssc"</span><span class="op">)</span>, show.image <span class="op">=</span> <span class="st">"slice1"</span>, assays <span class="op">=</span> <span class="st">"Spatial"</span>, between.layer.height <span class="op">=</span> <span class="fl">300</span>, names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Metabolite"</span>, <span class="st">"Clusters"</span><span class="op">)</span>, plot.height <span class="op">=</span> <span class="fl">400</span>, plot.width <span class="op">=</span> <span class="fl">700</span>, downscale.image <span class="op">=</span> <span class="fl">2</span>, col.palette <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Reds"</span>, <span class="va">custom_colorscale</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<iframe src="vignette_data_files/Mouse_Urinary_Bladder/3d_plot4.html" width="100%" height="400px">
</iframe>
</div>
<div class="section level2">
<h2 id="section-selecting-data-bin-size">Selecting Data Bin Size<a class="anchor" aria-label="anchor" href="#section-selecting-data-bin-size"></a>
</h2>
<p>One important element in analysing spatial metabolic data is choosing
a bin size to reduce the dimensionality of the dataset. SpaMTP provides
an interactive plot that lets you change the size of the bin and see
which metabolites are included within that bin for a specific peak, and
how that changes the expression spatially across the tissue. We can
observe this <a href="https://agc888.shinyapps.io/InteractiveSpatialPlot/" class="external-link">below</a>:</p>
<div class="sourceCode" id="section-cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/InteractiveSpatialPlot.html">InteractiveSpatialPlot</a></span><span class="op">(</span><span class="va">ROI_HE</span><span class="op">)</span></span></code></pre></div>
<iframe src="https://agc888.shinyapps.io/InteractiveSpatialPlot/" width="100%" height="600px">
</iframe>
<p>Based on this plot, users can adjust the bin size of their data (by
running <code>BinSpaMTP</code>) or can adjust the <code>plusminus</code>
attribute when using <code>ImageMZPlot</code>.</p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="section-data-export">Data Export<a class="anchor" aria-label="anchor" href="#section-data-export"></a>
</h2>
<p>Exporting data can often be difficult when dealing with large
datasets of different data object types such as <em>R’s Seurat</em>
compared to <em>Python’s scanpy</em>. <strong><em>SpaMTP</em></strong>
can easily export the data object into matrix.mtx, barcodes.csv and
features.csv files, which can be easily read into all major spatial
analysis packages. In addition, cell/pixel and feature metadata is also
exported as .csv files. Below we demonstrate how to export the
files:</p>
<div class="sourceCode" id="section-cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/SaveSpaMTPData.html">SaveSpaMTPData</a></span><span class="op">(</span><span class="va">bladder_HE</span>, outdir <span class="op">=</span> <span class="st">"../Documents/SpaMTP_Output/"</span>, assay <span class="op">=</span> <span class="st">"Spatial"</span>, slot <span class="op">=</span> <span class="st">"counts"</span>, annotations <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Lets visualise the output files directory structure:</p>
<div class="sourceCode" id="section-cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/dir_tree.html" class="external-link">dir_tree</a></span><span class="op">(</span><span class="st">"../Documents/SpaMTP_Output/"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## ../Documents/SpaMTP_Output/</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ├── barcode_metadata.csv</span></span>
<span><span class="co">## ├── feature_metadata.csv</span></span>
<span><span class="co">## ├── filtered_feature_bc_matrix</span></span>
<span><span class="co">## │   ├── barcodes.tsv</span></span>
<span><span class="co">## │   ├── genes.tsv</span></span>
<span><span class="co">## │   └── matrix.mtx</span></span>
<span><span class="co">## └── spatial</span></span>
<span><span class="co">##     ├── scalefactors_json.json</span></span>
<span><span class="co">##     ├── tissue_lowres_image.png</span></span>
<span><span class="co">##     └── tissue_positions_list.csv</span></span></code></pre>
<p>This data can therefor be loaded into other applications such as
<em>AnnData</em> in python:</p>
<div class="sourceCode" id="section-cb38"><pre class="sourceCode python"><code class="sourceCode python"><span id="section-cb38-1"><a href="#section-cb38-1" tabindex="-1"></a><span class="im">import</span> anndata</span>
<span id="section-cb38-2"><a href="#section-cb38-2" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="section-cb38-3"><a href="#section-cb38-3" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="section-cb38-4"><a href="#section-cb38-4" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="section-cb38-5"><a href="#section-cb38-5" tabindex="-1"></a></span>
<span id="section-cb38-6"><a href="#section-cb38-6" tabindex="-1"></a>adata <span class="op">=</span> sc.read_10x_mtx(path <span class="op">=</span> <span class="st">"../Documents/SpaMTP_Output/"</span>)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="section-session-info">Session Info<a class="anchor" aria-label="anchor" href="#section-session-info"></a>
</h2>
<div class="sourceCode" id="section-cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">## Running under: macOS 15.5</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Europe/Dublin</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] EnhancedVolcano_1.24.0 ggrepel_0.9.6          ggplot2_4.0.0         </span></span>
<span><span class="co">##  [4] dplyr_1.1.4            Seurat_5.3.0           SeuratObject_5.2.0    </span></span>
<span><span class="co">##  [7] sp_2.2-0               Cardinal_3.8.3         S4Vectors_0.44.0      </span></span>
<span><span class="co">## [10] ProtGenerics_1.38.0    BiocGenerics_0.52.0    BiocParallel_1.40.2   </span></span>
<span><span class="co">## [13] SpaMTP_1.1.0          </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3     rstudioapi_0.17.1      jsonlite_2.0.0        </span></span>
<span><span class="co">##   [4] magrittr_2.0.4         spatstat.utils_3.2-0   farver_2.1.2          </span></span>
<span><span class="co">##   [7] rmarkdown_2.29         fs_1.6.6               ragg_1.5.0            </span></span>
<span><span class="co">##  [10] vctrs_0.6.5            ROCR_1.0-11            spatstat.explore_3.5-3</span></span>
<span><span class="co">##  [13] RCurl_1.98-1.17        htmltools_0.5.8.1      sass_0.4.10           </span></span>
<span><span class="co">##  [16] sctransform_0.4.2      parallelly_1.45.1      KernSmooth_2.23-26    </span></span>
<span><span class="co">##  [19] bslib_0.9.0            htmlwidgets_1.6.4      desc_1.4.3            </span></span>
<span><span class="co">##  [22] ica_1.0-3              plyr_1.8.9             plotly_4.11.0         </span></span>
<span><span class="co">##  [25] zoo_1.8-14             cachem_1.1.0           igraph_2.1.4          </span></span>
<span><span class="co">##  [28] mime_0.13              lifecycle_1.0.4        pkgconfig_2.0.3       </span></span>
<span><span class="co">##  [31] Matrix_1.7-4           R6_2.6.1               fastmap_1.2.0         </span></span>
<span><span class="co">##  [34] fitdistrplus_1.2-4     future_1.67.0          shiny_1.11.1          </span></span>
<span><span class="co">##  [37] digest_0.6.37          patchwork_1.3.2        tensor_1.5.1          </span></span>
<span><span class="co">##  [40] RSpectra_0.16-2        irlba_2.3.5.1          textshaping_1.0.3     </span></span>
<span><span class="co">##  [43] labeling_0.4.3         progressr_0.16.0       spatstat.sparse_3.1-0 </span></span>
<span><span class="co">##  [46] httr_1.4.7             polyclip_1.10-7        abind_1.4-8           </span></span>
<span><span class="co">##  [49] compiler_4.4.1         proxy_0.4-27           withr_3.0.2           </span></span>
<span><span class="co">##  [52] S7_0.2.0               tiff_0.1-12            viridis_0.6.5         </span></span>
<span><span class="co">##  [55] DBI_1.2.3              fastDummies_1.7.5      MASS_7.3-65           </span></span>
<span><span class="co">##  [58] classInt_0.4-11        units_0.8-7            tools_4.4.1           </span></span>
<span><span class="co">##  [61] lmtest_0.9-40          httpuv_1.6.16          future.apply_1.20.0   </span></span>
<span><span class="co">##  [64] goftest_1.2-3          glue_1.8.0             nlme_3.1-168          </span></span>
<span><span class="co">##  [67] EBImage_4.48.0         promises_1.3.3         sf_1.0-21             </span></span>
<span><span class="co">##  [70] grid_4.4.1             Rtsne_0.17             cluster_2.1.8.1       </span></span>
<span><span class="co">##  [73] reshape2_1.4.4         generics_0.1.4         gtable_0.3.6          </span></span>
<span><span class="co">##  [76] spatstat.data_3.1-8    class_7.3-23           tidyr_1.3.1           </span></span>
<span><span class="co">##  [79] data.table_1.17.8      spatstat.geom_3.6-0    RcppAnnoy_0.0.22      </span></span>
<span><span class="co">##  [82] RANN_2.6.2             pillar_1.11.1          stringr_1.5.2         </span></span>
<span><span class="co">##  [85] spam_2.11-1            RcppHNSW_0.6.0         limma_3.62.2          </span></span>
<span><span class="co">##  [88] later_1.4.4            splines_4.4.1          lattice_0.22-7        </span></span>
<span><span class="co">##  [91] survival_3.8-3         deldir_2.0-4           tidyselect_1.2.1      </span></span>
<span><span class="co">##  [94] CardinalIO_1.4.0       locfit_1.5-9.12        miniUI_0.1.2          </span></span>
<span><span class="co">##  [97] pbapply_1.7-4          knitr_1.50             gridExtra_2.3         </span></span>
<span><span class="co">## [100] matter_2.8.0           scattermore_1.2        xfun_0.53             </span></span>
<span><span class="co">## [103] Biobase_2.66.0         statmod_1.5.0          matrixStats_1.5.0     </span></span>
<span><span class="co">## [106] fftwtools_0.9-11       stringi_1.8.7          lazyeval_0.2.2        </span></span>
<span><span class="co">## [109] yaml_2.3.10            evaluate_1.0.5         codetools_0.2-20      </span></span>
<span><span class="co">## [112] tibble_3.3.0           cli_3.6.5              ontologyIndex_2.12    </span></span>
<span><span class="co">## [115] uwot_0.2.3             xtable_1.8-4           reticulate_1.43.0     </span></span>
<span><span class="co">## [118] systemfonts_1.2.3      jquerylib_0.1.4        Rcpp_1.1.0            </span></span>
<span><span class="co">## [121] globals_0.18.0         spatstat.random_3.4-2  zeallot_0.2.0         </span></span>
<span><span class="co">## [124] png_0.1-8              spatstat.univar_3.1-4  parallel_4.4.1        </span></span>
<span><span class="co">## [127] pkgdown_2.1.3          dotCall64_1.2          jpeg_0.1-11           </span></span>
<span><span class="co">## [130] bitops_1.0-9           listenv_0.9.1          viridisLite_0.4.2     </span></span>
<span><span class="co">## [133] e1071_1.7-16           scales_1.4.0           ggridges_0.5.7        </span></span>
<span><span class="co">## [136] purrr_1.1.0            rlang_1.1.6            cowplot_1.2.0         </span></span>
<span><span class="co">## [139] shinyjs_2.1.0</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/agc888" class="external-link">Andrew Causer</a>, <a href="https://github.com/BCRL-tylu" class="external-link">Tianyao Lu</a>, <a href="https://github.com/ChemCharles" class="external-link">Christopher C.J. Fitzgerald</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
