#' Plot Set Enrichment Analysis Results from `SpaMTP::FindRegionalPathways()`.
#'
#' This function generates a plot for visualizing set enrichment analysis (SEA) results 
#' based on the output from the `SpaMTP::FindRegionalPathways()` function.
#'
#' @param SpaMTP A `SpaMTP` Seurat object containing spatial metabolomics (SM) and/or spatial transcriptomics (ST) data. If SM data is included, it must be annotated using the `SpaMTP::AnnotateSM()` function.
#' @param ident A character string specifying the cluster identifier used to group regions, corresponding to a column name in the `SpaMTP@meta.data` slot.
#' @param DE.list A list containing differential expression results from the `FindAllMarkers()` function, with items matching the order of the `analyte_types` argument.
#' @param regpathway A dataframe generated by the `SpaMTP::FindRegionalPathways()` function, containing identified regional pathways.
#' @param selected_pathways A character vector specifying the names or IDs of pathways to be included in the analysis (e.g., `c("Amino acid metabolism", "WP1902", "Aspartate and asparagine metabolism")`). This argument is not case-sensitive.
#' @param pval_cutoff_pathway A numeric value between 0 and 1, defining the cutoff for the adjusted p-value from the permutation test used to compute significant pathways. 
#' @param num_display An integer specifying the number of pathways to display in the plot.
#' @param text_size A numeric value controlling the size of the text elements in the plot.
#'
#' @return A `ggplot` object representing the set enrichment analysis results.
#' @export
#'
#' @examples
#' plotsea(SpaMTP, ident = "Custom_ident", regpathway = regpathway, 
#'         selected_pathways = selected_pathways, pval_cutoff_pathway = NULL, 
#'         num_display = NULL, text_size = NULL)

plotsea = function(SpaMTP,
                   ident,
                   regpathway,
                   selected_pathways = NULL,
                   pval_cutoff_pathway = NULL,
                   num_display = NULL,
                   text_size = NULL) {
  ## Checks for ident in SpaMTP Object
  if (!(ident %in% colnames(SpaMTP@meta.data))) {
    stop(
      "Ident: ",
      ident,
      " not found in SpaMTP object's @meta.data slot ... Make sure the ident column is in your @metadata and is a factor!"
    )
  }
  
  clus = as.factor(gsub("\\,.*", "", SpaMTP@meta.data[[ident]]))
  coordnate = sapply(
    SpaMTP@meta.data[, which(grepl(
      colnames(SpaMTP@meta.data),
      pattern = "coord",
      ignore.case = T
    ))],
    FUN = function(x) {
      as.numeric(gsub("\\,.*", "", x))
    }
  )
  palette <- suppressWarnings({
    brewer.pal(length(levels(clus)), "Set3")
  })
  if (length(levels(clus)) > length(palette)) {
    palette <- colorRampPalette(brewer.pal(9, "Set3"))(length(levels(clus)))
  }
  names(palette) = levels(clus)
  # SPM coordiante
  uid = unique(regpathway$Cluster_id)
  clu_names = data.frame(cluster = levels(clus),
                         clu_name = paste0("Cluster", levels(clus)))
  colnames(regpathway)[1] = "pathwayName"
  regpathway = regpathway %>% dplyr::group_by(pathwayName) %>% dplyr::mutate(
    Significance = ifelse(
      pval <= 0.05,
      "Significant at 5% significance level",
      "Not statistically significant"
    )
  ) %>% dplyr::mutate(group_importance = sum(abs(NES)))
  if (is.null(selected_pathways)) {
    regpathway = regpathway %>%  dplyr::filter(group_importance %in% sort(unique(regpathway$group_importance), decreasing = T)[1:(num_display %||% min(10, length(unique(
      regpathway$pathwayName
    ))))])
  } else{
    regpathway = regpathway %>% dplyr::filter((pathwayName %in% selected_pathways) |
                                                (sourceId %in% selected_pathways))
    regpathway = regpathway %>%  dplyr::filter(group_importance %in% sort(unique(regpathway$group_importance), decreasing = T)[1:(num_display %||% min(10, length(unique(
      regpathway$pathwayName
    ))))])
  }
  
  regpathway = regpathway %>% dplyr::mutate(pathnameid = paste0(pathwayName, "(", sourceId, ")"))
  pathwaynames = unique(regpathway$pathnameid)
  n = length(unique(regpathway$pathnameid))
  jaccard_matrix = matrix(nrow = n, ncol = n)
  colnames(jaccard_matrix) = rownames(jaccard_matrix) = pathwaynames
  verbose_message(message_text = "computing jaccard distance between pathways" , verbose = verbose)
  for (i in 1:(n - 1)) {
    pathway_id_i = sub(".*\\(([^)]+)\\).*", "\\1", pathwaynames[i])
    pathway_content_i = unique(analytehaspathway$rampId[which(analytehaspathway$pathwayRampId == pathway$pathwayRampId[which(pathway$sourceId == pathway_id_i)])])
    for (j in (i + 1):n) {
      pathway_id_j = sub(".*\\(([^)]+)\\).*", "\\1", pathwaynames[j])
      pathway_content_j = unique(analytehaspathway$rampId[which(analytehaspathway$pathwayRampId == pathway$pathwayRampId[which(pathway$sourceId == pathway_id_j)])])
      jc_simi = length(intersect(pathway_content_i, pathway_content_j)) / length(union(pathway_content_i, pathway_content_j))
      jaccard_matrix[i, j] = jaccard_matrix[j, i] = jc_simi
    }
  }
  diag(jaccard_matrix) = 1
  # Generate a dendrogram
  hc <- as.dendrogram(hclust(as.dist(jaccard_matrix)))
  # dendro <- ggtree(as.phylo(hc), layout = "rectangular")+scale_x_reverse()
  segment_hc <- with(ggdendro::segment(dendro_data(hc)),
                     data.frame(
                       x = y,
                       y = x,
                       xend = yend,
                       yend = xend
                     ))
  pos_table <- with(dendro_data(hc)$labels,
                    data.frame(
                      y_center = x,
                      path = as.character(label),
                      height = 1
                    ))
  
  axis_limits <- with(pos_table, c(min(y_center - 0.5 * height), max(y_center + 0.5 * height))) + 0.1 * c(-1, 1)
  plt_dendr <- ggplot(segment_hc) +
    geom_segment(aes(
      x = sqrt(x),
      y = y,
      xend = sqrt(xend),
      yend = yend
    )) +
    scale_x_continuous(expand = c(0, 0.1), limits = c(0, sqrt(max(segment_hc$xend)))) +
    scale_y_continuous(
      expand = c(0, 0),
      breaks = pos_table$y_center,
      labels = str_wrap(pos_table$path, width = 40),
      limits = axis_limits,
      position = "right"
    ) +
    labs(
      x = "Jacard distance",
      y = "",
      colour = "",
      size = ""
    ) +
    theme_bw() +
    theme(panel.grid.minor = element_blank(),
          axis.text = element_text(size = ((text_size) %||% 12) - 3))
  #### ggplot
  suppressWarnings({
    gg_dot = ggplot(data = regpathway, aes(
      x = factor(Cluster_id, levels = sort(unique(Cluster_id))),
      y = factor(pathnameid, levels = dendro_data(hc)$labels$label)
    )) +
      geom_point(aes(colour = as.numeric(NES), size = as.numeric(sqrt(size)) * 5.1)) +
      scale_colour_gradient2(
        name = "Normalised enrichment score",
        low = "blue",
        # Original low color
        high = "red",
        # Original high color
        limits = c(min(regpathway$NES), max(regpathway$NES))
      ) +
      scale_size_continuous(name = "Number of altered metabolites \n in the pathway") +
      new_scale_colour() +
      geom_point(shape = 1,
                 aes(colour = Significance, size = as.numeric(size) * 2.1 +
                       0.1)) +
      scale_color_manual(
        values = c(
          "Significant at 5% significance level" = "green",
          "Not statistically significant" = "black"
        )
      ) +
      #labs(title = "Comparason of pathways expression between different cluster", y = "Pathways", x = "Clusters") +
      theme(
        title = element_text(size = text_size %||% 12, face = 'bold'),
        axis.text.x = element_text(
          size = text_size %||% 12,
          angle = 90,
          vjust = 0.5,
          hjust = 1,
        ),
        #20
        axis.title = element_text(size = text_size %||% 12),
        legend.key.size = unit(1, 'cm'),
        legend.title = element_text(size = ((text_size) %||% 12) - 3),
        legend.text = element_text(size = ((text_size) %||% 12) - 3)
      ) +
      new_scale_colour() + theme(
        panel.background = element_rect(fill = "white"),
        panel.grid = element_line(color = "grey")
      ) +   theme(
        legend.position = "left",
        axis.text.y = element_blank(),
        #axis.text.x = element_blank(),
        #axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        legend.direction = "horizontal"
      ) + xlab("Clusters")
  })
  u = ggarrange(gg_dot,
                plt_dendr,
                ncol = 2,
                widths = c(1, 0.6))
  print(u)
  svglite::svglite(file = "~/figure.svg",
                   width = 12,
                   height = 7)
  print(u)
  dev.off()
  return(u)
}
